// @remotion/studio-server/dist/routes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleRoutes = void 0;
const fs_1 = __importStar(require("fs"));
const node_fs_1 = require("node:fs");
const node_path_1 = __importStar(require("node:path"));
const node_url_1 = require("node:url");
const bundler_1 = require("@remotion/bundler");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const client_render_queue_1 = require("./client-render-queue");
const get_file_source_1 = require("./helpers/get-file-source");
const get_installed_installable_packages_1 = require("./helpers/get-installed-installable-packages");
const resolve_output_path_1 = require("./helpers/resolve-output-path");
const api_routes_1 = require("./preview-server/api-routes");
const element_install_state_1 = require("./preview-server/element-install-state");
const get_package_manager_1 = require("./preview-server/get-package-manager");
const get_static_file_fallback_hint_1 = require("./preview-server/get-static-file-fallback-hint");
const handler_1 = require("./preview-server/handler");
const parse_body_1 = require("./preview-server/parse-body");
const public_folder_1 = require("./preview-server/public-folder");
const open_in_editor_1 = require("./preview-server/routes/open-in-editor");
const serve_static_1 = require("./preview-server/serve-static");
const validate_same_origin_1 = require("./preview-server/validate-same-origin");
const watch_ignore_next_change_1 = require("./preview-server/watch-ignore-next-change");
const loggedStaticFileHints = new Set();
const ELEMENT_INSTALL_FOCUS_MAX_AGE = 5 * 60 * 1000;
const ELEMENT_INSTALL_TARGET_RESPONSE_WAIT = 250;
const static404 = (response) => {
    response.writeHead(404);
    response.end('The static/ prefix has been changed, this URL is no longer valid.');
    return Promise.resolve();
};
const output404 = (response) => {
    response.writeHead(404);
    response.end('The outputs/ prefix has been changed, this URL is no longer valid.');
    return Promise.resolve();
};
const handleRemotionConfig = (response, remotionRoot) => {
    var _a;
    response.writeHead(200, {
        'Content-Type': 'application/json',
    });
    const body = {
        isRemotion: true,
        cwd: remotionRoot,
        version: (_a = process.env.REMOTION_VERSION) !== null && _a !== void 0 ? _a : null,
    };
    response.end(JSON.stringify(body));
    return Promise.resolve();
};
const isAllowedElementInstallOrigin = (origin) => {
    if (!origin) {
        return false;
    }
    try {
        const url = new URL(origin);
        return ((url.protocol === 'https:' &&
            (url.hostname === 'remotion.dev' ||
                url.hostname === 'www.remotion.dev')) ||
            (url.protocol === 'http:' &&
                (url.hostname === 'localhost' || url.hostname === '127.0.0.1')));
    }
    catch (_a) {
        return false;
    }
};
const setElementInstallCorsHeaders = ({ request, response, }) => {
    const { origin } = request.headers;
    if (isAllowedElementInstallOrigin(origin)) {
        response.setHeader('Access-Control-Allow-Origin', origin);
        response.setHeader('Vary', 'Origin');
        response.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
        response.setHeader('Access-Control-Allow-Headers', 'Content-Type');
        response.setHeader('Access-Control-Max-Age', '600');
        response.setHeader('Access-Control-Allow-Private-Network', 'true');
    }
};
const handleElementInstallOptions = ({ request, response, }) => {
    setElementInstallCorsHeaders({ request, response });
    response.writeHead(isAllowedElementInstallOrigin(request.headers.origin) ? 204 : 403);
    response.end();
    return Promise.resolve();
};
const handleElementInstallTarget = ({ liveEventsServer, request, response, remotionRoot, gitSource, }) => {
    if (!isAllowedElementInstallOrigin(request.headers.origin)) {
        response.writeHead(403);
        response.end(JSON.stringify({ success: false, reason: 'Origin not allowed' }));
        return Promise.resolve();
    }
    const requestId = `${Date.now()}-${Math.random().toString(16).slice(2)}`;
    liveEventsServer.sendEventToClient({
        type: 'request-element-install-target',
        requestId,
    });
    return new Promise((resolve) => {
        setTimeout(() => {
            var _a, _b, _c, _d, _e;
            const target = (0, element_install_state_1.getElementInstallTarget)(requestId);
            const now = Date.now();
            const targetIsLive = target !== null &&
                now - target.updatedAt < element_install_state_1.ELEMENT_INSTALL_TARGET_MAX_AGE;
            const host = (_a = request.headers.host) !== null && _a !== void 0 ? _a : null;
            const port = (_b = host === null || host === void 0 ? void 0 : host.split(':').at(-1)) !== null && _b !== void 0 ? _b : null;
            setElementInstallCorsHeaders({ request, response });
            response.writeHead(200, { 'Content-Type': 'application/json' });
            response.end(JSON.stringify({
                type: 'remotion-studio',
                projectName: (0, studio_shared_1.getProjectName)({
                    basename: node_path_1.default.basename,
                    gitSource,
                    resolvedRemotionRoot: remotionRoot,
                }),
                port: port === null ? null : Number(port),
                lastFocusedAt: (_c = target === null || target === void 0 ? void 0 : target.lastFocusedAt) !== null && _c !== void 0 ? _c : null,
                canInstall: target !== null && target.canInstall && targetIsLive,
                activeCompositionId: (_d = target === null || target === void 0 ? void 0 : target.compositionId) !== null && _d !== void 0 ? _d : null,
                readOnly: (_e = target === null || target === void 0 ? void 0 : target.readOnly) !== null && _e !== void 0 ? _e : false,
            }));
            resolve();
        }, ELEMENT_INSTALL_TARGET_RESPONSE_WAIT);
    });
};
const handleRequestElementInstall = async ({ liveEventsServer, request, response, }) => {
    if (!isAllowedElementInstallOrigin(request.headers.origin)) {
        response.writeHead(403);
        response.end(JSON.stringify({ success: false, reason: 'Origin not allowed' }));
        return;
    }
    setElementInstallCorsHeaders({ request, response });
    response.setHeader('Content-Type', 'application/json');
    try {
        const body = await (0, parse_body_1.parseRequestBody)(request);
        const parsed = (0, studio_shared_1.parseElementDragData)(JSON.stringify({
            type: 'remotion-element',
            version: 1,
            element: body.element,
        }));
        if (parsed === null) {
            response.writeHead(400);
            response.end(JSON.stringify({ success: false, reason: 'Invalid Element payload' }));
            return;
        }
        const target = (0, element_install_state_1.getElementInstallTarget)(null);
        const now = Date.now();
        const targetIsLive = target !== null &&
            now - target.updatedAt < element_install_state_1.ELEMENT_INSTALL_TARGET_MAX_AGE;
        const targetWasRecentlyFocused = target !== null &&
            target.lastFocusedAt !== null &&
            now - target.lastFocusedAt < ELEMENT_INSTALL_FOCUS_MAX_AGE;
        if (target === null ||
            !target.canInstall ||
            target.compositionFile === null ||
            target.compositionId === null ||
            !targetIsLive ||
            !targetWasRecentlyFocused) {
            response.writeHead(409);
            response.end(JSON.stringify({
                success: false,
                reason: 'No focused writable Remotion Studio composition',
            }));
            return;
        }
        const installRequest = {
            id: `${Date.now()}-${Math.random().toString(16).slice(2)}`,
            clientId: target.clientId,
            createdAt: Date.now(),
            compositionFile: target.compositionFile,
            compositionId: target.compositionId,
            element: parsed.element,
            position: null,
        };
        const delivered = liveEventsServer.sendEventToClientId(target.clientId, {
            type: 'element-install-request',
            request: installRequest,
        });
        if (delivered === false) {
            response.writeHead(409);
            response.end(JSON.stringify({
                success: false,
                reason: 'The selected Remotion Studio tab is no longer connected',
            }));
            return;
        }
        response.writeHead(200);
        response.end(JSON.stringify({ success: true, status: 'sent' }));
    }
    catch (err) {
        response.writeHead(500);
        response.end(JSON.stringify({ success: false, reason: err.message }));
    }
};
const handleFallback = async ({ remotionRoot, hash, response, request, getCurrentInputProps, getEnvVariables, publicDir, getRenderQueue, getRenderDefaults, numberOfAudioTags, audioLatencyHint, previewSampleRate, gitSource, logLevel, enableCrossSiteIsolation, getStudioRuntimeConfig, }) => {
    var _a;
    const acceptsHtml = ((_a = request.headers.accept) !== null && _a !== void 0 ? _a : '').includes('text/html');
    if (request.method === 'GET' && acceptsHtml) {
        await (0, watch_ignore_next_change_1.reloadPreviouslySuppressedFiles)();
    }
    const requestUrl = new URL(request.url, 'http://localhost');
    const { pathname } = requestUrl;
    const staticFileHint = (0, get_static_file_fallback_hint_1.getStaticFileFallbackHint)({
        method: request.method,
        pathname,
        publicDir,
    });
    if (staticFileHint &&
        pathname.includes('.') &&
        !loggedStaticFileHints.has(staticFileHint)) {
        loggedStaticFileHints.add(staticFileHint);
        renderer_1.RenderInternals.Log.error({ indent: false, logLevel }, [
            `"${pathname}" was requested but not found.`,
            'To import assets from the public/ folder, you must wrap them in staticFile(): https://www.remotion.dev/docs/assets',
            `Change \`"${pathname}"\` to \`staticFile("${pathname}")\` to fix the error.`,
        ].join('\n'));
    }
    const displayName = await (0, open_in_editor_1.getEditorName)();
    response.setHeader('content-type', 'text/html');
    if (enableCrossSiteIsolation) {
        response.setHeader('Cross-Origin-Opener-Policy', 'same-origin');
        response.setHeader('Cross-Origin-Embedder-Policy', 'credentialless');
    }
    const packageManager = (0, get_package_manager_1.getPackageManager)({
        remotionRoot,
        packageManager: undefined,
        dirUp: 0,
        logLevel,
    });
    (0, public_folder_1.fetchFolder)({ publicDir, staticHash: hash });
    const installedDependencies = (0, get_installed_installable_packages_1.getInstalledInstallablePackages)(remotionRoot);
    response.end(bundler_1.BundlerInternals.indexHtml({
        staticHash: hash,
        publicPath: '/',
        editorName: displayName,
        envVariables: getEnvVariables(),
        inputProps: getCurrentInputProps(),
        remotionRoot,
        studioServerCommand: packageManager === 'unknown' ? null : packageManager.startCommand,
        renderQueue: getRenderQueue(),
        completedClientRenders: (0, client_render_queue_1.getCompletedClientRenders)(),
        numberOfAudioTags,
        publicFiles: (0, public_folder_1.getFiles)(),
        includeFavicon: true,
        title: 'Remotion Studio',
        renderDefaults: getRenderDefaults(),
        publicFolderExists: (0, node_fs_1.existsSync)(publicDir) ? publicDir : null,
        gitSource,
        projectName: (0, studio_shared_1.getProjectName)({
            basename: node_path_1.default.basename,
            gitSource,
            resolvedRemotionRoot: remotionRoot,
        }),
        installedDependencies,
        packageManager: packageManager === 'unknown' ? 'unknown' : packageManager.manager,
        logLevel,
        mode: 'dev',
        audioLatencyHint: audioLatencyHint !== null && audioLatencyHint !== void 0 ? audioLatencyHint : 'playback',
        sampleRate: previewSampleRate,
        studioRuntimeConfig: getStudioRuntimeConfig(),
    }));
};
const handleFileSource = async ({ method, remotionRoot, search, response, request, }) => {
    if (method === 'OPTIONS') {
        response.writeHead(200);
        response.end();
        return Promise.resolve();
    }
    (0, validate_same_origin_1.validateSameOrigin)(request);
    if (!search.startsWith('?')) {
        throw new Error('query must start with ?');
    }
    const query = new node_url_1.URLSearchParams(search);
    const f = query.get('f');
    if (typeof f !== 'string') {
        throw new Error('must pass `f` parameter');
    }
    const data = await (0, get_file_source_1.getFileSource)(remotionRoot, decodeURIComponent(f));
    response.writeHead(200);
    response.write(data);
    response.end();
    return Promise.resolve();
};
const handleAddAsset = ({ req, res, search, publicDir, }) => {
    try {
        (0, validate_same_origin_1.validateSameOrigin)(req);
        const query = new node_url_1.URLSearchParams(search);
        const filePath = query.get('filePath');
        if (typeof filePath !== 'string') {
            throw new Error('No `filePath` provided');
        }
        const absolutePath = node_path_1.default.join(publicDir, filePath);
        const relativeToPublicDir = node_path_1.default.relative(publicDir, absolutePath);
        if (relativeToPublicDir.startsWith('..')) {
            throw new Error(`Not allowed to write to ${relativeToPublicDir}`);
        }
        fs_1.default.mkdirSync(node_path_1.default.dirname(absolutePath), { recursive: true });
        const writeStream = (0, fs_1.createWriteStream)(absolutePath);
        writeStream.on('close', () => {
            res.end(JSON.stringify({ success: true }));
        });
        req.pipe(writeStream);
    }
    catch (err) {
        res.statusCode = 500;
        res.end(JSON.stringify({ error: err.message }));
    }
    return Promise.resolve();
};
const handleUploadOutput = ({ req, res, search, remotionRoot, }) => {
    try {
        (0, validate_same_origin_1.validateSameOrigin)(req);
        const query = new node_url_1.URLSearchParams(search);
        const filePath = query.get('filePath');
        if (typeof filePath !== 'string') {
            throw new Error('No `filePath` provided');
        }
        const absolutePath = (0, resolve_output_path_1.resolveOutputPath)(remotionRoot, filePath);
        fs_1.default.mkdirSync(node_path_1.default.dirname(absolutePath), { recursive: true });
        const writeStream = (0, fs_1.createWriteStream)(absolutePath);
        writeStream.on('close', () => {
            res.end(JSON.stringify({ success: true }));
        });
        writeStream.on('error', (err) => {
            res.statusCode = 500;
            res.end(JSON.stringify({ error: err.message }));
        });
        req.on('error', (err) => {
            writeStream.destroy();
            res.statusCode = 500;
            res.end(JSON.stringify({ error: err.message }));
        });
        req.pipe(writeStream);
    }
    catch (err) {
        res.statusCode = 500;
        res.end(JSON.stringify({ error: err.message }));
    }
    return Promise.resolve();
};
const handleFavicon = (_, response) => {
    const filePath = node_path_1.default.join(__dirname, '..', 'web', 'favicon.png');
    const stat = (0, node_fs_1.statSync)(filePath);
    response.writeHead(200, {
        'Content-Type': 'image/png',
        'Content-Length': stat.size,
    });
    const readStream = (0, node_fs_1.createReadStream)(filePath);
    readStream.pipe(response);
    return Promise.resolve();
};
const handleBeep = (_, response) => {
    const filePath = node_path_1.default.join(__dirname, '..', 'web', 'beep.wav');
    const stat = (0, node_fs_1.statSync)(filePath);
    response.writeHead(200, {
        'Content-Type': 'audio/wav',
        'Content-Length': stat.size,
    });
    const readStream = (0, node_fs_1.createReadStream)(filePath);
    readStream.pipe(response);
    return Promise.resolve();
};
const handleRoutes = ({ staticHash, staticHashPrefix, outputHash, outputHashPrefix, request, response, liveEventsServer, getCurrentInputProps, getEnvVariables, remotionRoot, entryPoint, publicDir, logLevel, getRenderQueue, getRenderDefaults, numberOfAudioTags, queueMethods: methods, gitSource, binariesDirectory, audioLatencyHint, previewSampleRate, enableCrossSiteIsolation, getStudioRuntimeConfig, }) => {
    const url = new URL(request.url, 'http://localhost');
    if (url.pathname === '/api/file-source') {
        return handleFileSource({
            remotionRoot,
            search: url.search,
            method: request.method,
            response,
            request,
        });
    }
    if (url.pathname === `${staticHash}/api/add-asset`) {
        return handleAddAsset({
            req: request,
            res: response,
            search: url.search,
            publicDir,
        });
    }
    if (url.pathname === '/api/upload-output') {
        return handleUploadOutput({
            req: request,
            res: response,
            search: url.search,
            remotionRoot,
        });
    }
    if (url.pathname === '/api/element-install-target' ||
        url.pathname === '/api/request-element-install') {
        if (request.method === 'OPTIONS') {
            return handleElementInstallOptions({ request, response });
        }
        if (url.pathname === '/api/element-install-target') {
            return handleElementInstallTarget({
                liveEventsServer,
                request,
                response,
                remotionRoot,
                gitSource,
            });
        }
        return handleRequestElementInstall({
            liveEventsServer,
            request,
            response,
        });
    }
    for (const [key, value] of Object.entries(api_routes_1.allApiRoutes)) {
        if (url.pathname === key) {
            return (0, handler_1.handleRequest)({
                remotionRoot,
                entryPoint,
                handler: value,
                request,
                response,
                logLevel,
                methods,
                binariesDirectory,
                publicDir,
            });
        }
    }
    if (url.pathname === '/favicon.ico') {
        return handleFavicon(request, response);
    }
    if (url.pathname === '/beep.wav') {
        return handleBeep(request, response);
    }
    if (url.pathname === '/__remotion_config') {
        return handleRemotionConfig(response, remotionRoot);
    }
    if (url.pathname === '/events') {
        return liveEventsServer.router(request, response);
    }
    if (url.pathname.startsWith(staticHash)) {
        const filename = new URL(request.url, 'http://localhost').pathname.replace(new RegExp(`^${staticHash}`), '');
        const filePath = (0, node_path_1.join)(publicDir, decodeURIComponent(filename));
        return (0, serve_static_1.serveStatic)({
            root: publicDir,
            path: filePath,
            req: request,
            res: response,
            allowOutsidePublicFolder: false,
        });
    }
    if (url.pathname.startsWith(staticHashPrefix)) {
        return static404(response);
    }
    if (url.pathname.startsWith(outputHash)) {
        const filename = new URL(request.url, 'http://localhost').pathname.replace(new RegExp(`^${outputHash}`), '');
        const filePath = (0, node_path_1.join)(remotionRoot, decodeURIComponent(filename));
        return (0, serve_static_1.serveStatic)({
            root: remotionRoot,
            path: filePath,
            req: request,
            res: response,
            allowOutsidePublicFolder: false,
        });
    }
    if (url.pathname.startsWith(outputHashPrefix)) {
        return output404(response);
    }
    return handleFallback({
        remotionRoot,
        hash: staticHash,
        response,
        request,
        getCurrentInputProps,
        getEnvVariables,
        publicDir,
        getRenderQueue,
        getRenderDefaults,
        numberOfAudioTags,
        gitSource,
        logLevel,
        audioLatencyHint,
        previewSampleRate,
        enableCrossSiteIsolation,
        getStudioRuntimeConfig,
    });
};
exports.handleRoutes = handleRoutes;
