// @remotion/studio-server/dist/preview-server/live-events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setLiveEventsListener = exports.waitForLiveEventsListener = exports.makeLiveEventsRouter = exports.clearPrintPortMessageTimeout = void 0;
const server_ready_1 = require("../server-ready");
const default_props_watchers_1 = require("./default-props-watchers");
const file_existence_watchers_1 = require("./file-existence-watchers");
const sequence_props_watchers_1 = require("./sequence-props-watchers");
const serializeMessage = (message) => {
    return `data: ${JSON.stringify(message)}\n\n`;
};
let printPortMessageTimeout = null;
const clearPrintPortMessageTimeout = () => {
    if (printPortMessageTimeout) {
        clearTimeout(printPortMessageTimeout);
        printPortMessageTimeout = null;
    }
};
exports.clearPrintPortMessageTimeout = clearPrintPortMessageTimeout;
const makeLiveEventsRouter = (logLevel, getInitialUndoRedoState) => {
    let clients = [];
    let newClientListeners = [];
    const router = (request, response) => {
        const headers = {
            'content-type': 'text/event-stream;charset=utf-8',
            connection: 'keep-alive',
            'cache-control': 'no-cache',
        };
        response.writeHead(200, headers);
        response.write('\n');
        if (request.method === 'OPTIONS') {
            response.end();
            return Promise.resolve();
        }
        const clientId = String(Math.random());
        const { undoFile, redoFile } = getInitialUndoRedoState();
        response.write(serializeMessage({ type: 'init', clientId, undoFile, redoFile }));
        const newClient = {
            id: clientId,
            response,
        };
        clients.push(newClient);
        newClientListeners.forEach((cb) => cb());
        (0, exports.clearPrintPortMessageTimeout)();
        request.on('close', () => {
            (0, default_props_watchers_1.unsubscribeClientDefaultPropsWatchers)(clientId);
            (0, file_existence_watchers_1.unsubscribeClientFileExistenceWatchers)(clientId);
            (0, sequence_props_watchers_1.unsubscribeClientSequencePropsWatchers)(clientId);
            clients = clients.filter((client) => client.id !== clientId);
            // If all clients disconnected, print a comment so user can easily restart it.
            if (clients.length === 0) {
                (0, exports.clearPrintPortMessageTimeout)();
                printPortMessageTimeout = setTimeout(() => {
                    (0, server_ready_1.printServerReadyComment)('To restart', logLevel);
                }, 2500);
            }
        });
        return Promise.resolve();
    };
    const sendEventToClient = (event) => {
        clients.forEach((client) => {
            client.response.write(serializeMessage(event));
        });
    };
    const sendEventToClientId = (clientId, event) => {
        const client = clients.find((c) => c.id === clientId);
        if (!client) {
            return false;
        }
        client.response.write(serializeMessage(event));
        return true;
    };
    const addNewClientListener = (cb) => {
        newClientListeners.push(cb);
        return () => {
            newClientListeners = newClientListeners.filter((l) => l !== cb);
        };
    };
    return {
        sendEventToClient,
        sendEventToClientId,
        router,
        addNewClientListener,
        closeConnections: () => {
            return Promise.all(clients.map((client) => {
                return new Promise((resolve) => {
                    client.response.end(() => {
                        resolve();
                    });
                });
            })).then(() => undefined);
        },
    };
};
exports.makeLiveEventsRouter = makeLiveEventsRouter;
let liveEventsListener = null;
const waiters = [];
const waitForLiveEventsListener = () => {
    if (liveEventsListener) {
        return Promise.resolve(liveEventsListener);
    }
    return new Promise((resolve) => {
        waiters.push((list) => {
            resolve(list);
        });
    });
};
exports.waitForLiveEventsListener = waitForLiveEventsListener;
const setLiveEventsListener = (listener) => {
    liveEventsListener = listener;
    waiters.forEach((w) => w(listener));
    return () => {
        liveEventsListener = null;
    };
};
exports.setLiveEventsListener = setLiveEventsListener;
