// @remotion/studio/dist/helpers/studio-runtime-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStudioBufferStateDelayInMilliseconds = exports.getStudioMaxTimelineTracks = exports.getStudioExperimentalClientSideRenderingEnabled = exports.getStudioKeyboardShortcutsEnabled = exports.getStudioInteractivityEnabled = exports.getStudioAskAIEnabled = exports.DEFAULT_BUFFER_STATE_DELAY_IN_MILLISECONDS = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
exports.DEFAULT_BUFFER_STATE_DELAY_IN_MILLISECONDS = 300;
const defaultStudioRuntimeConfig = {
    askAIEnabled: false,
    bufferStateDelayInMilliseconds: null,
    experimentalClientSideRenderingEnabled: false,
    interactivityEnabled: true,
    keyboardShortcutsEnabled: true,
    maxTimelineTracks: null,
};
const getStudioRuntimeConfig = () => {
    var _a;
    if (typeof window === 'undefined') {
        return defaultStudioRuntimeConfig;
    }
    return (_a = window.remotion_studioConfig) !== null && _a !== void 0 ? _a : defaultStudioRuntimeConfig;
};
const getStudioAskAIEnabled = () => {
    return getStudioRuntimeConfig().askAIEnabled;
};
exports.getStudioAskAIEnabled = getStudioAskAIEnabled;
const getStudioInteractivityEnabled = () => {
    return getStudioRuntimeConfig().interactivityEnabled;
};
exports.getStudioInteractivityEnabled = getStudioInteractivityEnabled;
const getStudioKeyboardShortcutsEnabled = () => {
    return getStudioRuntimeConfig().keyboardShortcutsEnabled;
};
exports.getStudioKeyboardShortcutsEnabled = getStudioKeyboardShortcutsEnabled;
const getStudioExperimentalClientSideRenderingEnabled = () => {
    return getStudioRuntimeConfig().experimentalClientSideRenderingEnabled;
};
exports.getStudioExperimentalClientSideRenderingEnabled = getStudioExperimentalClientSideRenderingEnabled;
const getStudioMaxTimelineTracks = () => {
    var _a;
    return (_a = getStudioRuntimeConfig().maxTimelineTracks) !== null && _a !== void 0 ? _a : studio_shared_1.DEFAULT_TIMELINE_TRACKS;
};
exports.getStudioMaxTimelineTracks = getStudioMaxTimelineTracks;
const getStudioBufferStateDelayInMilliseconds = () => {
    var _a;
    return ((_a = getStudioRuntimeConfig().bufferStateDelayInMilliseconds) !== null && _a !== void 0 ? _a : exports.DEFAULT_BUFFER_STATE_DELAY_IN_MILLISECONDS);
};
exports.getStudioBufferStateDelayInMilliseconds = getStudioBufferStateDelayInMilliseconds;
