// @remotion/studio-shared/dist/drag-mime-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRemotionDragMimeType = exports.REMOTION_DRAG_MIME_TYPES = exports.SFX_DRAG_MIME_TYPE = exports.ELEMENT_DRAG_MIME_TYPE = exports.EFFECT_DRAG_MIME_TYPE = exports.COMPOSITION_DRAG_MIME_TYPE = exports.COMPONENT_DRAG_MIME_TYPE = exports.ASSET_DRAG_MIME_TYPE = void 0;
exports.ASSET_DRAG_MIME_TYPE = 'application/vnd.remotion.asset+json';
exports.COMPONENT_DRAG_MIME_TYPE = 'application/vnd.remotion.component+json';
exports.COMPOSITION_DRAG_MIME_TYPE = 'application/vnd.remotion.composition+json';
exports.EFFECT_DRAG_MIME_TYPE = 'application/vnd.remotion.effect+json';
exports.ELEMENT_DRAG_MIME_TYPE = 'application/vnd.remotion.element+json';
exports.SFX_DRAG_MIME_TYPE = 'application/vnd.remotion.sfx+json';
exports.REMOTION_DRAG_MIME_TYPES = [
    exports.ASSET_DRAG_MIME_TYPE,
    exports.COMPONENT_DRAG_MIME_TYPE,
    exports.COMPOSITION_DRAG_MIME_TYPE,
    exports.EFFECT_DRAG_MIME_TYPE,
    exports.ELEMENT_DRAG_MIME_TYPE,
    exports.SFX_DRAG_MIME_TYPE,
];
const isRemotionDragMimeType = (type) => {
    return exports.REMOTION_DRAG_MIME_TYPES.some((mimeType) => mimeType === type);
};
exports.isRemotionDragMimeType = isRemotionDragMimeType;
