// @remotion/renderer/dist/options/interactivity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interactivityOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
let interactivityEnabled = true;
const cliFlag = 'disable-interactivity';
exports.interactivityOption = {
    name: 'Disable or enable Studio interactivity',
    cliFlag,
    description: () => (jsx_runtime_1.jsx(jsx_runtime_1.Fragment, { children: "Enable or disable interactive editing in the Remotion Studio. When disabled, the Studio keeps previewing and source navigation available, but disables preview outlines, the sequence inspector, visual controls, timeline selection and timeline editing gestures." })),
    ssrName: null,
    docLink: 'https://www.remotion.dev/docs/config#setinteractivityenabled',
    type: false,
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined && commandLine[cliFlag] !== null) {
            interactivityEnabled = commandLine[cliFlag] === false;
            return {
                value: interactivityEnabled,
                source: 'cli',
            };
        }
        return {
            value: interactivityEnabled,
            source: 'config',
        };
    },
    setConfig(value) {
        interactivityEnabled = value;
    },
    id: cliFlag,
};
