// @remotion/studio/dist/components/RenderModal/SchemaEditor/Fieldset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fieldset = exports.getSchemaEditorRootInset = exports.getSchemaEditorFieldsetPadding = exports.SCHEMA_EDITOR_FIELDSET_PADDING = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
exports.SCHEMA_EDITOR_FIELDSET_PADDING = 6;
const getSchemaEditorFieldsetPadding = () => {
    return exports.SCHEMA_EDITOR_FIELDSET_PADDING;
};
exports.getSchemaEditorFieldsetPadding = getSchemaEditorFieldsetPadding;
const getSchemaEditorRootInset = (contentInset) => {
    const fieldsetPadding = (0, exports.getSchemaEditorFieldsetPadding)();
    return Math.max(0, contentInset - fieldsetPadding);
};
exports.getSchemaEditorRootInset = getSchemaEditorRootInset;
const AlreadyPaddedRightContext = (0, react_1.createContext)(false);
const Fieldset = ({ children, shouldPad }) => {
    const alreadyPadded = (0, react_1.useContext)(AlreadyPaddedRightContext);
    const style = (0, react_1.useMemo)(() => {
        if (shouldPad) {
            const padding = (0, exports.getSchemaEditorFieldsetPadding)();
            return {
                padding,
                paddingTop: padding / 2,
                paddingRight: alreadyPadded ? 0 : padding,
            };
        }
        return {};
    }, [alreadyPadded, shouldPad]);
    const content = jsx_runtime_1.jsx("div", { style: style, children: children });
    if (shouldPad) {
        return (jsx_runtime_1.jsx(AlreadyPaddedRightContext.Provider, { value: true, children: content }));
    }
    return content;
};
exports.Fieldset = Fieldset;
