// @remotion/media/dist/esm/index.mjs
var __dispose = Symbol.dispose || /* @__PURE__ */ Symbol.for("Symbol.dispose");
var __asyncDispose = Symbol.asyncDispose || /* @__PURE__ */ Symbol.for("Symbol.asyncDispose");
var __using = (stack, value, async) => {
  if (value != null) {
    if (typeof value !== "object" && typeof value !== "function")
      throw TypeError('Object expected to be assigned to "using" declaration');
    var dispose;
    if (async)
      dispose = value[__asyncDispose];
    if (dispose === undefined)
      dispose = value[__dispose];
    if (typeof dispose !== "function")
      throw TypeError("Object not disposable");
    stack.push([async, dispose, value]);
  } else if (async) {
    stack.push([async]);
  }
  return value;
};
var __callDispose = (stack, error, hasError) => {
  var E = typeof SuppressedError === "function" ? SuppressedError : function(e, s, m, _) {
    return _ = Error(m), _.name = "SuppressedError", _.error = e, _.suppressed = s, _;
  }, fail = (e) => error = hasError ? new E(e, error, "An error was suppressed during disposal") : (hasError = true, e), next = (it) => {
    while (it = stack.pop()) {
      try {
        var result = it[1] && it[1].call(it[2]);
        if (it[0])
          return Promise.resolve(result).then(next, (e) => (fail(e), next()));
      } catch (e) {
        fail(e);
      }
    }
    if (hasError)
      throw error;
  };
  return next();
};

// src/audio/audio.tsx
import { useMemo as useMemo3, useState as useState3 } from "react";
import {
  Internals as Internals17,
  Interactive,
  Sequence,
  useRemotionEnvironment as useRemotionEnvironment2,
  useVideoConfig as useVideoConfig2
} from "remotion";

// src/show-in-timeline.ts
import { Internals } from "remotion";
var getLoopDisplay = ({
  loop,
  mediaDurationInSeconds,
  playbackRate,
  trimAfter,
  trimBefore,
  sequenceDurationInFrames,
  compFps
}) => {
  if (!loop || !mediaDurationInSeconds) {
    return;
  }
  const durationInFrames = Internals.calculateMediaDuration({
    mediaDurationInFrames: mediaDurationInSeconds * compFps,
    playbackRate,
    trimAfter,
    trimBefore
  });
  const maxTimes = sequenceDurationInFrames / durationInFrames;
  return {
    numberOfTimes: maxTimes,
    startOffset: 0,
    durationInFrames
  };
};

// src/audio/audio-for-preview.tsx
import { useContext as useContext2, useEffect, useMemo, useRef, useState } from "react";
import {
  Internals as Internals7,
  Audio as RemotionAudio,
  useBufferState,
  useCurrentFrame,
  useVideoConfig
} from "remotion";

// src/get-time-in-seconds.ts
import { Internals as Internals2 } from "remotion";
var getTimeInSeconds = ({
  loop,
  mediaDurationInSeconds,
  unloopedTimeInSeconds,
  src,
  trimAfter,
  trimBefore,
  fps,
  playbackRate,
  ifNoMediaDuration
}) => {
  if (mediaDurationInSeconds === null && loop && ifNoMediaDuration === "fail") {
    throw new Error(`Could not determine duration of ${src}, but "loop" was set.`);
  }
  const loopDurationInFrames = loop ? Internals2.calculateMediaDuration({
    trimAfter,
    mediaDurationInFrames: mediaDurationInSeconds ? mediaDurationInSeconds * fps : Infinity,
    playbackRate: 1,
    trimBefore
  }) : Infinity;
  const timeInSeconds = loop ? unloopedTimeInSeconds * playbackRate * fps % loopDurationInFrames / fps : unloopedTimeInSeconds * playbackRate;
  if ((trimAfter ?? null) !== null && !loop) {
    const time = (trimAfter - (trimBefore ?? 0)) / fps;
    if (timeInSeconds >= time) {
      return null;
    }
  }
  return timeInSeconds + (trimBefore ?? 0) / fps;
};
var calculateEndTime = ({
  mediaDurationInSeconds,
  ifNoMediaDuration,
  src,
  trimAfter,
  trimBefore,
  fps
}) => {
  if (mediaDurationInSeconds === null && ifNoMediaDuration === "fail") {
    throw new Error(`Could not determine duration of ${src}, but "loop" was set.`);
  }
  const mediaDuration = Internals2.calculateMediaDuration({
    trimAfter,
    mediaDurationInFrames: mediaDurationInSeconds ? mediaDurationInSeconds * fps : Infinity,
    playbackRate: 1,
    trimBefore
  }) / fps;
  return mediaDuration + (trimBefore ?? 0) / fps;
};

// src/media-player.ts
import { ALL_FORMATS, Input, UrlSource } from "mediabunny";
import { Internals as Internals5 } from "remotion";

// src/audio-iterator-manager.ts
import {
  AudioBufferSink,
  InputDisposedError
} from "mediabunny";
import { Internals as Internals3 } from "remotion";

// src/audio/audio-preview-iterator.ts
var ALLOWED_GLOBAL_TIME_ANCHOR_SHIFT = 0.1;
var makeAudioIterator = ({
  startFromSecond,
  iterator,
  unscheduleAudioNode
}) => {
  let destroyed = false;
  const queuedAudioNodes = [];
  let mostRecentTimestamp = -Infinity;
  const cleanupAudioQueue = () => {
    for (const { node } of queuedAudioNodes) {
      unscheduleAudioNode(node);
      try {
        node.stop();
      } catch {}
    }
    queuedAudioNodes.length = 0;
  };
  const getNextFn = async () => {
    const next = await iterator.next();
    if (next.value) {
      mostRecentTimestamp = Math.max(mostRecentTimestamp, next.value.timelineTimestamp + next.value.sourceDurationInSeconds);
    }
    return next;
  };
  return {
    destroy: () => {
      cleanupAudioQueue();
      destroyed = true;
      iterator.return().catch(() => {
        return;
      });
    },
    getNextFn,
    isDestroyed: () => {
      return destroyed;
    },
    addQueuedAudioNode: (queuedNode) => {
      queuedAudioNodes.push(queuedNode);
    },
    guessNextTimestamp: () => {
      return !Number.isFinite(mostRecentTimestamp) ? startFromSecond : mostRecentTimestamp;
    },
    getQueuedPeriod: () => {
      let until = -Infinity;
      let from = Infinity;
      for (const node of queuedAudioNodes) {
        until = Math.max(until, node.timestamp + node.sourceDurationInSeconds);
        from = Math.min(from, node.timestamp);
      }
      if (!Number.isFinite(from) || !Number.isFinite(until)) {
        return null;
      }
      return {
        from,
        until
      };
    }
  };
};
var isAlreadyQueued = (time, queuedPeriod) => {
  if (!queuedPeriod) {
    return false;
  }
  return time >= queuedPeriod.from && time < queuedPeriod.until;
};

// src/audio/get-scheduled-time.ts
var getScheduledTime = ({
  mediaTimestamp,
  targetTime,
  currentTime,
  sequenceStartTime
}) => {
  const needsTrimStart = mediaTimestamp < sequenceStartTime;
  const offsetBecauseOfTrim = needsTrimStart ? sequenceStartTime - mediaTimestamp : 0;
  const offsetBecauseOfTooLate = targetTime < 0 ? -targetTime : 0;
  const offset = offsetBecauseOfTrim + offsetBecauseOfTooLate;
  const scheduledTime = targetTime + currentTime + offset;
  return scheduledTime;
};
var getDurationOfNode = ({
  sourceDurationInSeconds,
  sourceOffsetInSeconds,
  offset
}) => {
  return Math.max(0, sourceDurationInSeconds - (offset - sourceOffsetInSeconds));
};
var getTrimStartForAudioNode = ({
  mediaTimestamp,
  targetTime,
  sequenceStartTime,
  combinedPlaybackRate,
  sourceStartOffsetInSeconds
}) => {
  const needsTrimStart = mediaTimestamp < sequenceStartTime;
  const offsetBecauseOfTrim = needsTrimStart ? sequenceStartTime - mediaTimestamp : 0;
  const offsetBecauseOfTooLate = targetTime < 0 ? -targetTime * combinedPlaybackRate : 0;
  return sourceStartOffsetInSeconds + offsetBecauseOfTrim + offsetBecauseOfTooLate;
};

// src/audio/sort-by-priority.ts
class StaleWaiterError extends Error {
  constructor() {
    super("Waiter became stale before it got its turn");
    this.name = "StaleWaiterError";
  }
}
var CONCURRENCY = 1;
var waiters = [];
var running = 0;
var runningEntry = null;
var processNext = () => {
  if (running >= CONCURRENCY) {
    if (runningEntry?.waiter.getPriority() === null) {
      runningEntry.cancel();
    } else {
      return;
    }
  }
  const staleWaiters = [];
  for (let i = waiters.length - 1;i >= 0; i--) {
    if (waiters[i].getPriority() === null) {
      const [stale] = waiters.splice(i, 1);
      staleWaiters.push(stale);
    }
  }
  for (const stale of staleWaiters) {
    stale.onError(new StaleWaiterError);
  }
  if (waiters.length === 0) {
    return;
  }
  let bestIndex = 0;
  let bestPriority = waiters[0].getPriority();
  if (bestPriority === null) {
    throw new Error("Stale waiter should have been removed");
  }
  for (let i = 1;i < waiters.length; i++) {
    const priority = waiters[i].getPriority();
    if (priority === null) {
      throw new Error("Stale waiter should have been removed");
    }
    if (priority < bestPriority) {
      bestPriority = priority;
      bestIndex = i;
    }
  }
  if (bestPriority > 2) {
    return;
  }
  const [next] = waiters.splice(bestIndex, 1);
  running++;
  let settled = false;
  let cancelled = false;
  const entry = {
    waiter: next,
    cancel: () => {
      cancelled = true;
      entry.settle();
    },
    settle: () => {
      if (settled) {
        return;
      }
      settled = true;
      running--;
      if (runningEntry === entry) {
        runningEntry = null;
      }
    }
  };
  runningEntry = entry;
  next.fn().then((value) => {
    entry.settle();
    if (cancelled) {
      return;
    }
    next.onDone(value, processNext);
  }, (err) => {
    entry.settle();
    if (cancelled) {
      return;
    }
    next.onError(err);
  });
};
var waitForTurn = ({
  getPriority,
  fn,
  onDone,
  onError
}) => {
  waiters.push({
    getPriority,
    fn,
    onDone,
    onError
  });
  processNext();
};

// src/make-iterator-with-priming.ts
var AUDIO_PRIMING_SECONDS = 0.5;
async function* makeIteratorWithPriming({
  audioSink,
  timeToSeek,
  maximumTimestamp
}) {
  const primingStart = Math.max(0, timeToSeek - AUDIO_PRIMING_SECONDS);
  const iterator = audioSink.buffers(primingStart, maximumTimestamp);
  for await (const buffer of iterator) {
    const sourceStart = Math.max(buffer.timestamp, timeToSeek);
    const sourceEnd = Math.min(buffer.timestamp + buffer.duration, maximumTimestamp);
    if (sourceEnd <= sourceStart) {
      continue;
    }
    yield {
      buffer,
      timelineTimestamp: sourceStart,
      sourceOffsetInSeconds: sourceStart - buffer.timestamp,
      sourceDurationInSeconds: sourceEnd - sourceStart
    };
  }
}
async function* makeLoopingIterator({
  audioSink,
  seekTimeInSeconds,
  loopStartInSeconds,
  segmentEndInSeconds,
  maximumContinuousTimestamp
}) {
  if (segmentEndInSeconds <= loopStartInSeconds) {
    throw new Error(`Cannot loop audio over an empty range (${loopStartInSeconds} to ${segmentEndInSeconds})`);
  }
  let passStartInSeconds = seekTimeInSeconds;
  let passBaseTimestamp = seekTimeInSeconds;
  while (true) {
    let yieldedInPass = false;
    for await (const item of makeIteratorWithPriming({
      audioSink,
      timeToSeek: passStartInSeconds,
      maximumTimestamp: segmentEndInSeconds
    })) {
      yieldedInPass = true;
      const timelineTimestamp = passBaseTimestamp + (item.timelineTimestamp - passStartInSeconds);
      const sourceDurationInSeconds = Math.min(item.sourceDurationInSeconds, maximumContinuousTimestamp - timelineTimestamp);
      if (sourceDurationInSeconds <= 0) {
        return;
      }
      yield {
        ...item,
        timelineTimestamp,
        sourceDurationInSeconds
      };
      if (timelineTimestamp + sourceDurationInSeconds >= maximumContinuousTimestamp) {
        return;
      }
    }
    if (!yieldedInPass) {
      throw new Error(`Audio loop pass from ${passStartInSeconds} to ${segmentEndInSeconds} yielded no buffers`);
    }
    passBaseTimestamp += segmentEndInSeconds - passStartInSeconds;
    passStartInSeconds = loopStartInSeconds;
  }
}

// src/audio-iterator-manager.ts
var anchorToContinuousTime = ({
  anchor,
  unloopedTimeInSeconds,
  playbackRate
}) => {
  return anchor.mediaStartInSeconds + (unloopedTimeInSeconds - anchor.unloopedStartInSeconds) * playbackRate;
};
var audioIteratorManager = ({
  audioTrack,
  delayPlaybackHandleIfNotPremounting,
  sharedAudioContext,
  getSequenceEndTimestamp,
  getSequenceDurationInSeconds,
  getMediaEndTimestamp,
  getStartTime,
  initialMuted,
  drawDebugOverlay,
  initialPlaybackRate,
  initialTrimBefore,
  initialTrimAfter,
  initialSequenceOffset,
  initialSequenceDurationInFrames,
  initialLoop,
  initialFps
}) => {
  let muted = initialMuted;
  let currentVolume = 1;
  let currentSeek = {
    time: -1,
    playbackRate: initialPlaybackRate,
    trimBefore: initialTrimBefore,
    trimAfter: initialTrimAfter,
    sequenceOffset: initialSequenceOffset,
    sequenceDurationInFrames: initialSequenceDurationInFrames,
    loop: initialLoop,
    fps: initialFps
  };
  const gainNode = sharedAudioContext.audioContext.createGain();
  gainNode.connect(sharedAudioContext.gainNode);
  const audioSink = new AudioBufferSink(audioTrack);
  let audioBufferIterator = null;
  let currentAnchor = null;
  let audioIteratorsCreated = 0;
  let totalAudioScheduledInSeconds = 0;
  let currentDelayHandle = null;
  const unblockCurrentDelayHandle = () => {
    if (currentDelayHandle) {
      currentDelayHandle.unblock();
      currentDelayHandle = null;
    }
  };
  const pendingScheduleWaiters = [];
  const notifyNodeScheduled = () => {
    for (let i = pendingScheduleWaiters.length - 1;i >= 0; i--) {
      const waiter = pendingScheduleWaiters[i];
      waiter.remaining--;
      if (waiter.remaining <= 0) {
        waiter.resolve();
        pendingScheduleWaiters.splice(i, 1);
      }
    }
  };
  const waitForNScheduledNodes = (n) => {
    if (n <= 0) {
      return Promise.resolve();
    }
    return new Promise((resolve) => {
      pendingScheduleWaiters.push({ remaining: n, resolve });
    });
  };
  const scheduleAudioChunk = ({
    buffer,
    mediaTimestamp,
    originalUnloopedMediaTimestamp,
    sourceOffsetInSeconds,
    sourceDurationInSeconds,
    playbackRate,
    scheduleAudioNode,
    logLevel
  }) => {
    if (!audioBufferIterator) {
      throw new Error("Audio buffer iterator not found");
    }
    if (muted) {
      return;
    }
    const node = sharedAudioContext.audioContext.createBufferSource();
    node.buffer = buffer;
    node.playbackRate.value = playbackRate;
    node.connect(gainNode);
    const started = scheduleAudioNode(node, mediaTimestamp, originalUnloopedMediaTimestamp, sourceOffsetInSeconds, sourceDurationInSeconds);
    if (started.type === "not-started") {
      Internals3.Log.verbose({ logLevel, tag: "audio-scheduling" }, "not started, disconnected: %s %s", mediaTimestamp.toFixed(3), buffer.duration.toFixed(3));
      node.disconnect();
      return;
    }
    audioBufferIterator.addQueuedAudioNode({
      node,
      timestamp: mediaTimestamp,
      buffer,
      sourceDurationInSeconds,
      scheduledTime: started.scheduledTime,
      playbackRate,
      scheduledAtAnchor: sharedAudioContext.audioSyncAnchor.value
    });
  };
  const onAudioChunk = ({
    buffer,
    playbackRate,
    scheduleAudioNode,
    logLevel
  }) => {
    if (muted) {
      return;
    }
    const startTime = getStartTime();
    const sequenceEndTime = getSequenceEndTimestamp();
    if (buffer.timelineTimestamp + buffer.sourceDurationInSeconds <= startTime) {
      return;
    }
    if (buffer.timelineTimestamp >= sequenceEndTime) {
      return;
    }
    const scheduledStart = Math.max(buffer.timelineTimestamp, startTime);
    const scheduledEnd = Math.min(buffer.timelineTimestamp + buffer.sourceDurationInSeconds, sequenceEndTime);
    totalAudioScheduledInSeconds += Math.max(0, scheduledEnd - scheduledStart);
    scheduleAudioChunk({
      buffer: buffer.buffer.buffer,
      mediaTimestamp: buffer.timelineTimestamp,
      playbackRate,
      scheduleAudioNode,
      logLevel,
      originalUnloopedMediaTimestamp: buffer.buffer.timestamp,
      sourceOffsetInSeconds: buffer.sourceOffsetInSeconds,
      sourceDurationInSeconds: buffer.sourceDurationInSeconds
    });
    drawDebugOverlay();
  };
  const proceedScheduling = ({
    iterator,
    nonce,
    getTargetTime,
    playbackRate,
    scheduleAudioNode,
    onScheduled,
    onDestroyed,
    onDone,
    logLevel,
    currentTime,
    getAudioContextCurrentTimeMockedInTest
  }) => {
    waitForTurn({
      getPriority: () => {
        if (iterator.isDestroyed()) {
          onDestroyed();
          return null;
        }
        const guessedNextTimestamp = iterator.guessNextTimestamp();
        const targetTime = getTargetTime(guessedNextTimestamp, currentTime);
        if (targetTime === null) {
          return null;
        }
        const scheduledTime = getScheduledTime({
          mediaTimestamp: guessedNextTimestamp,
          targetTime,
          currentTime,
          sequenceStartTime: getStartTime()
        });
        return scheduledTime - getAudioContextCurrentTimeMockedInTest();
      },
      fn: () => iterator.getNextFn(),
      onDone: (result, next) => {
        if (iterator.isDestroyed()) {
          next();
          onDestroyed();
          return;
        }
        if (!result.value) {
          next();
          onDone();
          return;
        }
        onScheduled(result.value.timelineTimestamp);
        notifyNodeScheduled();
        onAudioChunk({
          buffer: result.value,
          playbackRate,
          scheduleAudioNode,
          logLevel
        });
        proceedScheduling({
          iterator,
          nonce,
          getTargetTime,
          playbackRate,
          scheduleAudioNode,
          onScheduled,
          onDestroyed,
          onDone,
          logLevel,
          currentTime,
          getAudioContextCurrentTimeMockedInTest
        });
        next();
      },
      onError: (e) => {
        if (e instanceof InputDisposedError) {
          onDestroyed();
          return;
        }
        if (e instanceof StaleWaiterError) {
          onDestroyed();
          return;
        }
        throw e;
      }
    });
  };
  const startAudioIterator = ({
    nonce,
    playbackRate,
    startFromSecond,
    unloopedStartFromSecond,
    scheduleAudioNode,
    getTargetTime,
    logLevel,
    loop,
    unscheduleAudioNode,
    getAudioContextCurrentTimeMockedInTest
  }) => {
    if (muted) {
      return;
    }
    const maximumTimestamp = getMediaEndTimestamp();
    if (startFromSecond >= maximumTimestamp) {
      return;
    }
    audioBufferIterator?.destroy();
    unblockCurrentDelayHandle();
    const delayHandle = delayPlaybackHandleIfNotPremounting();
    currentDelayHandle = delayHandle;
    currentAnchor = {
      unloopedStartInSeconds: unloopedStartFromSecond,
      mediaStartInSeconds: startFromSecond
    };
    const maximumContinuousTimestamp = startFromSecond + getSequenceDurationInSeconds() * playbackRate;
    const source = loop ? makeLoopingIterator({
      audioSink,
      seekTimeInSeconds: startFromSecond,
      loopStartInSeconds: getStartTime(),
      segmentEndInSeconds: maximumTimestamp,
      maximumContinuousTimestamp
    }) : makeIteratorWithPriming({
      audioSink,
      timeToSeek: startFromSecond,
      maximumTimestamp
    });
    const iterator = makeAudioIterator({
      startFromSecond,
      iterator: source,
      unscheduleAudioNode
    });
    audioIteratorsCreated++;
    audioBufferIterator = iterator;
    let chunksScheduled = 0;
    proceedScheduling({
      iterator,
      nonce,
      getTargetTime,
      playbackRate,
      scheduleAudioNode,
      onScheduled: () => {
        chunksScheduled++;
        if (chunksScheduled === 6) {
          delayHandle.unblock();
        }
      },
      onDestroyed: () => {
        delayHandle.unblock();
      },
      onDone: () => {
        delayHandle.unblock();
      },
      logLevel,
      currentTime: sharedAudioContext.audioContext.currentTime,
      getAudioContextCurrentTimeMockedInTest
    });
  };
  const seek = ({
    newTime,
    unloopedNewTime,
    nonce,
    playbackRate,
    localPlaybackRate,
    scheduleAudioNode,
    getTargetTime,
    logLevel,
    loop,
    trimBefore,
    trimAfter,
    sequenceOffset,
    sequenceDurationInFrames,
    fps,
    getAudioContextCurrentTimeMockedInTest
  }) => {
    if (nonce.isStale()) {
      return;
    }
    if (currentSeek.time === newTime && currentSeek.playbackRate === playbackRate && currentSeek.trimBefore === trimBefore && currentSeek.trimAfter === trimAfter && currentSeek.sequenceOffset === sequenceOffset && currentSeek.sequenceDurationInFrames === sequenceDurationInFrames && currentSeek.loop === loop && currentSeek.fps === fps) {
      return;
    }
    currentSeek = {
      time: newTime,
      playbackRate,
      trimBefore,
      trimAfter,
      sequenceOffset,
      sequenceDurationInFrames,
      loop,
      fps
    };
    if (muted) {
      return;
    }
    if (audioBufferIterator && !audioBufferIterator.isDestroyed()) {
      const timeToCheck = loop && currentAnchor ? anchorToContinuousTime({
        anchor: currentAnchor,
        unloopedTimeInSeconds: unloopedNewTime,
        playbackRate: localPlaybackRate
      }) : newTime;
      const queuedPeriod = audioBufferIterator.getQueuedPeriod();
      const queuedPeriodMinusLatency = queuedPeriod ? {
        from: queuedPeriod.from - ALLOWED_GLOBAL_TIME_ANCHOR_SHIFT - sharedAudioContext.audioContext.baseLatency - sharedAudioContext.audioContext.outputLatency,
        until: queuedPeriod.until
      } : null;
      const currentTimeIsAlreadyQueued = isAlreadyQueued(timeToCheck, queuedPeriodMinusLatency);
      if (currentTimeIsAlreadyQueued) {
        processNext();
        return;
      }
      const currentIteratorTimestamp = audioBufferIterator.guessNextTimestamp();
      if (currentIteratorTimestamp < timeToCheck && Math.abs(currentIteratorTimestamp - timeToCheck) < 1) {
        processNext();
        return;
      }
    }
    startAudioIterator({
      nonce,
      playbackRate,
      startFromSecond: newTime,
      unloopedStartFromSecond: unloopedNewTime,
      scheduleAudioNode,
      getTargetTime,
      logLevel,
      loop,
      unscheduleAudioNode: sharedAudioContext.unscheduleAudioNode,
      getAudioContextCurrentTimeMockedInTest
    });
  };
  return {
    startAudioIterator,
    getAudioBufferIterator: () => audioBufferIterator,
    getCurrentAnchor: () => currentAnchor,
    destroyIterator: () => {
      audioBufferIterator?.destroy();
      audioBufferIterator = null;
      currentAnchor = null;
      unblockCurrentDelayHandle();
    },
    seek,
    getAudioIteratorsCreated: () => audioIteratorsCreated,
    getTotalAudioScheduledInSeconds: () => totalAudioScheduledInSeconds,
    setMuted: (newMuted) => {
      muted = newMuted;
      gainNode.gain.value = muted ? 0 : currentVolume;
    },
    setVolume: (volume) => {
      currentVolume = Math.max(0, volume);
      gainNode.gain.value = muted ? 0 : currentVolume;
    },
    scheduleAudioChunk,
    waitForNScheduledNodes
  };
};

// src/debug-overlay/preview-overlay.ts
var drawPreviewOverlay = ({
  context,
  audioTime,
  audioContextState,
  audioSyncAnchor,
  playing,
  audioIteratorManager: audioIteratorManager2,
  videoIteratorManager,
  playbackRate
}) => {
  const anchorValue = audioSyncAnchor?.value ?? 0;
  const lines = [
    "Debug overlay",
    `Video iterators created: ${videoIteratorManager?.getVideoIteratorsCreated()}`,
    `Audio iterators created: ${audioIteratorManager2?.getAudioIteratorsCreated()}`,
    `Audio scheduled: ${(audioIteratorManager2?.getTotalAudioScheduledInSeconds() ?? 0).toFixed(3)}s`,
    `Frames rendered: ${videoIteratorManager?.getFramesRendered()}`,
    `Audio context state: ${audioContextState}`,
    audioTime ? `Audio time: ${((audioTime - anchorValue) * playbackRate).toFixed(3)}s` : null
  ].filter(Boolean);
  if (audioIteratorManager2) {
    const queuedPeriod = audioIteratorManager2.getAudioBufferIterator()?.getQueuedPeriod();
    if (queuedPeriod) {
      const aheadText = audioTime ? ` (${(queuedPeriod.until - (audioTime - anchorValue) * playbackRate).toFixed(3)}s ahead)` : "";
      lines.push(`Audio queued until ${queuedPeriod.until.toFixed(3)}s${aheadText}`);
    }
    lines.push(`Playing: ${playing}`);
  }
  const lineHeight = 30;
  const boxPaddingX = 10;
  const boxPaddingY = 10;
  const boxLeft = 20;
  const boxTop = 20;
  const boxWidth = 600;
  const boxHeight = lines.length * lineHeight + 2 * boxPaddingY;
  context.fillStyle = "rgba(0, 0, 0, 1)";
  context.fillRect(boxLeft, boxTop, boxWidth, boxHeight);
  context.fillStyle = "white";
  context.font = "24px sans-serif";
  context.textBaseline = "top";
  for (let i = 0;i < lines.length; i++) {
    context.fillText(lines[i], boxLeft + boxPaddingX, boxTop + boxPaddingY + i * lineHeight);
  }
};

// src/get-duration-or-compute.ts
var getDurationOrCompute = async (input) => {
  return await input.getDurationFromMetadata(undefined, {
    skipLiveWait: true
  }) ?? input.computeDuration(undefined, { skipLiveWait: true });
};

// src/helpers/resolve-audio-track.ts
var resolveAudioTrack = async ({
  videoTrack,
  audioTracks,
  audioStreamIndex
}) => {
  if (audioStreamIndex !== null) {
    return audioTracks[audioStreamIndex] ?? null;
  }
  if (videoTrack) {
    return await videoTrack.getPrimaryPairableAudioTrack() ?? null;
  }
  return audioTracks[0] ?? null;
};

// src/is-type-of-error.ts
function isNetworkError(error) {
  if (error.message.includes("Failed to fetch") || error.message.includes("Load failed") || error.message.includes("NetworkError when attempting to fetch resource")) {
    return true;
  }
  return false;
}
function isUnsupportedConfigurationError(error) {
  return error.message.includes("Unsupported configuration");
}

// src/nonce-manager.ts
var makeNonceManager = () => {
  let nonce = 0;
  const createAsyncOperation = () => {
    nonce++;
    const currentNonce = nonce;
    return {
      isStale: () => nonce !== currentNonce
    };
  };
  return {
    createAsyncOperation
  };
};

// src/premount-aware-delay-playback.ts
class PremountAwareDelayPlayback {
  isPremounting;
  isPostmounting;
  activeHandles = new Set;
  delayPlayback;
  constructor({
    bufferState,
    isPremounting,
    isPostmounting
  }) {
    this.delayPlayback = bufferState.delayPlayback;
    this.isPremounting = isPremounting;
    this.isPostmounting = isPostmounting;
  }
  shouldDelayPlayback() {
    return !this.isPremounting && !this.isPostmounting;
  }
  syncHandles() {
    for (const handle of this.activeHandles) {
      if (this.shouldDelayPlayback()) {
        handle.arm();
      } else {
        handle.disarm();
      }
    }
  }
  setIsPremounting(isPremounting) {
    this.isPremounting = isPremounting;
    this.syncHandles();
  }
  setIsPostmounting(isPostmounting) {
    this.isPostmounting = isPostmounting;
    this.syncHandles();
  }
  createHandle() {
    let armed = false;
    let unblock = null;
    let disposed = false;
    const arm = () => {
      if (armed || disposed) {
        return;
      }
      unblock = this.delayPlayback().unblock;
      armed = true;
    };
    const disarm = () => {
      if (!armed) {
        return;
      }
      unblock?.();
      unblock = null;
      armed = false;
    };
    const entry = {
      arm,
      disarm,
      dispose: () => {}
    };
    entry.dispose = () => {
      if (disposed) {
        return;
      }
      disposed = true;
      disarm();
      this.activeHandles.delete(entry);
    };
    this.activeHandles.add(entry);
    if (this.shouldDelayPlayback()) {
      arm();
    }
    return {
      unblock: entry.dispose,
      [Symbol.dispose]: entry.dispose
    };
  }
}

// src/request-init.ts
var normalizeMediaHeaders = (headers) => {
  if (!headers) {
    return;
  }
  const entries = [];
  if (headers instanceof Headers) {
    headers.forEach((value, key) => {
      entries.push([key.toLowerCase(), value]);
    });
  } else if (Array.isArray(headers)) {
    for (const [key, value] of headers) {
      entries.push([key.toLowerCase(), value]);
    }
  } else {
    for (const [key, value] of Object.entries(headers)) {
      entries.push([key.toLowerCase(), value]);
    }
  }
  entries.sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0);
  return entries;
};
var normalizeMediaRequestInit = (requestInit) => {
  if (!requestInit) {
    return;
  }
  const headers = normalizeMediaHeaders(requestInit.headers);
  const normalized = {
    ...requestInit.cache === undefined ? null : { cache: requestInit.cache },
    ...requestInit.credentials === undefined ? null : { credentials: requestInit.credentials },
    ...headers === undefined ? null : { headers },
    ...requestInit.integrity === undefined ? null : { integrity: requestInit.integrity },
    ...requestInit.mode === undefined ? null : { mode: requestInit.mode },
    ...requestInit.redirect === undefined ? null : { redirect: requestInit.redirect },
    ...requestInit.referrer === undefined ? null : { referrer: requestInit.referrer },
    ...requestInit.referrerPolicy === undefined ? null : { referrerPolicy: requestInit.referrerPolicy }
  };
  return Object.keys(normalized).length === 0 ? undefined : normalized;
};
var getMediaRequestInitFingerprint = (requestInit) => {
  const normalized = normalizeMediaRequestInit(requestInit);
  if (!normalized) {
    return null;
  }
  return [
    normalized.cache ?? null,
    normalized.credentials ?? null,
    normalized.integrity ?? null,
    normalized.mode ?? null,
    normalized.redirect ?? null,
    normalized.referrer ?? null,
    normalized.referrerPolicy ?? null,
    normalized.headers ?? null
  ];
};
var resolveRequestInit = ({
  credentials,
  requestInit
}) => {
  if (credentials === undefined) {
    return normalizeMediaRequestInit(requestInit);
  }
  return normalizeMediaRequestInit({
    credentials,
    ...requestInit
  });
};

// src/video-iterator-manager.ts
import { CanvasSink } from "mediabunny";
import { Internals as Internals4 } from "remotion";

// src/helpers/round-to-4-digits.ts
var roundTo4Digits = (timestamp) => {
  return Math.round(timestamp * 1000) / 1000;
};

// src/canvas-ahead-of-time.ts
var BUFFER_SIZE = 3;
var releaseCanvas = Symbol("releaseCanvas");
var releaseStableFrame = (frame) => {
  frame?.[releaseCanvas]?.();
};
var makeStableFramePool = () => {
  const availableCanvases = [];
  const makeStableFrame = (frame) => {
    const { canvas } = frame;
    const stableCanvas = availableCanvases.pop() ?? new OffscreenCanvas(canvas.width, canvas.height);
    if (stableCanvas.width !== canvas.width) {
      stableCanvas.width = canvas.width;
    }
    if (stableCanvas.height !== canvas.height) {
      stableCanvas.height = canvas.height;
    }
    const context = stableCanvas.getContext("2d");
    if (!context) {
      throw new Error("Could not create canvas context");
    }
    context.drawImage(canvas, 0, 0);
    let released = false;
    const stableFrame = {
      canvas: stableCanvas,
      duration: frame.duration,
      timestamp: frame.timestamp,
      [releaseCanvas]: () => {
        if (released) {
          return;
        }
        released = true;
        availableCanvases.push(stableCanvas);
      }
    };
    return stableFrame;
  };
  return { makeStableFrame };
};
var canvasesAheadOfTime = (videoSink, startTimestamp) => {
  const iterator = videoSink.canvases(startTimestamp);
  const buffer = [];
  let chaining = false;
  let reachedEnd = false;
  let closed = false;
  let inFlight = null;
  const stableFramePool = makeStableFramePool();
  const takeFrame = (frame) => {
    return stableFramePool.makeStableFrame(frame);
  };
  const fillNext = () => {
    if (chaining || reachedEnd || closed)
      return;
    if (buffer.length >= BUFFER_SIZE)
      return;
    chaining = true;
    const slot = { promise: iterator.next(), resolved: null };
    buffer.push(slot);
    inFlight = slot.promise.then((result) => {
      chaining = false;
      inFlight = null;
      if (result.done) {
        slot.resolved = result;
        reachedEnd = true;
        return;
      }
      if (closed) {
        slot.resolved = { done: true, value: undefined };
        return;
      }
      slot.resolved = { ...result, value: takeFrame(result.value) };
      fillNext();
    }, () => {
      chaining = false;
      inFlight = null;
    });
  };
  fillNext();
  const next = () => {
    const slot = buffer.shift();
    fillNext();
    if (!slot) {
      if (reachedEnd || closed) {
        return { type: "ready", frame: null };
      }
      const chain = inFlight;
      return {
        type: "pending",
        wait: async () => {
          await chain;
          const next2 = buffer.shift();
          fillNext();
          if (!next2)
            return null;
          if (next2.resolved) {
            return next2.resolved.done ? null : next2.resolved.value;
          }
          const result = await next2.promise;
          return result.done ? null : takeFrame(result.value);
        }
      };
    }
    if (slot.resolved) {
      if (slot.resolved.done) {
        return { type: "ready", frame: null };
      }
      return { type: "ready", frame: slot.resolved.value };
    }
    return {
      type: "pending",
      wait: async () => {
        const result = await slot.promise;
        if (slot.resolved) {
          return slot.resolved.done ? null : slot.resolved.value;
        }
        return result.done ? null : takeFrame(result.value);
      }
    };
  };
  const closeIterator = async () => {
    closed = true;
    for (const slot of buffer) {
      if (slot.resolved && !slot.resolved.done) {
        releaseStableFrame(slot.resolved.value);
      }
    }
    buffer.length = 0;
    await iterator.return();
  };
  return { next, closeIterator };
};

// src/prewarm-iterator-for-looping.ts
var makePrewarmedVideoIteratorCache = (videoSink) => {
  const prewarmedVideoIterators = new Map;
  const prewarmIteratorForLooping = ({ timeToSeek }) => {
    if (!prewarmedVideoIterators.has(timeToSeek)) {
      prewarmedVideoIterators.set(timeToSeek, canvasesAheadOfTime(videoSink, timeToSeek));
    }
  };
  const makeIteratorOrUsePrewarmed = (timeToSeek) => {
    const prewarmedIterator = prewarmedVideoIterators.get(timeToSeek);
    if (prewarmedIterator) {
      prewarmedVideoIterators.delete(timeToSeek);
      return prewarmedIterator;
    }
    return canvasesAheadOfTime(videoSink, timeToSeek);
  };
  const destroy = () => {
    for (const iterator of prewarmedVideoIterators.values()) {
      iterator.closeIterator();
    }
    prewarmedVideoIterators.clear();
  };
  return {
    prewarmIteratorForLooping,
    makeIteratorOrUsePrewarmed,
    destroy
  };
};

// src/video/video-preview-iterator.ts
var MAXIMUM_AWAITED_PEEK_DISTANCE_SECONDS = 0.05;
var createVideoIterator = async (timeToSeek, cache) => {
  let destroyed = false;
  const iterator = cache.makeIteratorOrUsePrewarmed(timeToSeek);
  let iteratorEnded = false;
  const firstAwait = iterator.next();
  const initialFrame = firstAwait && firstAwait.type === "ready" ? firstAwait.frame : await firstAwait.wait();
  let lastReturnedFrame = initialFrame;
  let peekedFrame = null;
  const setLastReturnedFrame = (frame) => {
    if (lastReturnedFrame !== frame) {
      releaseStableFrame(lastReturnedFrame);
    }
    lastReturnedFrame = frame;
  };
  const setPeekedFrame = (frame) => {
    peekedFrame = frame;
    if (peekedFrame === null) {
      iteratorEnded = true;
    }
    return peekedFrame;
  };
  const peekIfReady = () => {
    if (peekedFrame) {
      return { type: "ready", frame: peekedFrame };
    }
    if (iteratorEnded) {
      return { type: "ready", frame: null };
    }
    const next = iterator.next();
    if (next.type === "ready") {
      return { type: "ready", frame: setPeekedFrame(next.frame) };
    }
    return {
      type: "pending",
      wait: next.wait
    };
  };
  const peek = async () => {
    const peeked = peekIfReady();
    if (peeked.type === "ready") {
      return peeked.frame;
    }
    return setPeekedFrame(await peeked.wait());
  };
  const getFrameEndTimestampFromPeek = (frame) => {
    return frame ? roundTo4Digits(frame.timestamp) : Infinity;
  };
  const getFrameEndTimestamp = async () => {
    return getFrameEndTimestampFromPeek(await peek());
  };
  const getFrameEndTimestampIfCloseEnough = async ({
    timestamp,
    frameTimestamp
  }) => {
    const peeked = peekIfReady();
    if (peeked.type === "ready") {
      return {
        type: "ready",
        timestamp: getFrameEndTimestampFromPeek(peeked.frame)
      };
    }
    if (timestamp - frameTimestamp > MAXIMUM_AWAITED_PEEK_DISTANCE_SECONDS) {
      return { type: "pending" };
    }
    const awaitedPeeked = setPeekedFrame(await peeked.wait());
    return {
      type: "ready",
      timestamp: getFrameEndTimestampFromPeek(awaitedPeeked)
    };
  };
  const getNextOrNullIfNotAvailable = () => {
    if (peekedFrame) {
      const frame = peekedFrame;
      setLastReturnedFrame(frame);
      const retValue = {
        type: "got-frame-or-end",
        frame
      };
      peekedFrame = null;
      return retValue;
    }
    const next = iterator.next();
    if (next.type === "pending") {
      return {
        type: "need-to-wait-for-it",
        waitPromise: async () => {
          const res = await next.wait();
          if (res) {
            setLastReturnedFrame(res);
          } else {
            iteratorEnded = true;
          }
          return res;
        }
      };
    }
    if (next.frame) {
      setLastReturnedFrame(next.frame);
    } else {
      iteratorEnded = true;
    }
    return {
      type: "got-frame-or-end",
      frame: next.frame ?? null
    };
  };
  const destroy = () => {
    destroyed = true;
    releaseStableFrame(lastReturnedFrame);
    if (peekedFrame !== lastReturnedFrame) {
      releaseStableFrame(peekedFrame);
    }
    lastReturnedFrame = null;
    peekedFrame = null;
    iterator.closeIterator().catch(() => {
      return;
    });
  };
  const tryToSatisfySeek = async (time) => {
    const timestamp = roundTo4Digits(time);
    if (lastReturnedFrame) {
      const frameTimestamp = roundTo4Digits(lastReturnedFrame.timestamp);
      if (timestamp < frameTimestamp) {
        const lastFrameWasInitialFrame = lastReturnedFrame === initialFrame;
        const firstFrameDoesSatisfy = lastFrameWasInitialFrame && roundTo4Digits(time) >= roundTo4Digits(timeToSeek);
        if (firstFrameDoesSatisfy) {
          return {
            type: "satisfied",
            frame: lastReturnedFrame
          };
        }
        return {
          type: "not-satisfied",
          reason: `iterator is too far, most recently returned ${frameTimestamp}`
        };
      }
      const frameEndTimestamp = await getFrameEndTimestamp();
      if (frameTimestamp <= timestamp && frameEndTimestamp > timestamp) {
        return {
          type: "satisfied",
          frame: lastReturnedFrame
        };
      }
    }
    if (iteratorEnded) {
      if (lastReturnedFrame) {
        return {
          type: "satisfied",
          frame: lastReturnedFrame
        };
      }
      return {
        type: "not-satisfied",
        reason: "iterator ended"
      };
    }
    while (true) {
      const frame = getNextOrNullIfNotAvailable();
      if (frame.type === "need-to-wait-for-it") {
        return {
          type: "not-satisfied",
          reason: "iterator did not have frame ready"
        };
      }
      if (frame.type === "got-frame-or-end") {
        if (frame.frame === null) {
          iteratorEnded = true;
          if (lastReturnedFrame) {
            return {
              type: "satisfied",
              frame: lastReturnedFrame
            };
          }
          return {
            type: "not-satisfied",
            reason: "iterator ended and did not have frame ready"
          };
        }
        const frameTimestamp = roundTo4Digits(frame.frame.timestamp);
        const frameEndTimestamp = await getFrameEndTimestampIfCloseEnough({
          frameTimestamp,
          timestamp
        });
        if (frameEndTimestamp.type === "pending") {
          return {
            type: "not-satisfied",
            reason: "iterator did not have next frame ready"
          };
        }
        if (frameTimestamp <= timestamp && frameEndTimestamp.timestamp > timestamp) {
          return {
            type: "satisfied",
            frame: frame.frame
          };
        }
        continue;
      }
      throw new Error("Unreachable");
    }
  };
  return {
    destroy,
    initialFrame,
    isDestroyed: () => {
      return destroyed;
    },
    tryToSatisfySeek
  };
};

// src/video-iterator-manager.ts
var { runEffectChain } = Internals4;
var videoIteratorManager = async ({
  delayPlaybackHandleIfNotPremounting,
  canvas,
  context,
  drawDebugOverlay,
  logLevel,
  getOnVideoFrameCallback,
  videoTrack,
  getLoopSegmentMediaEndTimestamp,
  getStartTime,
  getIsLooping,
  getEffects,
  getEffectChainState
}) => {
  let videoIteratorsCreated = 0;
  let videoFrameIterator = null;
  let framesRendered = 0;
  let currentDelayHandle = null;
  let lastDrawnFrame = null;
  let currentSeek = null;
  const clearLastDrawnFrame = () => {
    lastDrawnFrame = null;
  };
  if (canvas) {
    const displayWidth = await videoTrack.getDisplayWidth();
    const displayHeight = await videoTrack.getDisplayHeight();
    if (canvas.width !== displayWidth || canvas.height !== displayHeight) {
      canvas.width = displayWidth;
      canvas.height = displayHeight;
    }
  }
  const canvasSink = new CanvasSink(videoTrack, {
    poolSize: 3,
    fit: "contain",
    alpha: true
  });
  const prewarmedVideoIteratorCache = makePrewarmedVideoIteratorCache(canvasSink);
  const paintFrame = async (frame) => {
    if (context && canvas) {
      const effects = getEffects();
      const chainState = getEffectChainState(canvas.width, canvas.height);
      if (effects.length > 0 && chainState && canvas instanceof HTMLCanvasElement) {
        await runEffectChain({
          state: chainState,
          source: frame.canvas,
          effects,
          output: canvas,
          width: canvas.width,
          height: canvas.height
        });
      } else {
        context.clearRect(0, 0, canvas.width, canvas.height);
        context.drawImage(frame.canvas, 0, 0);
      }
    }
  };
  const drawFrame = async (frame) => {
    await paintFrame(frame);
    lastDrawnFrame = frame;
    framesRendered++;
    drawDebugOverlay();
    const callback = getOnVideoFrameCallback();
    if (callback) {
      callback(frame.canvas);
    }
    Internals4.Log.trace({ logLevel, tag: "@remotion/media" }, `[MediaPlayer] Drew frame ${frame.timestamp.toFixed(3)}s`);
  };
  const redrawCurrentFrame = async () => {
    if (!lastDrawnFrame) {
      return;
    }
    await paintFrame(lastDrawnFrame);
    drawDebugOverlay();
    const callback = getOnVideoFrameCallback();
    if (callback) {
      callback(lastDrawnFrame.canvas);
    }
    Internals4.Log.trace({ logLevel, tag: "@remotion/media" }, `[MediaPlayer] Redrew frame ${lastDrawnFrame.timestamp.toFixed(3)}s with updated effects`);
  };
  const startVideoIterator = async (timeToSeek, nonce) => {
    let __stack = [];
    try {
      clearLastDrawnFrame();
      videoFrameIterator?.destroy();
      const delayHandle = __using(__stack, delayPlaybackHandleIfNotPremounting(), 0);
      currentDelayHandle = delayHandle;
      currentSeek = timeToSeek;
      const iterator = await createVideoIterator(timeToSeek, prewarmedVideoIteratorCache);
      videoIteratorsCreated++;
      videoFrameIterator = iterator;
      if (iterator.isDestroyed()) {
        return;
      }
      if (nonce.isStale()) {
        return;
      }
      if (videoFrameIterator.isDestroyed()) {
        return;
      }
      if (!iterator.initialFrame) {
        return;
      }
      await drawFrame(iterator.initialFrame);
    } catch (_catch) {
      var _err = _catch, _hasErr = 1;
    } finally {
      __callDispose(__stack, _err, _hasErr);
    }
  };
  const seek = async ({ newTime, nonce }) => {
    if (!videoFrameIterator) {
      return;
    }
    if (currentSeek !== null && roundTo4Digits(currentSeek) === roundTo4Digits(newTime)) {
      return;
    }
    currentSeek = newTime;
    if (getIsLooping()) {
      if (getLoopSegmentMediaEndTimestamp() - newTime < 1) {
        prewarmedVideoIteratorCache.prewarmIteratorForLooping({
          timeToSeek: getStartTime()
        });
      }
    }
    const videoSatisfyResult = await videoFrameIterator.tryToSatisfySeek(newTime);
    if (videoSatisfyResult.type === "satisfied") {
      await drawFrame(videoSatisfyResult.frame);
      return;
    }
    if (nonce.isStale()) {
      return;
    }
    await startVideoIterator(newTime, nonce);
  };
  return {
    startVideoIterator,
    getVideoIteratorsCreated: () => videoIteratorsCreated,
    seek,
    destroy: () => {
      clearLastDrawnFrame();
      prewarmedVideoIteratorCache.destroy();
      videoFrameIterator?.destroy();
      if (context && canvas) {
        context.clearRect(0, 0, canvas.width, canvas.height);
      }
      if (currentDelayHandle) {
        currentDelayHandle.unblock();
        currentDelayHandle = null;
      }
      videoFrameIterator = null;
    },
    getVideoFrameIterator: () => videoFrameIterator,
    drawFrame,
    redrawCurrentFrame,
    getFramesRendered: () => framesRendered
  };
};

// src/media-player.ts
class MediaPlayer {
  tagType;
  canvas;
  context;
  src;
  logLevel;
  playbackRate;
  globalPlaybackRate;
  audioStreamIndex;
  sharedAudioContext;
  audioIteratorManager = null;
  videoIteratorManager = null;
  playing = false;
  loop = false;
  fps;
  trimBefore;
  trimAfter;
  sequenceDurationInFrames;
  sequenceOffset;
  totalDuration;
  debugOverlay = false;
  nonceManager;
  onVideoFrameCallback = null;
  getEffects;
  getEffectChainState;
  initializationPromise = null;
  premountAwareDelayPlayback;
  seekPromiseChain = Promise.resolve();
  constructor({
    canvas,
    src,
    logLevel,
    sharedAudioContext,
    loop,
    trimBefore,
    trimAfter,
    playbackRate,
    globalPlaybackRate,
    audioStreamIndex,
    fps,
    debugOverlay,
    bufferState,
    isPremounting,
    isPostmounting,
    durationInFrames,
    onVideoFrameCallback,
    playing,
    sequenceOffset,
    credentials,
    requestInit,
    tagType,
    getEffects,
    getEffectChainState
  }) {
    this.canvas = canvas ?? null;
    this.src = src;
    this.logLevel = logLevel;
    this.sharedAudioContext = sharedAudioContext;
    this.playbackRate = playbackRate;
    this.globalPlaybackRate = globalPlaybackRate;
    this.loop = loop;
    this.trimBefore = trimBefore;
    this.trimAfter = trimAfter;
    this.audioStreamIndex = audioStreamIndex;
    this.fps = fps;
    this.debugOverlay = debugOverlay;
    this.premountAwareDelayPlayback = new PremountAwareDelayPlayback({
      bufferState,
      isPremounting,
      isPostmounting
    });
    this.sequenceDurationInFrames = durationInFrames;
    this.nonceManager = makeNonceManager();
    this.onVideoFrameCallback = onVideoFrameCallback;
    this.playing = playing;
    this.sequenceOffset = sequenceOffset;
    const resolvedRequestInit = resolveRequestInit({ credentials, requestInit });
    this.input = new Input({
      source: new UrlSource(this.src, resolvedRequestInit ? {
        requestInit: resolvedRequestInit
      } : undefined),
      formats: ALL_FORMATS
    });
    this.tagType = tagType;
    this.getEffects = getEffects;
    this.getEffectChainState = getEffectChainState;
    if (canvas) {
      const context = canvas.getContext("2d", {
        alpha: true,
        desynchronized: true
      });
      if (!context) {
        throw new Error("Could not get 2D context from canvas");
      }
      this.context = context;
    } else {
      this.context = null;
    }
  }
  input;
  isDisposalError() {
    return this.input.disposed === true;
  }
  initialize(startTimeUnresolved, initialMuted) {
    const promise = this._initialize(startTimeUnresolved, initialMuted);
    this.initializationPromise = promise;
    this.seekPromiseChain = promise;
    return promise;
  }
  getStartTime() {
    return (this.trimBefore ?? 0) / this.fps;
  }
  getSequenceEndTimestamp() {
    return this.sequenceDurationInFrames / this.fps * this.playbackRate + this.getStartTime();
  }
  getSequenceDurationInSeconds() {
    return this.sequenceDurationInFrames / this.fps;
  }
  getMediaEndTimestamp() {
    return calculateEndTime({
      mediaDurationInSeconds: this.totalDuration,
      ifNoMediaDuration: "fail",
      src: this.src,
      trimAfter: this.trimAfter,
      trimBefore: this.trimBefore,
      fps: this.fps
    });
  }
  getLoopSegmentMediaEndTimestamp() {
    return Math.min(this.getMediaEndTimestamp(), this.getSequenceEndTimestamp());
  }
  async _initialize(startTimeUnresolved, initialMuted) {
    let __stack = [];
    try {
      const _ = __using(__stack, this.delayPlaybackHandleIfNotPremounting(), 0);
      try {
        if (this.input.disposed) {
          return { type: "disposed" };
        }
        try {
          await this.input.getFormat();
        } catch (error) {
          if (this.isDisposalError()) {
            return { type: "disposed" };
          }
          const err = error;
          if (isNetworkError(err)) {
            throw error;
          }
          Internals5.Log.error({ logLevel: this.logLevel, tag: "@remotion/media" }, `[MediaPlayer] Failed to recognize format for ${this.src}`, error);
          return { type: "unknown-container-format" };
        }
        const [durationInSeconds, videoTrack, audioTracks] = await Promise.all([
          getDurationOrCompute(this.input),
          this.input.getPrimaryVideoTrack(),
          this.input.getAudioTracks()
        ]);
        if (this.input.disposed) {
          return { type: "disposed" };
        }
        this.totalDuration = durationInSeconds;
        const audioTrack = await resolveAudioTrack({
          videoTrack,
          audioTracks,
          audioStreamIndex: this.audioStreamIndex
        });
        if (!videoTrack && !audioTrack) {
          return { type: "no-tracks" };
        }
        if (videoTrack && this.tagType === "video") {
          if (await videoTrack.isLive()) {
            throw new Error("Live streams are not currently supported by Remotion. Sorry! Source: " + this.src);
          }
          if (await videoTrack.isRelativeToUnixEpoch()) {
            throw new Error("Streams with UNIX timestamps are not currently supported by Remotion. Sorry! Source: " + this.src);
          }
          const canDecode = await videoTrack.canDecode();
          if (!canDecode) {
            return { type: "cannot-decode" };
          }
          if (this.input.disposed) {
            return { type: "disposed" };
          }
          this.videoIteratorManager = await videoIteratorManager({
            videoTrack,
            delayPlaybackHandleIfNotPremounting: this.delayPlaybackHandleIfNotPremounting,
            context: this.context,
            canvas: this.canvas,
            getOnVideoFrameCallback: () => this.onVideoFrameCallback,
            logLevel: this.logLevel,
            drawDebugOverlay: this.drawDebugOverlay,
            getLoopSegmentMediaEndTimestamp: () => this.getLoopSegmentMediaEndTimestamp(),
            getStartTime: () => this.getStartTime(),
            getIsLooping: () => this.loop,
            getEffects: this.getEffects,
            getEffectChainState: this.getEffectChainState
          });
        }
        const startTime = this.getTrimmedTime(startTimeUnresolved);
        if (startTime === null) {
          throw new Error(`should have asserted that the time is not null`);
        }
        if (audioTrack && this.sharedAudioContext) {
          if (await audioTrack.isLive()) {
            throw new Error("Live streams are not currently supported by Remotion. Sorry! Source: " + this.src);
          }
          if (await audioTrack.isRelativeToUnixEpoch()) {
            throw new Error("Streams with UNIX timestamps are not currently supported by Remotion. Sorry! Source: " + this.src);
          }
          const canDecode = await audioTrack.canDecode();
          if (!canDecode) {
            return { type: "cannot-decode" };
          }
          if (this.input.disposed) {
            return { type: "disposed" };
          }
          this.audioIteratorManager = audioIteratorManager({
            audioTrack,
            delayPlaybackHandleIfNotPremounting: this.delayPlaybackHandleIfNotPremounting,
            sharedAudioContext: this.sharedAudioContext,
            getMediaEndTimestamp: () => this.getMediaEndTimestamp(),
            getSequenceEndTimestamp: () => this.getSequenceEndTimestamp(),
            getStartTime: () => this.getStartTime(),
            initialMuted,
            drawDebugOverlay: this.drawDebugOverlay,
            initialPlaybackRate: this.playbackRate * this.globalPlaybackRate,
            getSequenceDurationInSeconds: () => this.getSequenceDurationInSeconds(),
            initialTrimBefore: this.trimBefore,
            initialTrimAfter: this.trimAfter,
            initialSequenceOffset: this.sequenceOffset,
            initialSequenceDurationInFrames: this.sequenceDurationInFrames,
            initialLoop: this.loop,
            initialFps: this.fps
          });
        }
        const nonce = this.nonceManager.createAsyncOperation();
        try {
          await Promise.all([
            this.audioIteratorManager ? this.audioIteratorManager.startAudioIterator({
              nonce,
              playbackRate: this.playbackRate * this.globalPlaybackRate,
              startFromSecond: startTime,
              unloopedStartFromSecond: startTimeUnresolved,
              scheduleAudioNode: this.scheduleAudioNode,
              getTargetTime: this.getTargetTime,
              logLevel: this.logLevel,
              loop: this.loop,
              unscheduleAudioNode: this.sharedAudioContext.unscheduleAudioNode,
              getAudioContextCurrentTimeMockedInTest: () => this.sharedAudioContext.audioContext.currentTime
            }) : Promise.resolve(),
            this.videoIteratorManager ? this.videoIteratorManager.startVideoIterator(startTime, nonce) : Promise.resolve()
          ]);
        } catch (error) {
          if (this.isDisposalError()) {
            return { type: "disposed" };
          }
          Internals5.Log.error({ logLevel: this.logLevel, tag: "@remotion/media" }, "[MediaPlayer] Failed to start audio and video iterators", error);
        }
        return { type: "success", durationInSeconds };
      } catch (error) {
        const err = error;
        if (isNetworkError(err)) {
          Internals5.Log.error({ logLevel: this.logLevel, tag: "@remotion/media" }, `[MediaPlayer] Network/CORS error for ${this.src}`, err);
          return { type: "network-error" };
        }
        Internals5.Log.error({ logLevel: this.logLevel, tag: "@remotion/media" }, "[MediaPlayer] Failed to initialize", error);
        throw error;
      }
    } catch (_catch) {
      var _err = _catch, _hasErr = 1;
    } finally {
      __callDispose(__stack, _err, _hasErr);
    }
  }
  seekToWithQueue = async (newTime, unloopedNewTime) => {
    const nonce = this.nonceManager.createAsyncOperation();
    await this.seekPromiseChain;
    this.seekPromiseChain = this.seekToDoNotCallDirectly(newTime, unloopedNewTime, nonce);
    await this.seekPromiseChain;
  };
  async seekTo(time) {
    const newTime = this.getTrimmedTime(time);
    if (newTime === null) {
      throw new Error(`should have asserted that the time is not null`);
    }
    await this.seekToWithQueue(newTime, time);
  }
  async seekToDoNotCallDirectly(newTime, unloopedNewTime, nonce) {
    if (nonce.isStale()) {
      return;
    }
    try {
      await Promise.all([
        this.videoIteratorManager?.seek({
          newTime,
          nonce
        }),
        this.audioIteratorManager?.seek({
          newTime,
          unloopedNewTime,
          nonce,
          playbackRate: this.playbackRate * this.globalPlaybackRate,
          localPlaybackRate: this.playbackRate,
          getTargetTime: this.getTargetTime,
          logLevel: this.logLevel,
          loop: this.loop,
          trimBefore: this.trimBefore,
          trimAfter: this.trimAfter,
          sequenceOffset: this.sequenceOffset,
          sequenceDurationInFrames: this.sequenceDurationInFrames,
          fps: this.fps,
          scheduleAudioNode: this.scheduleAudioNode,
          getAudioContextCurrentTimeMockedInTest: () => this.sharedAudioContext.audioContext.currentTime
        })
      ]);
    } catch (error) {
      if (this.isDisposalError()) {
        return;
      }
      throw error;
    }
  }
  play() {
    if (this.playing) {
      return;
    }
    this.playing = true;
    this.drawDebugOverlay();
  }
  delayPlaybackHandleIfNotPremounting = () => {
    return this.premountAwareDelayPlayback.createHandle();
  };
  pause() {
    if (!this.playing) {
      return;
    }
    this.playing = false;
    this.drawDebugOverlay();
  }
  setMuted(muted) {
    this.audioIteratorManager?.setMuted(muted);
  }
  setVolume(volume) {
    if (!this.audioIteratorManager) {
      return;
    }
    this.audioIteratorManager.setVolume(volume);
  }
  getTrimmedTime(unloopedTimeInSeconds) {
    return getTimeInSeconds({
      unloopedTimeInSeconds,
      playbackRate: this.playbackRate,
      loop: this.loop,
      trimBefore: this.trimBefore,
      trimAfter: this.trimAfter,
      mediaDurationInSeconds: this.totalDuration ?? null,
      fps: this.fps,
      ifNoMediaDuration: "infinity",
      src: this.src
    });
  }
  async setTrimBefore(trimBefore, unloopedTimeInSeconds) {
    if (this.trimBefore !== trimBefore) {
      this.trimBefore = trimBefore;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  async setTrimAfter(trimAfter, unloopedTimeInSeconds) {
    if (this.trimAfter !== trimAfter) {
      this.trimAfter = trimAfter;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  setDebugOverlay(debugOverlay) {
    this.debugOverlay = debugOverlay;
  }
  async setPlaybackRate(rate, unloopedTimeInSeconds) {
    const previousRate = this.playbackRate;
    if (previousRate !== rate) {
      this.playbackRate = rate;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  async setGlobalPlaybackRate(rate, unloopedTimeInSeconds) {
    const previousRate = this.globalPlaybackRate;
    if (previousRate !== rate) {
      this.globalPlaybackRate = rate;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  async setFps(fps, unloopedTimeInSeconds) {
    const previousFps = this.fps;
    if (previousFps !== fps) {
      this.fps = fps;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  setIsPremounting(isPremounting) {
    this.premountAwareDelayPlayback.setIsPremounting(isPremounting);
  }
  setIsPostmounting(isPostmounting) {
    this.premountAwareDelayPlayback.setIsPostmounting(isPostmounting);
  }
  async setLoop(loop, unloopedTimeInSeconds) {
    const previousLoop = this.loop;
    if (previousLoop !== loop) {
      this.loop = loop;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  async setSequenceOffset(offset, unloopedTimeInSeconds) {
    const previousOffset = this.sequenceOffset;
    if (previousOffset !== offset) {
      this.sequenceOffset = offset;
      this.audioIteratorManager?.destroyIterator();
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  async setSequenceDurationInFrames(sequenceDurationInFrames, unloopedTimeInSeconds) {
    const previousDuration = this.sequenceDurationInFrames;
    if (previousDuration !== sequenceDurationInFrames) {
      this.sequenceDurationInFrames = sequenceDurationInFrames;
      await this.seekTo(unloopedTimeInSeconds);
    }
  }
  async dispose() {
    if (this.initializationPromise) {
      try {
        await this.initializationPromise;
      } catch {}
    }
    this.nonceManager.createAsyncOperation();
    this.videoIteratorManager?.destroy();
    this.audioIteratorManager?.destroyIterator();
    this.input.dispose();
  }
  getTargetTime = (mediaTimestamp, currentTime) => {
    if (!this.sharedAudioContext) {
      throw new Error("Shared audio context not found");
    }
    const globalTime = (currentTime - this.sharedAudioContext.audioSyncAnchor.value) * this.globalPlaybackRate;
    const timeInSeconds = globalTime - this.sequenceOffset;
    const anchor = this.loop ? this.audioIteratorManager?.getCurrentAnchor() ?? null : null;
    const localTime = anchor ? anchorToContinuousTime({
      anchor,
      unloopedTimeInSeconds: timeInSeconds,
      playbackRate: this.playbackRate
    }) : this.getTrimmedTime(timeInSeconds);
    if (localTime === null) {
      return null;
    }
    const targetTime = (mediaTimestamp - localTime) / (this.playbackRate * this.globalPlaybackRate);
    return targetTime;
  };
  scheduleAudioNode = (node, mediaTimestamp, originalUnloopedMediaTimestamp, sourceOffsetInSeconds, sourceDurationInSeconds) => {
    if (!this.sharedAudioContext) {
      throw new Error("Shared audio context not found");
    }
    const targetTime = this.getTargetTime(mediaTimestamp, this.sharedAudioContext.audioContext.currentTime);
    const combinedPlaybackRate = this.playbackRate * this.globalPlaybackRate;
    if (targetTime === null) {
      return {
        type: "not-started",
        reason: "no target for" + mediaTimestamp.toFixed(3) + "," + this.sharedAudioContext.audioContext.currentTime.toFixed(3)
      };
    }
    const sequenceStartTime = this.getStartTime();
    const offset = getTrimStartForAudioNode({
      mediaTimestamp,
      targetTime,
      sequenceStartTime,
      combinedPlaybackRate,
      sourceStartOffsetInSeconds: sourceOffsetInSeconds
    });
    const duration = getDurationOfNode({
      sourceDurationInSeconds,
      sourceOffsetInSeconds,
      offset
    });
    const scheduledTime = getScheduledTime({
      mediaTimestamp,
      targetTime,
      sequenceStartTime,
      currentTime: this.sharedAudioContext.audioContext.currentTime
    });
    return this.sharedAudioContext.scheduleAudioNode({
      node,
      mediaTimestamp,
      sourceOffset: sourceOffsetInSeconds,
      scheduledTime,
      duration,
      offset,
      originalUnloopedMediaTimestamp
    });
  };
  setVideoFrameCallback(callback) {
    this.onVideoFrameCallback = callback;
  }
  async redrawVideoEffects() {
    await this.videoIteratorManager?.redrawCurrentFrame();
  }
  drawDebugOverlay = () => {
    if (!this.debugOverlay)
      return;
    if (this.context && this.canvas) {
      drawPreviewOverlay({
        context: this.context,
        audioTime: this.sharedAudioContext?.audioContext.currentTime ?? null,
        audioContextState: this.sharedAudioContext?.audioContext.state ?? null,
        audioSyncAnchor: this.sharedAudioContext?.audioSyncAnchor ?? null,
        audioIteratorManager: this.audioIteratorManager,
        playing: this.playing,
        videoIteratorManager: this.videoIteratorManager,
        playbackRate: this.playbackRate * this.globalPlaybackRate
      });
    }
  };
  audioSyncAnchorChanged = () => {
    if (!this.audioIteratorManager) {
      return;
    }
    this.audioIteratorManager.destroyIterator();
  };
}

// src/on-error.ts
var callOnErrorAndResolve = ({
  onError,
  error,
  disallowFallback,
  isClientSideRendering,
  clientSideError
}) => {
  const result = onError?.(error);
  if (isClientSideRendering) {
    return ["fail", clientSideError];
  }
  if (result) {
    return [result, error];
  }
  if (disallowFallback) {
    return ["fail", error];
  }
  return ["fallback", error];
};

// src/use-common-effects.ts
import { useContext, useLayoutEffect } from "react";
import { Internals as Internals6 } from "remotion";
var useCommonEffects = ({
  mediaPlayerRef,
  mediaPlayerReady,
  currentTimeRef,
  playing,
  isPlayerBuffering,
  frame,
  trimBefore,
  trimAfter,
  effectiveMuted,
  userPreferredVolume,
  playbackRate,
  globalPlaybackRate,
  fps,
  sequenceOffset,
  loop,
  durationInFrames,
  isPremounting,
  isPostmounting,
  currentTime,
  logLevel,
  label
}) => {
  const sharedAudioContext = useContext(Internals6.SharedAudioContext);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer)
      return;
    if (playing && !isPlayerBuffering) {
      mediaPlayer.play();
    } else {
      mediaPlayer.pause();
    }
  }, [
    isPlayerBuffering,
    playing,
    logLevel,
    mediaPlayerReady,
    frame,
    mediaPlayerRef
  ]);
  useLayoutEffect(() => {
    if (!sharedAudioContext)
      return;
    const { remove } = sharedAudioContext.audioSyncAnchorEmitter.subscribe((event) => {
      if (event === "changed") {
        mediaPlayerRef.current?.audioSyncAnchorChanged();
      }
    });
    return () => {
      remove();
    };
  }, [sharedAudioContext, mediaPlayerRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setTrimBefore(trimBefore, currentTimeRef.current);
  }, [trimBefore, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setTrimAfter(trimAfter, currentTimeRef.current);
  }, [trimAfter, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady)
      return;
    mediaPlayer.setMuted(effectiveMuted);
  }, [effectiveMuted, mediaPlayerReady, mediaPlayerRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setVolume(userPreferredVolume);
  }, [userPreferredVolume, mediaPlayerReady, mediaPlayerRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setPlaybackRate(playbackRate, currentTimeRef.current);
  }, [playbackRate, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setGlobalPlaybackRate(globalPlaybackRate, currentTimeRef.current);
  }, [globalPlaybackRate, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setLoop(loop, currentTimeRef.current);
  }, [loop, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setSequenceDurationInFrames(durationInFrames, currentTimeRef.current);
  }, [durationInFrames, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer) {
      return;
    }
    mediaPlayer.setIsPremounting(isPremounting);
  }, [isPremounting, mediaPlayerReady, mediaPlayerRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer) {
      return;
    }
    mediaPlayer.setIsPostmounting(isPostmounting);
  }, [isPostmounting, mediaPlayerReady, mediaPlayerRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setFps(fps, currentTimeRef.current);
  }, [fps, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setSequenceOffset(sequenceOffset, currentTimeRef.current);
  }, [sequenceOffset, mediaPlayerReady, mediaPlayerRef, currentTimeRef]);
  useLayoutEffect(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady)
      return;
    mediaPlayer.seekTo(currentTime).catch(() => {});
    Internals6.Log.trace({ logLevel, tag: "@remotion/media" }, `[${label}] Updating target time to ${currentTime.toFixed(3)}s`);
  }, [currentTime, logLevel, mediaPlayerReady, label, mediaPlayerRef]);
};

// src/audio/audio-for-preview.tsx
import { jsx } from "react/jsx-runtime";
var {
  useUnsafeVideoConfig,
  Timeline,
  SharedAudioContext,
  usePlayerMutedState,
  useMediaVolumeState,
  useFrameForVolumeProp,
  evaluateVolume,
  warnAboutTooHighVolume,
  usePreload,
  SequenceContext
} = Internals7;
var AudioForPreviewAssertedShowing = ({
  src,
  playbackRate,
  logLevel,
  muted,
  volume,
  loopVolumeCurveBehavior,
  loop,
  trimAfter,
  trimBefore,
  name,
  showInTimeline,
  stack,
  disallowFallbackToHtml5Audio,
  toneFrequency,
  audioStreamIndex,
  fallbackHtml5AudioProps,
  onError,
  credentials,
  requestInit,
  setMediaDurationInSeconds
}) => {
  const videoConfig = useUnsafeVideoConfig();
  const frame = useCurrentFrame();
  const mediaPlayerRef = useRef(null);
  const initialTrimBeforeRef = useRef(trimBefore);
  const initialTrimAfterRef = useRef(trimAfter);
  const [initialRequestInit] = useState(requestInit);
  const [mediaPlayerReady, setMediaPlayerReady] = useState(false);
  const [shouldFallbackToNativeAudio, setShouldFallbackToNativeAudio] = useState(false);
  const [playing] = Timeline.usePlayingState();
  const { playbackRate: globalPlaybackRate } = Internals7.usePlaybackRate();
  const sharedAudioContext = useContext2(SharedAudioContext);
  const buffer = useBufferState();
  const [playerMuted] = usePlayerMutedState();
  const [mediaVolume] = useMediaVolumeState();
  const volumePropFrame = useFrameForVolumeProp(loopVolumeCurveBehavior ?? "repeat");
  const userPreferredVolume = evaluateVolume({
    frame: volumePropFrame,
    volume,
    mediaVolume
  });
  warnAboutTooHighVolume(userPreferredVolume);
  if (!videoConfig) {
    throw new Error("No video config found");
  }
  if (!src) {
    throw new TypeError("No `src` was passed to <NewAudioForPreview>.");
  }
  const currentTime = frame / videoConfig.fps;
  const currentTimeRef = useRef(currentTime);
  currentTimeRef.current = currentTime;
  const preloadedSrc = usePreload(src);
  const parentSequence = useContext2(SequenceContext);
  const isPremounting = Boolean(parentSequence?.premounting);
  const isPostmounting = Boolean(parentSequence?.postmounting);
  const sequenceOffset = (parentSequence?.absoluteFrom ?? 0) / videoConfig.fps;
  const bufferingContext = useContext2(Internals7.BufferingContextReact);
  if (!bufferingContext) {
    throw new Error("useMediaPlayback must be used inside a <BufferingContext>");
  }
  const effectiveMuted = muted || playerMuted || userPreferredVolume <= 0;
  const isPlayerBuffering = Internals7.useIsPlayerBuffering(bufferingContext);
  const initialPlaying = useRef(playing && !isPlayerBuffering);
  const initialIsPremounting = useRef(isPremounting);
  const initialIsPostmounting = useRef(isPostmounting);
  const initialGlobalPlaybackRate = useRef(globalPlaybackRate);
  const initialPlaybackRate = useRef(playbackRate);
  const initialMuted = useRef(effectiveMuted);
  const initialDurationInFrames = useRef(videoConfig.durationInFrames);
  const initialSequenceOffset = useRef(sequenceOffset);
  useCommonEffects({
    mediaPlayerRef,
    mediaPlayerReady,
    currentTimeRef,
    playing,
    isPlayerBuffering,
    frame,
    trimBefore,
    trimAfter,
    effectiveMuted,
    userPreferredVolume,
    playbackRate,
    globalPlaybackRate,
    fps: videoConfig.fps,
    sequenceOffset,
    loop,
    durationInFrames: videoConfig.durationInFrames,
    isPremounting,
    isPostmounting,
    currentTime,
    logLevel,
    label: "AudioForPreview"
  });
  useEffect(() => {
    if (!sharedAudioContext)
      return;
    if (!sharedAudioContext.audioContext)
      return;
    const {
      audioContext,
      gainNode,
      audioSyncAnchor,
      scheduleAudioNode,
      unscheduleAudioNode
    } = sharedAudioContext;
    if (!gainNode) {
      return;
    }
    try {
      const player = new MediaPlayer({
        src: preloadedSrc,
        logLevel,
        sharedAudioContext: {
          audioContext,
          gainNode,
          audioSyncAnchor,
          scheduleAudioNode,
          unscheduleAudioNode
        },
        loop,
        trimAfter: initialTrimAfterRef.current,
        trimBefore: initialTrimBeforeRef.current,
        fps: videoConfig.fps,
        canvas: null,
        playbackRate: initialPlaybackRate.current,
        audioStreamIndex: audioStreamIndex ?? null,
        debugOverlay: false,
        bufferState: buffer,
        isPostmounting: initialIsPostmounting.current,
        isPremounting: initialIsPremounting.current,
        globalPlaybackRate: initialGlobalPlaybackRate.current,
        durationInFrames: initialDurationInFrames.current,
        onVideoFrameCallback: null,
        playing: initialPlaying.current,
        sequenceOffset: initialSequenceOffset.current,
        credentials,
        requestInit: initialRequestInit,
        tagType: "audio",
        getEffects: () => [],
        getEffectChainState: () => null
      });
      mediaPlayerRef.current = player;
      player.initialize(currentTimeRef.current, initialMuted.current).then((result) => {
        if (result.type === "disposed") {
          return;
        }
        const handleError = (error, fallbackMessage) => {
          const [action, errorToUse] = callOnErrorAndResolve({
            onError,
            error,
            disallowFallback: disallowFallbackToHtml5Audio,
            isClientSideRendering: false,
            clientSideError: error
          });
          if (action === "fail") {
            throw errorToUse;
          } else {
            Internals7.Log.warn({ logLevel, tag: "@remotion/media" }, fallbackMessage);
            setShouldFallbackToNativeAudio(true);
          }
        };
        if (result.type === "unknown-container-format") {
          handleError(new Error(`Unknown container format ${preloadedSrc}.`), `Unknown container format for ${preloadedSrc} (Supported formats: https://www.remotion.dev/docs/mediabunny/formats), falling back to <Html5Audio>`);
          return;
        }
        if (result.type === "network-error") {
          handleError(new Error(`Network error fetching ${preloadedSrc}.`), `Network error fetching ${preloadedSrc}, falling back to <Html5Audio>`);
          return;
        }
        if (result.type === "cannot-decode") {
          handleError(new Error(`Cannot decode ${preloadedSrc}.`), `Cannot decode ${preloadedSrc}, falling back to <Html5Audio>`);
          return;
        }
        if (result.type === "no-tracks") {
          handleError(new Error(`No video or audio tracks found for ${preloadedSrc}.`), `No video or audio tracks found for ${preloadedSrc}, falling back to <Html5Audio>`);
          return;
        }
        if (result.type === "success") {
          setMediaPlayerReady(true);
          setMediaDurationInSeconds(result.durationInSeconds);
          Internals7.Log.trace({ logLevel, tag: "@remotion/media" }, `[AudioForPreview] MediaPlayer initialized successfully`);
        }
      }).catch((error) => {
        const [action, errorToUse] = callOnErrorAndResolve({
          onError,
          error,
          disallowFallback: disallowFallbackToHtml5Audio,
          isClientSideRendering: false,
          clientSideError: error
        });
        if (action === "fail") {
          throw errorToUse;
        } else {
          Internals7.Log.error({ logLevel, tag: "@remotion/media" }, "[AudioForPreview] Failed to initialize MediaPlayer", error);
          setShouldFallbackToNativeAudio(true);
        }
      });
    } catch (error) {
      const [action, errorToUse] = callOnErrorAndResolve({
        error,
        onError,
        disallowFallback: disallowFallbackToHtml5Audio,
        isClientSideRendering: false,
        clientSideError: error
      });
      if (action === "fail") {
        throw errorToUse;
      }
      Internals7.Log.error({ logLevel, tag: "@remotion/media" }, "[AudioForPreview] MediaPlayer initialization failed", errorToUse);
      setShouldFallbackToNativeAudio(true);
    }
    return () => {
      if (mediaPlayerRef.current) {
        Internals7.Log.trace({ logLevel, tag: "@remotion/media" }, `[AudioForPreview] Disposing MediaPlayer`);
        mediaPlayerRef.current.dispose();
        mediaPlayerRef.current = null;
      }
      setMediaPlayerReady(false);
      setShouldFallbackToNativeAudio(false);
    };
  }, [
    preloadedSrc,
    logLevel,
    sharedAudioContext,
    loop,
    videoConfig.fps,
    audioStreamIndex,
    disallowFallbackToHtml5Audio,
    buffer,
    onError,
    credentials,
    initialRequestInit,
    setMediaDurationInSeconds
  ]);
  if (shouldFallbackToNativeAudio && !disallowFallbackToHtml5Audio) {
    return /* @__PURE__ */ jsx(RemotionAudio, {
      src,
      muted,
      volume,
      startFrom: trimBefore,
      endAt: trimAfter,
      playbackRate,
      loopVolumeCurveBehavior,
      name,
      loop,
      showInTimeline,
      stack: stack ?? undefined,
      toneFrequency,
      audioStreamIndex,
      pauseWhenBuffering: fallbackHtml5AudioProps?.pauseWhenBuffering,
      crossOrigin: fallbackHtml5AudioProps?.crossOrigin,
      ...fallbackHtml5AudioProps
    });
  }
  return null;
};
var AudioForPreview = ({
  loop = false,
  src,
  logLevel,
  muted,
  name,
  volume,
  loopVolumeCurveBehavior,
  playbackRate = 1,
  trimAfter,
  trimBefore,
  showInTimeline,
  stack,
  disallowFallbackToHtml5Audio,
  toneFrequency,
  audioStreamIndex,
  fallbackHtml5AudioProps,
  onError,
  credentials,
  requestInit,
  setMediaDurationInSeconds
}) => {
  const preloadedSrc = usePreload(src);
  const defaultLogLevel = Internals7.useLogLevel();
  const frame = useCurrentFrame();
  const videoConfig = useVideoConfig();
  const currentTime = frame / videoConfig.fps;
  const showShow = useMemo(() => {
    return getTimeInSeconds({
      unloopedTimeInSeconds: currentTime,
      playbackRate,
      loop,
      trimBefore,
      trimAfter,
      mediaDurationInSeconds: Infinity,
      fps: videoConfig.fps,
      ifNoMediaDuration: "infinity",
      src
    }) !== null;
  }, [
    currentTime,
    playbackRate,
    src,
    trimAfter,
    trimBefore,
    videoConfig.fps,
    loop
  ]);
  if (!showShow) {
    return null;
  }
  if (!setMediaDurationInSeconds) {
    throw new Error("setMediaDurationInSeconds is required");
  }
  return /* @__PURE__ */ jsx(AudioForPreviewAssertedShowing, {
    audioStreamIndex,
    src: preloadedSrc,
    playbackRate,
    logLevel: logLevel ?? defaultLogLevel,
    muted: muted ?? false,
    volume: volume ?? 1,
    loopVolumeCurveBehavior: loopVolumeCurveBehavior ?? "repeat",
    loop,
    trimAfter,
    trimBefore,
    name,
    showInTimeline: showInTimeline ?? true,
    stack,
    disallowFallbackToHtml5Audio: disallowFallbackToHtml5Audio ?? false,
    toneFrequency,
    onError,
    credentials,
    requestInit,
    fallbackHtml5AudioProps,
    setMediaDurationInSeconds
  });
};

// src/audio/audio-for-rendering.tsx
import { useContext as useContext3, useLayoutEffect as useLayoutEffect2, useMemo as useMemo2, useState as useState2 } from "react";
import {
  cancelRender as cancelRender2,
  Html5Audio,
  Internals as Internals16,
  random,
  useCurrentFrame as useCurrentFrame2,
  useDelayRender,
  useRemotionEnvironment
} from "remotion";

// src/caches.ts
import React2 from "react";
import { cancelRender, Internals as Internals12 } from "remotion";

// src/audio-extraction/audio-manager.ts
import { Internals as Internals9 } from "remotion";

// src/audio-extraction/audio-iterator.ts
import { Internals as Internals8 } from "remotion";

// src/audio-extraction/audio-cache.ts
var makeAudioCache = () => {
  const timestamps = [];
  const samples = {};
  const addFrame = (sample) => {
    timestamps.push(sample.timestamp);
    samples[sample.timestamp] = sample;
  };
  const clearBeforeThreshold = (threshold) => {
    for (const timestamp of timestamps.slice()) {
      const endTimestamp = timestamp + samples[timestamp].duration;
      if (endTimestamp < threshold) {
        const isLast = timestamp === timestamps[timestamps.length - 1];
        if (isLast) {
          continue;
        }
        samples[timestamp].close();
        delete samples[timestamp];
        timestamps.splice(timestamps.indexOf(timestamp), 1);
      }
    }
  };
  const deleteAll = () => {
    for (const timestamp of timestamps) {
      samples[timestamp].close();
      delete samples[timestamp];
    }
    timestamps.length = 0;
  };
  const getSamples = (timestamp, durationInSeconds) => {
    const selected = [];
    for (let i = 0;i < timestamps.length; i++) {
      const sampleTimestamp = timestamps[i];
      const sample = samples[sampleTimestamp];
      if (sample.timestamp + sample.duration - 0.0000000001 <= timestamp) {
        continue;
      }
      if (sample.timestamp >= timestamp + durationInSeconds - 0.0000000001) {
        break;
      }
      selected.push(sample);
    }
    return selected;
  };
  const getOpenTimestamps = () => {
    return timestamps;
  };
  const getOldestTimestamp = () => {
    return timestamps[0];
  };
  const getNewestTimestamp = () => {
    if (timestamps.length === 0) {
      return null;
    }
    const sample = samples[timestamps[timestamps.length - 1]];
    return sample.timestamp + sample.duration;
  };
  return {
    addFrame,
    clearBeforeThreshold,
    deleteAll,
    getSamples,
    getOldestTimestamp,
    getNewestTimestamp,
    getOpenTimestamps
  };
};

// src/audio-extraction/audio-iterator.ts
var EXTRA_THRESHOLD_IN_SECONDS = 1.5;
var safetyOutOfOrderThreshold = 0.2;
var warned = {};
var warnAboutMatroskaOnce = (src, logLevel) => {
  if (warned[src]) {
    return;
  }
  warned[src] = true;
  Internals8.Log.warn({ logLevel, tag: "@remotion/media" }, `Audio from ${src} will need to be read from the beginning. https://www.remotion.dev/docs/media/support#matroska-limitation`);
};
var makeAudioIterator2 = ({
  audioSampleSink,
  isMatroska,
  startTimestamp,
  src,
  actualMatroskaTimestamps,
  logLevel
}) => {
  const sampleIterator = audioSampleSink.samples(isMatroska ? 0 : Math.max(0, startTimestamp - EXTRA_THRESHOLD_IN_SECONDS));
  if (isMatroska) {
    warnAboutMatroskaOnce(src, logLevel);
  }
  let fullDuration = null;
  const cache = makeAudioCache();
  let lastUsed = Date.now();
  const getNextSample = async () => {
    lastUsed = Date.now();
    const { value: sample, done } = await sampleIterator.next();
    if (done) {
      fullDuration = cache.getNewestTimestamp();
      return null;
    }
    const realTimestamp = actualMatroskaTimestamps.getRealTimestamp(sample.timestamp);
    if (realTimestamp !== null && realTimestamp !== sample.timestamp) {
      sample.setTimestamp(realTimestamp);
    }
    actualMatroskaTimestamps.observeTimestamp(sample.timestamp);
    actualMatroskaTimestamps.observeTimestamp(sample.timestamp + sample.duration);
    cache.addFrame(sample);
    return sample;
  };
  const getSamples = async (timestamp, durationInSeconds) => {
    lastUsed = Date.now();
    if (fullDuration !== null && timestamp > fullDuration) {
      cache.clearBeforeThreshold(fullDuration - safetyOutOfOrderThreshold);
      return [];
    }
    const samples = cache.getSamples(timestamp, durationInSeconds);
    const newestTimestamp = cache.getNewestTimestamp();
    if (newestTimestamp !== null) {
      if (newestTimestamp >= timestamp + durationInSeconds - 0.0000000001) {
        return samples;
      }
    }
    while (true) {
      const sample = await getNextSample();
      const deleteBefore = fullDuration === null ? timestamp : Math.min(timestamp, fullDuration);
      cache.clearBeforeThreshold(deleteBefore - safetyOutOfOrderThreshold);
      if (sample === null) {
        break;
      }
      if (sample.timestamp + sample.duration - 0.0000000001 <= timestamp) {
        continue;
      }
      if (sample.timestamp >= timestamp + durationInSeconds - 0.0000000001) {
        break;
      }
      samples.push(sample);
    }
    return samples;
  };
  const logOpenFrames = () => {
    const openTimestamps = cache.getOpenTimestamps();
    if (openTimestamps.length > 0) {
      const first = openTimestamps[0];
      const last = openTimestamps[openTimestamps.length - 1];
      Internals8.Log.verbose({ logLevel, tag: "@remotion/media" }, "Open audio samples for src", src, `${first.toFixed(3)}...${last.toFixed(3)}`);
    }
  };
  const getCacheStats = () => {
    return {
      count: cache.getOpenTimestamps().length,
      size: cache.getOpenTimestamps().reduce((acc, t) => acc + t, 0)
    };
  };
  const canSatisfyRequestedTime = (timestamp) => {
    const oldestTimestamp = cache.getOldestTimestamp() ?? startTimestamp;
    if (fullDuration !== null && timestamp > fullDuration) {
      return true;
    }
    return oldestTimestamp < timestamp && Math.abs(oldestTimestamp - timestamp) < 10;
  };
  const prepareForDeletion = () => {
    cache.deleteAll();
    sampleIterator.return().then((value) => {
      if (value.value) {
        value.value.close();
      }
    });
    fullDuration = null;
  };
  let op = Promise.resolve([]);
  return {
    src,
    getSamples: (ts, dur) => {
      op = op.then(() => getSamples(ts, dur));
      return op;
    },
    waitForCompletion: async () => {
      await op;
      return true;
    },
    canSatisfyRequestedTime,
    logOpenFrames,
    getCacheStats,
    getLastUsed: () => lastUsed,
    prepareForDeletion,
    startTimestamp,
    clearBeforeThreshold: cache.clearBeforeThreshold,
    getOldestTimestamp: cache.getOldestTimestamp,
    getNewestTimestamp: cache.getNewestTimestamp
  };
};

// src/audio-extraction/audio-manager.ts
var makeAudioManager = () => {
  const iterators = [];
  const makeIterator = ({
    timeInSeconds,
    src,
    audioSampleSink,
    isMatroska,
    actualMatroskaTimestamps,
    logLevel
  }) => {
    const iterator = makeAudioIterator2({
      audioSampleSink,
      isMatroska,
      startTimestamp: timeInSeconds,
      src,
      actualMatroskaTimestamps,
      logLevel
    });
    iterators.push(iterator);
    return iterator;
  };
  const getIteratorMostInThePast = () => {
    let mostInThePast = null;
    let mostInThePastIterator = null;
    for (const iterator of iterators) {
      const lastUsed = iterator.getLastUsed();
      if (mostInThePast === null || lastUsed < mostInThePast) {
        mostInThePast = lastUsed;
        mostInThePastIterator = iterator;
      }
    }
    return mostInThePastIterator;
  };
  const deleteOldestIterator = () => {
    const iterator = getIteratorMostInThePast();
    if (iterator) {
      iterator.prepareForDeletion();
      iterators.splice(iterators.indexOf(iterator), 1);
    }
  };
  const deleteDuplicateIterators = (logLevel) => {
    const seenKeys = new Set;
    for (let i = 0;i < iterators.length; i++) {
      const iterator = iterators[i];
      const key = `${iterator.src}-${iterator.getOldestTimestamp()}-${iterator.getNewestTimestamp()}`;
      if (seenKeys.has(key)) {
        iterator.prepareForDeletion();
        iterators.splice(iterators.indexOf(iterator), 1);
        Internals9.Log.verbose({ logLevel, tag: "@remotion/media" }, `Deleted duplicate iterator for ${iterator.src}`);
      }
      seenKeys.add(key);
    }
  };
  const getIterator = async ({
    src,
    timeInSeconds,
    audioSampleSink,
    isMatroska,
    actualMatroskaTimestamps,
    logLevel,
    maxCacheSize
  }) => {
    let attempts = 0;
    const maxAttempts = 3;
    while ((await getTotalCacheStats()).totalSize > maxCacheSize && attempts < maxAttempts) {
      deleteOldestIterator();
      attempts++;
    }
    if ((await getTotalCacheStats()).totalSize > maxCacheSize && attempts >= maxAttempts) {
      Internals9.Log.warn({ logLevel, tag: "@remotion/media" }, `Audio cache: Exceeded max cache size after ${maxAttempts} attempts. Still ${(await getTotalCacheStats()).totalSize} bytes used, target was ${maxCacheSize} bytes.`);
    }
    for (const iterator of iterators) {
      if (iterator.src === src && await iterator.waitForCompletion() && iterator.canSatisfyRequestedTime(timeInSeconds)) {
        return iterator;
      }
    }
    for (let i = 0;i < iterators.length; i++) {
      const iterator = iterators[i];
      if (iterator.src === src && iterator.startTimestamp === timeInSeconds) {
        iterator.prepareForDeletion();
        iterators.splice(iterators.indexOf(iterator), 1);
      }
    }
    deleteDuplicateIterators(logLevel);
    return makeIterator({
      src,
      timeInSeconds,
      audioSampleSink,
      isMatroska,
      actualMatroskaTimestamps,
      logLevel
    });
  };
  const getCacheStats = () => {
    let totalCount = 0;
    let totalSize = 0;
    for (const iterator of iterators) {
      const { count, size } = iterator.getCacheStats();
      totalCount += count;
      totalSize += size;
    }
    return { count: totalCount, totalSize };
  };
  const logOpenFrames = () => {
    for (const iterator of iterators) {
      iterator.logOpenFrames();
    }
  };
  let queue = Promise.resolve(undefined);
  return {
    getIterator: ({
      src,
      timeInSeconds,
      audioSampleSink,
      isMatroska,
      actualMatroskaTimestamps,
      logLevel,
      maxCacheSize
    }) => {
      queue = queue.then(() => getIterator({
        src,
        timeInSeconds,
        audioSampleSink,
        isMatroska,
        actualMatroskaTimestamps,
        logLevel,
        maxCacheSize
      }));
      return queue;
    },
    getCacheStats,
    getIteratorMostInThePast,
    logOpenFrames,
    deleteDuplicateIterators
  };
};

// src/video-extraction/keyframe-manager.ts
import { Internals as Internals11 } from "remotion";

// src/render-timestamp-range.ts
var renderTimestampRange = (timestamps) => {
  if (timestamps.length === 0) {
    return "(none)";
  }
  if (timestamps.length === 1) {
    return timestamps[0].toFixed(3);
  }
  return `${timestamps[0].toFixed(3)}...${timestamps[timestamps.length - 1].toFixed(3)}`;
};

// src/video-extraction/keyframe-bank.ts
import { Internals as Internals10 } from "remotion";

// src/video-extraction/get-allocation-size.ts
var getAllocationSize = (sample) => {
  if (sample.format === null) {
    return sample.codedHeight * sample.codedWidth * 4;
  }
  return sample.allocationSize();
};

// src/video-extraction/keyframe-bank.ts
var BIGGEST_ALLOWED_JUMP_FORWARD_SECONDS = 3;
var makeKeyframeBank = async ({
  logLevel: parentLogLevel,
  src,
  videoSampleSink,
  initialTimestampRequest
}) => {
  const sampleIterator = videoSampleSink.samples(roundTo4Digits(initialTimestampRequest));
  const frames = {};
  const frameTimestamps = [];
  let hasReachedEndOfVideo = false;
  let lastUsed = Date.now();
  let allocationSize = 0;
  const getMeasuredDurationOfFrame = (timestamp) => {
    const index = frameTimestamps.indexOf(timestamp);
    if (index === -1) {
      throw new Error(`Frame ${timestamp} not found`);
    }
    const nextTimestamp = frameTimestamps[index + 1];
    if (nextTimestamp === undefined) {
      return null;
    }
    return nextTimestamp - timestamp;
  };
  const getKnownDurationOfFrame = (timestamp) => {
    const measuredDuration = getMeasuredDurationOfFrame(timestamp);
    if (measuredDuration !== null) {
      return measuredDuration;
    }
    if (!hasReachedEndOfVideo) {
      return null;
    }
    return frames[timestamp].duration ?? 0;
  };
  const getEstimatedDurationOfFrame = (timestamp) => {
    return getMeasuredDurationOfFrame(timestamp) ?? frames[timestamp].duration ?? 0;
  };
  const deleteFrameAtTimestamp = (timestamp) => {
    allocationSize -= getAllocationSize(frames[timestamp]);
    frameTimestamps.splice(frameTimestamps.indexOf(timestamp), 1);
    frames[timestamp].close();
    delete frames[timestamp];
  };
  const deleteFramesBeforeTimestamp = ({
    logLevel,
    timestampInSeconds
  }) => {
    const deletedTimestamps = [];
    for (const frameTimestamp of frameTimestamps.slice()) {
      if (hasReachedEndOfVideo) {
        const isLast = frameTimestamp === frameTimestamps[frameTimestamps.length - 1];
        if (isLast) {
          continue;
        }
      }
      if (!frames[frameTimestamp]) {
        continue;
      }
      const duration = getKnownDurationOfFrame(frameTimestamp);
      if (duration === null) {
        continue;
      }
      if (frameTimestamp + duration < timestampInSeconds) {
        deleteFrameAtTimestamp(frameTimestamp);
        deletedTimestamps.push(frameTimestamp);
      }
    }
    if (deletedTimestamps.length > 0) {
      Internals10.Log.verbose({ logLevel, tag: "@remotion/media" }, `Deleted ${deletedTimestamps.length} frame${deletedTimestamps.length === 1 ? "" : "s"} ${renderTimestampRange(deletedTimestamps)} for src ${src} because it is lower than ${timestampInSeconds}. Remaining: ${renderTimestampRange(frameTimestamps)}`);
    }
  };
  const hasDecodedEnoughForTimestamp = (timestamp) => {
    const lastFrameTimestamp = frameTimestamps[frameTimestamps.length - 1];
    if (lastFrameTimestamp === undefined) {
      return false;
    }
    const lastFrame = frames[lastFrameTimestamp];
    if (!lastFrame) {
      return true;
    }
    if (roundTo4Digits(lastFrameTimestamp) >= roundTo4Digits(timestamp)) {
      return true;
    }
    const duration = getKnownDurationOfFrame(lastFrameTimestamp);
    if (duration === null) {
      return false;
    }
    return roundTo4Digits(lastFrameTimestamp + duration) > roundTo4Digits(timestamp);
  };
  const addFrame = (frame, logLevel) => {
    if (frames[frame.timestamp]) {
      deleteFrameAtTimestamp(frame.timestamp);
    }
    frames[frame.timestamp] = frame;
    frameTimestamps.push(frame.timestamp);
    allocationSize += getAllocationSize(frame);
    lastUsed = Date.now();
    Internals10.Log.trace({ logLevel, tag: "@remotion/media" }, `Added frame at ${frame.timestamp}sec to bank`);
  };
  const ensureEnoughFramesForTimestamp = async (timestampInSeconds, logLevel, fps) => {
    while (!hasDecodedEnoughForTimestamp(timestampInSeconds)) {
      const sample = await sampleIterator.next();
      if (sample.value) {
        addFrame(sample.value, logLevel);
      }
      if (sample.done) {
        hasReachedEndOfVideo = true;
        break;
      }
      deleteFramesBeforeTimestamp({
        logLevel: parentLogLevel,
        timestampInSeconds: timestampInSeconds - getSafeWindowOfMonotonicity(fps)
      });
    }
    lastUsed = Date.now();
  };
  const getFrameFromTimestamp = async (timestampInSeconds, fps) => {
    lastUsed = Date.now();
    let adjustedTimestamp = timestampInSeconds;
    if (hasReachedEndOfVideo && roundTo4Digits(adjustedTimestamp) > roundTo4Digits(frameTimestamps[frameTimestamps.length - 1])) {
      adjustedTimestamp = frameTimestamps[frameTimestamps.length - 1];
    }
    await ensureEnoughFramesForTimestamp(adjustedTimestamp, parentLogLevel, fps);
    for (let i = frameTimestamps.length - 1;i >= 0; i--) {
      const sample = frames[frameTimestamps[i]];
      if (!sample) {
        return null;
      }
      if (roundTo4Digits(sample.timestamp) <= roundTo4Digits(adjustedTimestamp) || Math.abs(sample.timestamp - adjustedTimestamp) <= 0.001) {
        return sample;
      }
    }
    return frames[frameTimestamps[0]] ?? null;
  };
  const hasTimestampInSecond = async (timestamp, fps) => {
    return await getFrameFromTimestamp(timestamp, fps) !== null;
  };
  const getOpenFrameCount = () => {
    return {
      size: allocationSize,
      timestamps: frameTimestamps
    };
  };
  const getLastUsed = () => {
    return lastUsed;
  };
  let queue = Promise.resolve(undefined);
  const firstFrame = await sampleIterator.next();
  if (!firstFrame.value) {
    throw new Error("No first frame found");
  }
  const startTimestampInSeconds = firstFrame.value.timestamp;
  Internals10.Log.verbose({ logLevel: parentLogLevel, tag: "@remotion/media" }, `Creating keyframe bank from ${startTimestampInSeconds}sec`);
  addFrame(firstFrame.value, parentLogLevel);
  const getRangeOfTimestamps = () => {
    if (frameTimestamps.length === 0) {
      return null;
    }
    const firstTimestamp = frameTimestamps[0];
    const lastTimestamp = frameTimestamps[frameTimestamps.length - 1];
    const lastFrameDuration = getEstimatedDurationOfFrame(lastTimestamp);
    return {
      firstTimestamp,
      lastTimestamp: lastTimestamp + lastFrameDuration
    };
  };
  const prepareForDeletion = (logLevel, reason) => {
    const range = getRangeOfTimestamps();
    if (range) {
      Internals10.Log.verbose({ logLevel, tag: "@remotion/media" }, `Preparing for deletion (${reason}) of keyframe bank from ${range?.firstTimestamp}sec to ${range?.lastTimestamp}sec`);
    }
    let framesDeleted = 0;
    for (const frameTimestamp of frameTimestamps.slice()) {
      if (!frames[frameTimestamp]) {
        continue;
      }
      deleteFrameAtTimestamp(frameTimestamp);
      framesDeleted++;
    }
    sampleIterator.return();
    frameTimestamps.length = 0;
    return { framesDeleted };
  };
  const canSatisfyTimestamp = (timestamp) => {
    if (frameTimestamps.length === 0) {
      return false;
    }
    const roundedTimestamp = roundTo4Digits(timestamp);
    const firstFrameTimestamp = roundTo4Digits(frameTimestamps[0]);
    const range = getRangeOfTimestamps();
    if (!range) {
      return false;
    }
    const lastFrameTimestamp = roundTo4Digits(frameTimestamps[frameTimestamps.length - 1]);
    const lastFrameEndTimestamp = roundTo4Digits(range.lastTimestamp);
    if (hasReachedEndOfVideo && roundedTimestamp > lastFrameTimestamp) {
      return true;
    }
    if (roundedTimestamp < firstFrameTimestamp) {
      const firstFrameIsInitialFrame = firstFrameTimestamp === startTimestampInSeconds;
      const firstFrameDoesSatisfy = firstFrameIsInitialFrame && roundedTimestamp >= initialTimestampRequest;
      return firstFrameDoesSatisfy;
    }
    if (roundedTimestamp - BIGGEST_ALLOWED_JUMP_FORWARD_SECONDS > lastFrameEndTimestamp) {
      return false;
    }
    return true;
  };
  const keyframeBank = {
    getFrameFromTimestamp: (timestamp, fps) => {
      queue = queue.then(() => getFrameFromTimestamp(timestamp, fps));
      return queue;
    },
    prepareForDeletion,
    hasTimestampInSecond: (timestamp, fps) => {
      queue = queue.then(() => hasTimestampInSecond(timestamp, fps));
      return queue;
    },
    addFrame,
    deleteFramesBeforeTimestamp,
    src,
    getOpenFrameCount,
    getLastUsed,
    canSatisfyTimestamp,
    getRangeOfTimestamps
  };
  return keyframeBank;
};

// src/video-extraction/keyframe-manager.ts
var makeKeyframeManager = () => {
  let sources = {};
  const addKeyframeBank = ({ src, bank }) => {
    sources[src] = sources[src] ?? [];
    sources[src].push(bank);
  };
  const logCacheStats = (logLevel) => {
    let count = 0;
    let totalSize = 0;
    for (const src in sources) {
      for (const bank of sources[src]) {
        const { size, timestamps } = bank.getOpenFrameCount();
        count += timestamps.length;
        totalSize += size;
        if (size === 0) {
          continue;
        }
        Internals11.Log.verbose({ logLevel, tag: "@remotion/media" }, `Open frames for src ${src}: ${renderTimestampRange(timestamps)}`);
      }
    }
    Internals11.Log.verbose({ logLevel, tag: "@remotion/media" }, `Video cache stats: ${count} open frames, ${totalSize} bytes`);
  };
  const getCacheStats = () => {
    let count = 0;
    let totalSize = 0;
    for (const src in sources) {
      for (const bank of sources[src]) {
        const { timestamps, size } = bank.getOpenFrameCount();
        count += timestamps.length;
        totalSize += size;
        if (size === 0) {
          continue;
        }
      }
    }
    return { count, totalSize };
  };
  const getTheKeyframeBankMostInThePast = () => {
    let mostInThePast = null;
    let mostInThePastBank = null;
    let numberOfBanks = 0;
    for (const src in sources) {
      for (const bank of sources[src]) {
        const index = sources[src].indexOf(bank);
        const lastUsed = bank.getLastUsed();
        if (mostInThePast === null || lastUsed < mostInThePast) {
          mostInThePast = lastUsed;
          mostInThePastBank = { src, bank, index };
        }
        numberOfBanks++;
      }
    }
    if (!mostInThePastBank) {
      throw new Error("No keyframe bank found");
    }
    return { mostInThePastBank, numberOfBanks };
  };
  const deleteOldestKeyframeBank = (logLevel) => {
    const {
      mostInThePastBank: {
        bank: mostInThePastBank,
        src: mostInThePastSrc,
        index: mostInThePastIndex
      },
      numberOfBanks
    } = getTheKeyframeBankMostInThePast();
    if (numberOfBanks < 2) {
      return { finish: true };
    }
    if (mostInThePastBank) {
      const range = mostInThePastBank.getRangeOfTimestamps();
      const { framesDeleted } = mostInThePastBank.prepareForDeletion(logLevel, "deleted oldest keyframe bank to stay under max cache size");
      sources[mostInThePastSrc].splice(mostInThePastIndex, 1);
      if (range) {
        Internals11.Log.verbose({ logLevel, tag: "@remotion/media" }, `Deleted ${framesDeleted} frames for src ${mostInThePastSrc} from ${range?.firstTimestamp}sec to ${range?.lastTimestamp}sec to free up memory.`);
      }
    }
    return { finish: false };
  };
  const ensureToStayUnderMaxCacheSize = (logLevel, maxCacheSize) => {
    let cacheStats = getTotalCacheStats();
    let attempts = 0;
    const maxAttempts = 3;
    while (cacheStats.totalSize > maxCacheSize && attempts < maxAttempts) {
      const { finish } = deleteOldestKeyframeBank(logLevel);
      if (finish) {
        break;
      }
      Internals11.Log.verbose({ logLevel, tag: "@remotion/media" }, "Deleted oldest keyframe bank to stay under max cache size", (cacheStats.totalSize / 1024 / 1024).toFixed(1), "out of", (maxCacheSize / 1024 / 1024).toFixed(1));
      cacheStats = getTotalCacheStats();
      attempts++;
    }
    if (cacheStats.totalSize > maxCacheSize && attempts >= maxAttempts) {
      Internals11.Log.warn({ logLevel, tag: "@remotion/media" }, `Exceeded max cache size after ${maxAttempts} attempts. Remaining cache size: ${(cacheStats.totalSize / 1024 / 1024).toFixed(1)} MB, target was ${(maxCacheSize / 1024 / 1024).toFixed(1)} MB.`);
    }
  };
  const clearKeyframeBanksBeforeTime = ({
    timestampInSeconds,
    src,
    logLevel,
    fps
  }) => {
    const threshold = timestampInSeconds - getSafeWindowOfMonotonicity(fps);
    if (!sources[src]) {
      return;
    }
    const banks = sources[src];
    for (const bank of banks) {
      const range = bank.getRangeOfTimestamps();
      if (!range) {
        continue;
      }
      if (range.lastTimestamp < threshold) {
        bank.prepareForDeletion(logLevel, "cleared before threshold " + threshold);
        Internals11.Log.verbose({ logLevel, tag: "@remotion/media" }, `[Video] Cleared frames for src ${src} from ${range.firstTimestamp}sec to ${range.lastTimestamp}sec`);
        const bankIndex = banks.indexOf(bank);
        delete sources[src][bankIndex];
      } else {
        bank.deleteFramesBeforeTimestamp({
          timestampInSeconds: threshold,
          logLevel
        });
      }
    }
    sources[src] = sources[src].filter((bank) => bank !== undefined);
    logCacheStats(logLevel);
  };
  const getKeyframeBankOrRefetch = async ({
    timestamp,
    videoSampleSink,
    src,
    logLevel
  }) => {
    const existingBanks = sources[src] ?? [];
    const existingBank = existingBanks?.find((bank) => bank.canSatisfyTimestamp(timestamp));
    if (!existingBank) {
      Internals11.Log.trace({ logLevel, tag: "@remotion/media" }, `Creating new keyframe bank for src ${src} at timestamp ${timestamp}`);
      const newKeyframeBank = await makeKeyframeBank({
        videoSampleSink,
        logLevel,
        src,
        initialTimestampRequest: timestamp
      });
      addKeyframeBank({ src, bank: newKeyframeBank });
      return newKeyframeBank;
    }
    if (existingBank.canSatisfyTimestamp(timestamp)) {
      Internals11.Log.trace({ logLevel, tag: "@remotion/media" }, `Keyframe bank exists and satisfies timestamp ${timestamp}`);
      return existingBank;
    }
    Internals11.Log.verbose({ logLevel, tag: "@remotion/media" }, `Keyframe bank exists but frame at time ${timestamp} does not exist anymore.`);
    existingBank.prepareForDeletion(logLevel, "already existed but evicted");
    sources[src] = sources[src].filter((bank) => bank !== existingBank);
    const replacementKeybank = await makeKeyframeBank({
      videoSampleSink,
      initialTimestampRequest: timestamp,
      logLevel,
      src
    });
    addKeyframeBank({ src, bank: replacementKeybank });
    return replacementKeybank;
  };
  const requestKeyframeBank = async ({
    timestamp,
    videoSampleSink,
    src,
    logLevel,
    maxCacheSize,
    fps
  }) => {
    ensureToStayUnderMaxCacheSize(logLevel, maxCacheSize);
    clearKeyframeBanksBeforeTime({
      timestampInSeconds: timestamp,
      src,
      logLevel,
      fps
    });
    const keyframeBank = await getKeyframeBankOrRefetch({
      timestamp,
      videoSampleSink,
      src,
      logLevel
    });
    return keyframeBank;
  };
  const clearAll = (logLevel) => {
    const srcs = Object.keys(sources);
    for (const src of srcs) {
      const banks = sources[src];
      for (const bank of banks) {
        bank.prepareForDeletion(logLevel, "clearAll");
      }
      sources[src] = [];
    }
    sources = {};
  };
  let queue = Promise.resolve(undefined);
  return {
    requestKeyframeBank: ({
      timestamp,
      videoSampleSink,
      src,
      logLevel,
      maxCacheSize,
      fps
    }) => {
      queue = queue.then(() => requestKeyframeBank({
        timestamp,
        videoSampleSink,
        src,
        logLevel,
        maxCacheSize,
        fps
      }));
      return queue;
    },
    getCacheStats,
    clearAll
  };
};

// src/caches.ts
var getSafeWindowOfMonotonicity = (fps) => 0.2 * 30 / fps;
var keyframeManager = makeKeyframeManager();
var audioManager = makeAudioManager();
var getTotalCacheStats = () => {
  const keyframeManagerCacheStats = keyframeManager.getCacheStats();
  const audioManagerCacheStats = audioManager.getCacheStats();
  return {
    count: keyframeManagerCacheStats.count + audioManagerCacheStats.count,
    totalSize: keyframeManagerCacheStats.totalSize + audioManagerCacheStats.totalSize
  };
};
var getUncachedMaxCacheSize = (logLevel) => {
  if (typeof window !== "undefined" && window.remotion_mediaCacheSizeInBytes !== undefined && window.remotion_mediaCacheSizeInBytes !== null) {
    if (window.remotion_mediaCacheSizeInBytes < 240 * 1024 * 1024) {
      cancelRender(new Error(`The minimum value for the "mediaCacheSizeInBytes" prop is 240MB (${240 * 1024 * 1024}), got: ${window.remotion_mediaCacheSizeInBytes}`));
    }
    if (window.remotion_mediaCacheSizeInBytes > 20000 * 1024 * 1024) {
      cancelRender(new Error(`The maximum value for the "mediaCacheSizeInBytes" prop is 20GB (${20000 * 1024 * 1024}), got: ${window.remotion_mediaCacheSizeInBytes}`));
    }
    Internals12.Log.verbose({ logLevel, tag: "@remotion/media" }, `Using cache size set using "mediaCacheSizeInBytes": ${(window.remotion_mediaCacheSizeInBytes / 1024 / 1024).toFixed(1)} MB`);
    return window.remotion_mediaCacheSizeInBytes;
  }
  if (typeof window !== "undefined" && window.remotion_initialMemoryAvailable !== undefined && window.remotion_initialMemoryAvailable !== null) {
    const value = window.remotion_initialMemoryAvailable / 2;
    if (value < 500 * 1024 * 1024) {
      Internals12.Log.verbose({ logLevel, tag: "@remotion/media" }, `Using cache size set based on minimum value of 500MB (which is more than half of the available system memory!)`);
      return 500 * 1024 * 1024;
    }
    if (value > 20000 * 1024 * 1024) {
      Internals12.Log.verbose({ logLevel, tag: "@remotion/media" }, `Using cache size set based on maximum value of 20GB (which is less than half of the available system memory)`);
      return 20000 * 1024 * 1024;
    }
    Internals12.Log.verbose({ logLevel, tag: "@remotion/media" }, `Using cache size set based on available memory (50% of available memory): ${(value / 1024 / 1024).toFixed(1)} MB`);
    return value;
  }
  return 1000 * 1000 * 1000;
};
var cachedMaxCacheSize = null;
var getMaxVideoCacheSize = (logLevel) => {
  if (cachedMaxCacheSize !== null) {
    return cachedMaxCacheSize;
  }
  cachedMaxCacheSize = getUncachedMaxCacheSize(logLevel);
  return cachedMaxCacheSize;
};
var useMaxMediaCacheSize = (logLevel) => {
  const context = React2.useContext(Internals12.MaxMediaCacheSizeContext);
  if (context === null) {
    return getMaxVideoCacheSize(logLevel);
  }
  return context;
};

// src/convert-audiodata/apply-volume.ts
var applyVolume = (array, volume) => {
  if (volume === 1) {
    return;
  }
  for (let i = 0;i < array.length; i++) {
    const newValue = array[i] * volume;
    if (newValue < -32768) {
      array[i] = -32768;
    } else if (newValue > 32767) {
      array[i] = 32767;
    } else {
      array[i] = newValue;
    }
  }
};

// src/convert-audiodata/resample-audiodata.ts
var TARGET_NUMBER_OF_CHANNELS = 2;
var getTargetSampleRate = () => {
  if (typeof window !== "undefined" && window.remotion_sampleRate) {
    return window.remotion_sampleRate;
  }
  return 48000;
};
var fixFloatingPoint = (value) => {
  if (value % 1 < 0.0000001) {
    return Math.floor(value);
  }
  if (value % 1 > 0.9999999) {
    return Math.ceil(value);
  }
  return value;
};
var resampleAudioData = ({
  srcNumberOfChannels,
  sourceChannels,
  destination,
  targetFrames,
  chunkSize
}) => {
  const getSourceValues = (startUnfixed, endUnfixed, channelIndex) => {
    const start = fixFloatingPoint(startUnfixed);
    const end = fixFloatingPoint(endUnfixed);
    const startFloor = Math.floor(start);
    const startCeil = Math.ceil(start);
    const startFraction = start - startFloor;
    const endFraction = end - Math.floor(end);
    const endFloor = Math.floor(end);
    let weightedSum = 0;
    let totalWeight = 0;
    if (startFraction > 0) {
      const firstSample = sourceChannels[startFloor * srcNumberOfChannels + channelIndex];
      weightedSum += firstSample * (1 - startFraction);
      totalWeight += 1 - startFraction;
    }
    for (let k = startCeil;k < endFloor; k++) {
      const num = sourceChannels[k * srcNumberOfChannels + channelIndex];
      weightedSum += num;
      totalWeight += 1;
    }
    if (endFraction > 0) {
      const lastSample = sourceChannels[endFloor * srcNumberOfChannels + channelIndex];
      weightedSum += lastSample * endFraction;
      totalWeight += endFraction;
    }
    const average = weightedSum / totalWeight;
    return average;
  };
  for (let newFrameIndex = 0;newFrameIndex < targetFrames; newFrameIndex++) {
    const start = newFrameIndex * chunkSize;
    const end = start + chunkSize;
    if (TARGET_NUMBER_OF_CHANNELS === srcNumberOfChannels) {
      for (let i = 0;i < srcNumberOfChannels; i++) {
        destination[newFrameIndex * srcNumberOfChannels + i] = getSourceValues(start, end, i);
      }
    }
    if (srcNumberOfChannels === 1) {
      const m = getSourceValues(start, end, 0);
      destination[newFrameIndex * 2 + 0] = m;
      destination[newFrameIndex * 2 + 1] = m;
    } else if (srcNumberOfChannels === 4) {
      const l = getSourceValues(start, end, 0);
      const r = getSourceValues(start, end, 1);
      const sl = getSourceValues(start, end, 2);
      const sr = getSourceValues(start, end, 3);
      const l2 = 0.5 * (l + sl);
      const r2 = 0.5 * (r + sr);
      destination[newFrameIndex * 2 + 0] = l2;
      destination[newFrameIndex * 2 + 1] = r2;
    } else if (srcNumberOfChannels === 6) {
      const l = getSourceValues(start, end, 0);
      const r = getSourceValues(start, end, 1);
      const c = getSourceValues(start, end, 2);
      const sl = getSourceValues(start, end, 3);
      const sr = getSourceValues(start, end, 4);
      const sq = Math.sqrt(1 / 2);
      const l2 = l + sq * (c + sl);
      const r2 = r + sq * (c + sr);
      destination[newFrameIndex * 2 + 0] = l2;
      destination[newFrameIndex * 2 + 1] = r2;
    } else {
      for (let i = 0;i < srcNumberOfChannels; i++) {
        destination[newFrameIndex * TARGET_NUMBER_OF_CHANNELS + i] = getSourceValues(start, end, i);
      }
    }
  }
};

// src/looped-frame.ts
var frameForVolumeProp = ({
  behavior,
  loop,
  assetDurationInSeconds,
  fps,
  frame,
  startsAt
}) => {
  if (!loop) {
    return frame + startsAt;
  }
  if (behavior === "extend") {
    return frame + startsAt;
  }
  const assetDurationInFrames = Math.floor(assetDurationInSeconds * fps) - startsAt;
  return frame % assetDurationInFrames + startsAt;
};

// src/convert-audiodata/convert-audiodata.ts
var FORMAT = "s16";
var fixFloatingPoint2 = (value) => {
  const decimal = Math.abs(value % 1);
  if (decimal < 0.0000001) {
    return value < 0 ? Math.ceil(value) : Math.floor(value);
  }
  if (decimal > 0.9999999) {
    return value < 0 ? Math.floor(value) : Math.ceil(value);
  }
  return value;
};
var ceilButNotIfFloatingPointIssue = (value) => {
  const fixed = fixFloatingPoint2(value);
  return Math.ceil(fixed);
};
var convertAudioData = ({
  audioData,
  trimStartInSeconds,
  trimEndInSeconds,
  playbackRate,
  audioDataTimestamp,
  isLast
}) => {
  const {
    numberOfChannels: srcNumberOfChannels,
    sampleRate: currentSampleRate,
    numberOfFrames
  } = audioData;
  const ratio = currentSampleRate / getTargetSampleRate();
  const frameOffset = Math.floor(fixFloatingPoint2(trimStartInSeconds * audioData.sampleRate));
  const unroundedFrameCount = numberOfFrames - trimEndInSeconds * audioData.sampleRate - frameOffset;
  const frameCount = isLast ? ceilButNotIfFloatingPointIssue(unroundedFrameCount) : Math.round(unroundedFrameCount);
  const newNumberOfFrames = isLast ? ceilButNotIfFloatingPointIssue(unroundedFrameCount / ratio / playbackRate) : Math.round(unroundedFrameCount / ratio / playbackRate);
  if (newNumberOfFrames === 0) {
    throw new Error("Cannot resample - the given sample rate would result in less than 1 sample");
  }
  const srcChannels = new Int16Array(srcNumberOfChannels * frameCount);
  const isF32 = audioData.format === "f32" || audioData.format === "f32-planar";
  if (isF32) {
    const bytesPerPlane = frameCount * 4;
    const f32Buffer = new ArrayBuffer(srcNumberOfChannels * bytesPerPlane);
    for (let ch = 0;ch < srcNumberOfChannels; ch++) {
      audioData.copyTo(new Float32Array(f32Buffer, ch * bytesPerPlane, frameCount), { planeIndex: ch, frameOffset, frameCount, format: "f32-planar" });
    }
    const f32AudioData = new AudioData({
      format: "f32-planar",
      sampleRate: currentSampleRate,
      numberOfFrames: frameCount,
      numberOfChannels: srcNumberOfChannels,
      timestamp: audioData.timestamp,
      data: f32Buffer
    });
    f32AudioData.copyTo(srcChannels, {
      planeIndex: 0,
      format: FORMAT,
      frameOffset: 0,
      frameCount
    });
    f32AudioData.close();
  } else {
    audioData.copyTo(srcChannels, {
      planeIndex: 0,
      format: FORMAT,
      frameOffset,
      frameCount
    });
  }
  const data = new Int16Array(newNumberOfFrames * TARGET_NUMBER_OF_CHANNELS);
  const chunkSize = frameCount / newNumberOfFrames;
  const timestampOffsetMicroseconds = frameOffset / audioData.sampleRate * 1e6;
  if (newNumberOfFrames === frameCount && TARGET_NUMBER_OF_CHANNELS === srcNumberOfChannels && playbackRate === 1) {
    return {
      data: srcChannels,
      numberOfFrames: newNumberOfFrames,
      timestamp: audioDataTimestamp * 1e6 + fixFloatingPoint2(timestampOffsetMicroseconds),
      durationInMicroSeconds: fixFloatingPoint2(newNumberOfFrames / getTargetSampleRate() * 1e6)
    };
  }
  resampleAudioData({
    srcNumberOfChannels,
    sourceChannels: srcChannels,
    destination: data,
    targetFrames: newNumberOfFrames,
    chunkSize
  });
  const newAudioData = {
    data,
    numberOfFrames: newNumberOfFrames,
    timestamp: audioDataTimestamp * 1e6 + fixFloatingPoint2(timestampOffsetMicroseconds),
    durationInMicroSeconds: fixFloatingPoint2(newNumberOfFrames / getTargetSampleRate() * 1e6)
  };
  return newAudioData;
};

// src/convert-audiodata/combine-audiodata.ts
var combineAudioDataAndClosePrevious = (audioDataArray) => {
  let numberOfFrames = 0;
  let durationInMicroSeconds = 0;
  const { timestamp } = audioDataArray[0];
  for (const audioData of audioDataArray) {
    numberOfFrames += audioData.numberOfFrames;
    durationInMicroSeconds += audioData.durationInMicroSeconds;
  }
  const arr = new Int16Array(numberOfFrames * TARGET_NUMBER_OF_CHANNELS);
  let offset = 0;
  for (const audioData of audioDataArray) {
    arr.set(audioData.data, offset);
    offset += audioData.data.length;
  }
  return {
    data: arr,
    numberOfFrames,
    timestamp: fixFloatingPoint2(timestamp),
    durationInMicroSeconds: fixFloatingPoint2(durationInMicroSeconds)
  };
};

// src/get-sink.ts
import { Internals as Internals14 } from "remotion";

// src/video-extraction/get-frames-since-keyframe.ts
import {
  ALL_FORMATS as ALL_FORMATS2,
  AudioSampleSink,
  EncodedPacketSink,
  Input as Input2,
  MATROSKA,
  UrlSource as UrlSource2,
  VideoSampleSink,
  WEBM
} from "mediabunny";
import { Internals as Internals13 } from "remotion";

// src/browser-can-use-webgl2.ts
var browserCanUseWebGl2 = null;
var browserCanUseWebGl2Uncached = () => {
  const canvas = new OffscreenCanvas(1, 1);
  const context = canvas.getContext("webgl2");
  return context !== null;
};
var canBrowserUseWebGl2 = () => {
  if (browserCanUseWebGl2 !== null) {
    return browserCanUseWebGl2;
  }
  browserCanUseWebGl2 = browserCanUseWebGl2Uncached();
  return browserCanUseWebGl2;
};

// src/video-extraction/remember-actual-matroska-timestamps.ts
var rememberActualMatroskaTimestamps = (isMatroska) => {
  const observations = [];
  const observeTimestamp = (startTime) => {
    if (!isMatroska) {
      return;
    }
    observations.push(startTime);
  };
  const getRealTimestamp = (observedTimestamp) => {
    if (!isMatroska) {
      return observedTimestamp;
    }
    return observations.find((observation) => Math.abs(observedTimestamp - observation) < 0.001) ?? null;
  };
  return {
    observeTimestamp,
    getRealTimestamp
  };
};

// src/video-extraction/get-frames-since-keyframe.ts
var getRetryDelay = () => {
  return null;
};
var getFormatOrNullOrNetworkError = async (input) => {
  try {
    return await input.getFormat();
  } catch (err) {
    if (isNetworkError(err)) {
      return "network-error";
    }
    return null;
  }
};
var getSinks = async (src, logLevel, credentials, requestInit) => {
  const resolvedRequestInit = resolveRequestInit({ credentials, requestInit });
  const input = new Input2({
    formats: ALL_FORMATS2,
    source: new UrlSource2(src, {
      getRetryDelay,
      ...resolvedRequestInit ? { requestInit: resolvedRequestInit } : undefined
    })
  });
  const format = await getFormatOrNullOrNetworkError(input);
  const isMatroska = format === MATROSKA || format === WEBM;
  const getVideoSinks = async () => {
    if (format === "network-error") {
      return "network-error";
    }
    if (format === null) {
      return "unknown-container-format";
    }
    const videoTrack = await input.getPrimaryVideoTrack();
    if (!videoTrack) {
      return "no-video-track";
    }
    if (await videoTrack.isLive()) {
      throw new Error("Live streams are not currently supported by Remotion. Sorry! Source: " + src);
    }
    if (await videoTrack.isRelativeToUnixEpoch()) {
      throw new Error("Streams with UNIX timestamps are not currently supported by Remotion. Sorry! Source: " + src);
    }
    const canDecode = await videoTrack.canDecode();
    if (!canDecode) {
      return "cannot-decode";
    }
    const sampleSink = new VideoSampleSink(videoTrack);
    const packetSink = new EncodedPacketSink(videoTrack);
    const startPacket = await packetSink.getFirstPacket({
      verifyKeyPackets: true
    });
    const hasAlpha = startPacket?.sideData.alpha;
    if (hasAlpha && !canBrowserUseWebGl2()) {
      Internals13.Log.warn({ logLevel, tag: "@remotion/media" }, `WebGL2 is not available, using the non-fast CPU path to decode alpha for ${src}.`);
    }
    return {
      sampleSink
    };
  };
  let videoSinksPromise = null;
  const getVideoSinksPromise = () => {
    if (videoSinksPromise) {
      return videoSinksPromise;
    }
    videoSinksPromise = getVideoSinks();
    return videoSinksPromise;
  };
  const audioSinksPromise = {};
  const getAudioSinks = async (index) => {
    if (format === null) {
      return "unknown-container-format";
    }
    if (format === "network-error") {
      return "network-error";
    }
    const [videoTrack, audioTracks] = await Promise.all([
      input.getPrimaryVideoTrack(),
      input.getAudioTracks()
    ]);
    const audioTrack = await resolveAudioTrack({
      videoTrack,
      audioTracks,
      audioStreamIndex: index
    });
    if (!audioTrack) {
      return "no-audio-track";
    }
    const canDecode = await audioTrack.canDecode();
    if (!canDecode) {
      return "cannot-decode-audio";
    }
    return {
      sampleSink: new AudioSampleSink(audioTrack)
    };
  };
  const getAudioSinksPromise = (index) => {
    const keyIndex = index === null ? -1 : index;
    if (audioSinksPromise[keyIndex]) {
      return audioSinksPromise[keyIndex];
    }
    audioSinksPromise[keyIndex] = getAudioSinks(index);
    return audioSinksPromise[keyIndex];
  };
  return {
    getVideo: () => getVideoSinksPromise(),
    getAudio: (index) => getAudioSinksPromise(index),
    actualMatroskaTimestamps: rememberActualMatroskaTimestamps(isMatroska),
    isMatroska,
    getDuration: () => {
      return getDurationOrCompute(input);
    }
  };
};

// src/get-sink.ts
var sinkPromises = {};
var getSinkCacheKey = ({
  src,
  credentials,
  requestInit
}) => JSON.stringify([
  src,
  credentials,
  getMediaRequestInitFingerprint(requestInit)
]);
var getSink = (src, logLevel, credentials, requestInit) => {
  const normalizedRequestInit = normalizeMediaRequestInit(requestInit);
  const cacheKey = getSinkCacheKey({
    src,
    credentials,
    requestInit: normalizedRequestInit
  });
  let promise = sinkPromises[cacheKey];
  if (!promise) {
    Internals14.Log.verbose({
      logLevel,
      tag: "@remotion/media"
    }, `Sink for ${src} was not found, creating new sink`);
    promise = getSinks(src, logLevel, credentials, normalizedRequestInit);
    sinkPromises[cacheKey] = promise;
  }
  return promise;
};

// src/audio-extraction/extract-audio.ts
var extractAudioInternal = async ({
  src,
  timeInSeconds: unloopedTimeInSeconds,
  durationInSeconds: durationNotYetApplyingPlaybackRate,
  logLevel,
  loop,
  playbackRate,
  audioStreamIndex,
  trimBefore,
  trimAfter,
  fps,
  maxCacheSize,
  credentials,
  requestInit
}) => {
  const { getAudio, actualMatroskaTimestamps, isMatroska, getDuration } = await getSink(src, logLevel, credentials, requestInit);
  let mediaDurationInSeconds = null;
  if (loop) {
    mediaDurationInSeconds = await getDuration();
  }
  const audio = await getAudio(audioStreamIndex);
  if (audio === "network-error") {
    return "network-error";
  }
  if (audio === "no-audio-track") {
    return { data: null, durationInSeconds: null };
  }
  if (audio === "cannot-decode-audio") {
    return "cannot-decode";
  }
  if (audio === "unknown-container-format") {
    return "unknown-container-format";
  }
  const timeInSeconds = getTimeInSeconds({
    loop,
    mediaDurationInSeconds,
    unloopedTimeInSeconds,
    src,
    trimAfter,
    playbackRate,
    trimBefore,
    fps,
    ifNoMediaDuration: "fail"
  });
  if (timeInSeconds === null) {
    return { data: null, durationInSeconds: mediaDurationInSeconds };
  }
  try {
    const sampleIterator = await audioManager.getIterator({
      src,
      timeInSeconds,
      audioSampleSink: audio.sampleSink,
      isMatroska,
      actualMatroskaTimestamps,
      logLevel,
      maxCacheSize
    });
    const durationInSeconds = durationNotYetApplyingPlaybackRate * playbackRate;
    const samples = await sampleIterator.getSamples(timeInSeconds, durationInSeconds);
    audioManager.logOpenFrames();
    const audioDataArray = [];
    for (let i = 0;i < samples.length; i++) {
      const sample = samples[i];
      if (Math.abs(sample.timestamp - (timeInSeconds + durationInSeconds)) * sample.sampleRate < 1) {
        continue;
      }
      if (sample.timestamp + sample.duration <= timeInSeconds) {
        continue;
      }
      const isFirstSample = i === 0;
      const isLastSample = i === samples.length - 1;
      const audioDataRaw = sample.toAudioData();
      let trimStartInSeconds = 0;
      let trimEndInSeconds = 0;
      let leadingSilence = null;
      if (isFirstSample) {
        trimStartInSeconds = fixFloatingPoint2(timeInSeconds - sample.timestamp);
        if (trimStartInSeconds < 0) {
          const silenceFrames = Math.ceil(fixFloatingPoint2(-trimStartInSeconds * getTargetSampleRate()));
          leadingSilence = {
            data: new Int16Array(silenceFrames * TARGET_NUMBER_OF_CHANNELS),
            numberOfFrames: silenceFrames,
            timestamp: timeInSeconds * 1e6,
            durationInMicroSeconds: silenceFrames / getTargetSampleRate() * 1e6
          };
          trimStartInSeconds = 0;
        }
      }
      if (isLastSample) {
        trimEndInSeconds = Math.max(0, sample.timestamp + sample.duration - (timeInSeconds + durationInSeconds));
      }
      const audioData = convertAudioData({
        audioData: audioDataRaw,
        trimStartInSeconds,
        trimEndInSeconds,
        playbackRate,
        audioDataTimestamp: sample.timestamp,
        isLast: isLastSample
      });
      audioDataRaw.close();
      if (audioData.numberOfFrames === 0) {
        continue;
      }
      if (leadingSilence) {
        audioDataArray.push(leadingSilence);
      }
      audioDataArray.push(audioData);
    }
    if (audioDataArray.length === 0) {
      return { data: null, durationInSeconds: mediaDurationInSeconds };
    }
    const combined = combineAudioDataAndClosePrevious(audioDataArray);
    return { data: combined, durationInSeconds: mediaDurationInSeconds };
  } catch (err) {
    const error = err;
    if (isNetworkError(error)) {
      return "network-error";
    }
    if (isUnsupportedConfigurationError(error)) {
      return "cannot-decode";
    }
    throw err;
  }
};
var queue = Promise.resolve(undefined);
var extractAudio = (params) => {
  queue = queue.then(() => extractAudioInternal(params));
  return queue;
};

// src/video-extraction/extract-frame.ts
import { Internals as Internals15 } from "remotion";
var extractFrameInternal = async ({
  src,
  timeInSeconds: unloopedTimeInSeconds,
  logLevel,
  loop,
  trimAfter,
  trimBefore,
  playbackRate,
  fps,
  maxCacheSize,
  credentials,
  requestInit
}) => {
  const sink = await getSink(src, logLevel, credentials, requestInit);
  const [video, mediaDurationInSecondsRaw] = await Promise.all([
    sink.getVideo(),
    loop ? sink.getDuration() : Promise.resolve(null)
  ]);
  const mediaDurationInSeconds = loop ? mediaDurationInSecondsRaw : null;
  if (video === "no-video-track") {
    throw new Error(`No video track found for ${src}`);
  }
  if (video === "cannot-decode") {
    return { type: "cannot-decode", durationInSeconds: mediaDurationInSeconds };
  }
  if (video === "unknown-container-format") {
    return { type: "unknown-container-format" };
  }
  if (video === "network-error") {
    return { type: "network-error" };
  }
  if (video === "cannot-decode-alpha") {
    return {
      type: "cannot-decode-alpha",
      durationInSeconds: mediaDurationInSeconds
    };
  }
  const timeInSeconds = getTimeInSeconds({
    loop,
    mediaDurationInSeconds,
    unloopedTimeInSeconds,
    src,
    trimAfter,
    playbackRate,
    trimBefore,
    fps,
    ifNoMediaDuration: "fail"
  });
  if (timeInSeconds === null) {
    return {
      type: "success",
      frame: null,
      rotation: 0,
      durationInSeconds: await sink.getDuration()
    };
  }
  try {
    const keyframeBank = await keyframeManager.requestKeyframeBank({
      videoSampleSink: video.sampleSink,
      timestamp: timeInSeconds,
      src,
      logLevel,
      maxCacheSize,
      fps
    });
    if (!keyframeBank) {
      return {
        type: "success",
        frame: null,
        rotation: 0,
        durationInSeconds: await sink.getDuration()
      };
    }
    const frame = await keyframeBank.getFrameFromTimestamp(timeInSeconds, fps);
    const rotation = frame?.rotation ?? 0;
    return {
      type: "success",
      frame: frame?.toVideoFrame() ?? null,
      rotation,
      durationInSeconds: await sink.getDuration()
    };
  } catch (err) {
    Internals15.Log.info({ logLevel, tag: "@remotion/media" }, `Error decoding ${src} at time ${timeInSeconds}: ${err}`, err);
    return { type: "cannot-decode", durationInSeconds: mediaDurationInSeconds };
  }
};
var queue2 = Promise.resolve(undefined);
var extractFrame = (params) => {
  queue2 = queue2.then(() => extractFrameInternal(params));
  return queue2;
};

// src/video-extraction/rotate-frame.ts
var rotateFrame = async ({
  frame,
  rotation
}) => {
  if (rotation === 0) {
    const directBitmap = await createImageBitmap(frame);
    frame.close();
    return directBitmap;
  }
  const width = rotation === 90 || rotation === 270 ? frame.displayHeight : frame.displayWidth;
  const height = rotation === 90 || rotation === 270 ? frame.displayWidth : frame.displayHeight;
  const canvas = new OffscreenCanvas(width, height);
  const ctx = canvas.getContext("2d");
  if (!ctx) {
    throw new Error("Could not get 2d context");
  }
  canvas.width = width;
  canvas.height = height;
  if (rotation === 90) {
    ctx.translate(width, 0);
  } else if (rotation === 180) {
    ctx.translate(width, height);
  } else if (rotation === 270) {
    ctx.translate(0, height);
  }
  ctx.rotate(rotation * (Math.PI / 180));
  ctx.drawImage(frame, 0, 0);
  const bitmap = await createImageBitmap(canvas);
  frame.close();
  return bitmap;
};

// src/extract-frame-and-audio.ts
var extractFrameAndAudio = async ({
  src,
  timeInSeconds,
  logLevel,
  durationInSeconds,
  playbackRate,
  includeAudio,
  includeVideo,
  loop,
  audioStreamIndex,
  trimAfter,
  trimBefore,
  fps,
  maxCacheSize,
  credentials,
  requestInit
}) => {
  try {
    const [video, audio] = await Promise.all([
      includeVideo ? extractFrame({
        src,
        timeInSeconds,
        logLevel,
        loop,
        trimAfter,
        playbackRate,
        trimBefore,
        fps,
        maxCacheSize,
        credentials,
        requestInit
      }) : null,
      includeAudio ? extractAudio({
        src,
        timeInSeconds,
        durationInSeconds,
        logLevel,
        loop,
        playbackRate,
        audioStreamIndex,
        trimAfter,
        fps,
        trimBefore,
        maxCacheSize,
        credentials,
        requestInit
      }) : null
    ]);
    if (video?.type === "cannot-decode") {
      return {
        type: "cannot-decode",
        durationInSeconds: video.durationInSeconds
      };
    }
    if (video?.type === "unknown-container-format") {
      return { type: "unknown-container-format" };
    }
    if (video?.type === "cannot-decode-alpha") {
      return {
        type: "cannot-decode-alpha",
        durationInSeconds: video.durationInSeconds
      };
    }
    if (video?.type === "network-error") {
      return { type: "network-error" };
    }
    if (audio === "unknown-container-format") {
      return { type: "unknown-container-format" };
    }
    if (audio === "network-error") {
      return { type: "network-error" };
    }
    if (audio === "cannot-decode") {
      return {
        type: "cannot-decode",
        durationInSeconds: video?.type === "success" ? video.durationInSeconds : null
      };
    }
    return {
      type: "success",
      frame: video?.frame ? await rotateFrame({
        frame: video.frame,
        rotation: video.rotation
      }) : null,
      audio: audio?.data ?? null,
      durationInSeconds: audio?.durationInSeconds ?? null
    };
  } catch (err) {
    const error = err;
    if (isNetworkError(error)) {
      return { type: "network-error" };
    }
    throw err;
  }
};

// src/video-extraction/add-broadcast-channel-listener.ts
var emitReadiness = (channel) => {
  channel.postMessage({
    type: "main-tab-ready"
  });
  setInterval(() => {
    channel.postMessage({
      type: "main-tab-ready"
    });
  }, 300);
};
var addBroadcastChannelListener = () => {
  if (!(typeof window !== "undefined" && window.remotion_broadcastChannel && window.remotion_isMainTab)) {
    return;
  }
  window.remotion_broadcastChannel.addEventListener("message", async (event) => {
    const data = event.data;
    if (data.type === "request") {
      try {
        const result = await extractFrameAndAudio({
          src: data.src,
          timeInSeconds: data.timeInSeconds,
          logLevel: data.logLevel,
          durationInSeconds: data.durationInSeconds,
          playbackRate: data.playbackRate,
          includeAudio: data.includeAudio,
          includeVideo: data.includeVideo,
          loop: data.loop,
          audioStreamIndex: data.audioStreamIndex,
          trimAfter: data.trimAfter,
          trimBefore: data.trimBefore,
          fps: data.fps,
          maxCacheSize: data.maxCacheSize,
          credentials: data.credentials,
          requestInit: data.requestInit
        });
        if (result.type === "cannot-decode") {
          const cannotDecodeResponse = {
            type: "response-cannot-decode",
            id: data.id,
            durationInSeconds: result.durationInSeconds
          };
          window.remotion_broadcastChannel.postMessage(cannotDecodeResponse);
          return;
        }
        if (result.type === "cannot-decode-alpha") {
          const cannotDecodeAlphaResponse = {
            type: "response-cannot-decode-alpha",
            id: data.id,
            durationInSeconds: result.durationInSeconds
          };
          window.remotion_broadcastChannel.postMessage(cannotDecodeAlphaResponse);
          return;
        }
        if (result.type === "network-error") {
          const networkErrorResponse = {
            type: "response-network-error",
            id: data.id
          };
          window.remotion_broadcastChannel.postMessage(networkErrorResponse);
          return;
        }
        if (result.type === "unknown-container-format") {
          const unknownContainerFormatResponse = {
            type: "response-unknown-container-format",
            id: data.id
          };
          window.remotion_broadcastChannel.postMessage(unknownContainerFormatResponse);
          return;
        }
        const { frame, audio, durationInSeconds } = result;
        const imageBitmap = frame ? await createImageBitmap(frame) : null;
        if (frame) {
          frame.close();
        }
        const response = {
          type: "response-success",
          id: data.id,
          frame: imageBitmap,
          audio,
          durationInSeconds: durationInSeconds ?? null
        };
        window.remotion_broadcastChannel.postMessage(response);
      } catch (error) {
        const response = {
          type: "response-error",
          id: data.id,
          errorStack: error.stack ?? "No stack trace"
        };
        window.remotion_broadcastChannel.postMessage(response);
      }
    } else if (data.type === "main-tab-ready") {} else {
      throw new Error("Invalid message: " + JSON.stringify(data));
    }
  });
  emitReadiness(window.remotion_broadcastChannel);
};
var mainTabIsReadyProm = null;
var waitForMainTabToBeReady = (channel) => {
  if (mainTabIsReadyProm) {
    return mainTabIsReadyProm;
  }
  mainTabIsReadyProm = new Promise((resolve) => {
    const onMessage = (event) => {
      const data = event.data;
      if (data.type === "main-tab-ready") {
        resolve();
        channel.removeEventListener("message", onMessage);
      }
    };
    channel.addEventListener("message", onMessage);
  });
  return mainTabIsReadyProm;
};

// src/video-extraction/extract-frame-via-broadcast-channel.ts
addBroadcastChannelListener();
var extractFrameViaBroadcastChannel = async ({
  src,
  timeInSeconds,
  logLevel,
  durationInSeconds,
  playbackRate,
  includeAudio,
  includeVideo,
  isClientSideRendering,
  loop,
  audioStreamIndex,
  trimAfter,
  trimBefore,
  fps,
  maxCacheSize,
  credentials,
  requestInit
}) => {
  if (isClientSideRendering || window.remotion_isMainTab) {
    return extractFrameAndAudio({
      logLevel,
      src,
      timeInSeconds,
      durationInSeconds,
      playbackRate,
      includeAudio,
      includeVideo,
      loop,
      audioStreamIndex,
      trimAfter,
      trimBefore,
      fps,
      maxCacheSize,
      credentials,
      requestInit
    });
  }
  await waitForMainTabToBeReady(window.remotion_broadcastChannel);
  const requestId = crypto.randomUUID();
  const resolvePromise = new Promise((resolve, reject) => {
    const onMessage = (event) => {
      const data = event.data;
      if (!data) {
        return;
      }
      if (data.type === "main-tab-ready") {
        return;
      }
      if (data.id !== requestId) {
        return;
      }
      if (data.type === "response-success") {
        resolve({
          type: "success",
          frame: data.frame ? data.frame : null,
          audio: data.audio ? data.audio : null,
          durationInSeconds: data.durationInSeconds ? data.durationInSeconds : null
        });
        window.remotion_broadcastChannel.removeEventListener("message", onMessage);
        return;
      }
      if (data.type === "response-error") {
        reject(data.errorStack);
        window.remotion_broadcastChannel.removeEventListener("message", onMessage);
        return;
      }
      if (data.type === "response-cannot-decode") {
        resolve({
          type: "cannot-decode",
          durationInSeconds: data.durationInSeconds
        });
        window.remotion_broadcastChannel.removeEventListener("message", onMessage);
        return;
      }
      if (data.type === "response-network-error") {
        resolve({ type: "network-error" });
        window.remotion_broadcastChannel.removeEventListener("message", onMessage);
        return;
      }
      if (data.type === "response-unknown-container-format") {
        resolve({ type: "unknown-container-format" });
        window.remotion_broadcastChannel.removeEventListener("message", onMessage);
        return;
      }
      if (data.type === "response-cannot-decode-alpha") {
        resolve({
          type: "cannot-decode-alpha",
          durationInSeconds: data.durationInSeconds
        });
        window.remotion_broadcastChannel.removeEventListener("message", onMessage);
        return;
      }
      throw new Error(`Invalid message: ${JSON.stringify(data)}`);
    };
    window.remotion_broadcastChannel.addEventListener("message", onMessage);
  });
  const request = {
    type: "request",
    src,
    timeInSeconds,
    id: requestId,
    logLevel,
    durationInSeconds,
    playbackRate,
    includeAudio,
    includeVideo,
    loop,
    audioStreamIndex,
    trimAfter,
    trimBefore,
    fps,
    maxCacheSize,
    credentials,
    requestInit: normalizeMediaRequestInit(requestInit)
  };
  window.remotion_broadcastChannel.postMessage(request);
  let timeoutId;
  return Promise.race([
    resolvePromise.then((res) => {
      clearTimeout(timeoutId);
      return res;
    }),
    new Promise((_, reject) => {
      timeoutId = setTimeout(() => {
        reject(new Error(`Timeout while extracting frame at time ${timeInSeconds}sec from ${src}`));
      }, Math.max(3000, window.remotion_puppeteerTimeout - 5000));
    })
  ]);
};

// src/audio/audio-for-rendering.tsx
import { jsx as jsx2 } from "react/jsx-runtime";
var AudioForRendering = ({
  volume: volumeProp,
  playbackRate,
  src,
  muted,
  loopVolumeCurveBehavior,
  delayRenderRetries,
  delayRenderTimeoutInMilliseconds,
  logLevel: overriddenLogLevel,
  loop,
  fallbackHtml5AudioProps,
  audioStreamIndex,
  showInTimeline,
  style,
  name,
  disallowFallbackToHtml5Audio,
  toneFrequency,
  trimAfter,
  trimBefore,
  onError,
  credentials,
  requestInit
}) => {
  const defaultLogLevel = Internals16.useLogLevel();
  const logLevel = overriddenLogLevel ?? defaultLogLevel;
  const frame = useCurrentFrame2();
  const absoluteFrame = Internals16.useTimelinePosition();
  const videoConfig = Internals16.useUnsafeVideoConfig();
  const { registerRenderAsset, unregisterRenderAsset } = useContext3(Internals16.RenderAssetManager);
  const startsAt = Internals16.useMediaStartsAt();
  const environment = useRemotionEnvironment();
  if (!videoConfig) {
    throw new Error("No video config found");
  }
  if (!src) {
    throw new TypeError("No `src` was passed to <Audio>.");
  }
  const { fps } = videoConfig;
  const { delayRender, continueRender } = useDelayRender();
  const [replaceWithHtml5Audio, setReplaceWithHtml5Audio] = useState2(false);
  const [initialRequestInit] = useState2(requestInit);
  const sequenceContext = useContext3(Internals16.SequenceContext);
  const id = useMemo2(() => `media-audio-${random(src)}-${sequenceContext?.cumulatedFrom}-${sequenceContext?.relativeFrom}-${sequenceContext?.durationInFrames}`, [
    src,
    sequenceContext?.cumulatedFrom,
    sequenceContext?.relativeFrom,
    sequenceContext?.durationInFrames
  ]);
  const maxCacheSize = useMaxMediaCacheSize(logLevel);
  const audioEnabled = Internals16.useAudioEnabled();
  useLayoutEffect2(() => {
    const timestamp = frame / fps;
    const durationInSeconds = 1 / fps;
    const shouldRenderAudio = (() => {
      if (!audioEnabled) {
        return false;
      }
      if (muted) {
        return false;
      }
      return true;
    })();
    if (!shouldRenderAudio) {
      return;
    }
    if (replaceWithHtml5Audio) {
      return;
    }
    const newHandle = delayRender(`Extracting audio for frame ${frame}`, {
      retries: delayRenderRetries ?? undefined,
      timeoutInMilliseconds: delayRenderTimeoutInMilliseconds ?? undefined
    });
    extractFrameViaBroadcastChannel({
      src,
      timeInSeconds: timestamp,
      durationInSeconds,
      playbackRate: playbackRate ?? 1,
      logLevel,
      includeAudio: shouldRenderAudio,
      includeVideo: false,
      isClientSideRendering: environment.isClientSideRendering,
      loop: loop ?? false,
      audioStreamIndex: audioStreamIndex ?? null,
      trimAfter,
      trimBefore,
      fps,
      maxCacheSize,
      credentials,
      requestInit: initialRequestInit
    }).then((result) => {
      const handleError = (error, clientSideError, fallbackMessage) => {
        const [action, errorToUse] = callOnErrorAndResolve({
          onError,
          error,
          disallowFallback: disallowFallbackToHtml5Audio ?? false,
          isClientSideRendering: environment.isClientSideRendering,
          clientSideError
        });
        if (action === "fail") {
          cancelRender2(errorToUse);
        }
        Internals16.Log.warn({ logLevel, tag: "@remotion/media" }, fallbackMessage);
        setReplaceWithHtml5Audio(true);
      };
      if (result.type === "unknown-container-format") {
        handleError(new Error(`Unknown container format ${src}.`), new Error(`Cannot render audio "${src}": Unknown container format. See supported formats: https://www.remotion.dev/docs/mediabunny/formats`), `Unknown container format for ${src} (Supported formats: https://www.remotion.dev/docs/mediabunny/formats), falling back to <Html5Audio>`);
        return;
      }
      if (result.type === "cannot-decode") {
        handleError(new Error(`Cannot decode ${src}.`), new Error(`Cannot render audio "${src}": The audio could not be decoded by the browser.`), `Cannot decode ${src}, falling back to <Html5Audio>`);
        return;
      }
      if (result.type === "cannot-decode-alpha") {
        throw new Error(`Cannot decode alpha component for ${src}, and 'disallowFallbackToHtml5Audio' was set. But this should never happen, since you used the <Audio> tag. Please report this as a bug.`);
      }
      if (result.type === "network-error") {
        handleError(new Error(`Network error fetching ${src}.`), new Error(`Cannot render audio "${src}": Network error while fetching the audio (possibly CORS).`), `Network error fetching ${src}, falling back to <Html5Audio>`);
        return;
      }
      const { audio, durationInSeconds: assetDurationInSeconds } = result;
      const volumePropsFrame = frameForVolumeProp({
        behavior: loopVolumeCurveBehavior ?? "repeat",
        loop: loop ?? false,
        assetDurationInSeconds: assetDurationInSeconds ?? 0,
        fps,
        frame,
        startsAt
      });
      const volume = Internals16.evaluateVolume({
        volume: volumeProp,
        frame: volumePropsFrame,
        mediaVolume: 1
      });
      Internals16.warnAboutTooHighVolume(volume);
      if (audio && volume > 0) {
        applyVolume(audio.data, volume);
        registerRenderAsset({
          type: "inline-audio",
          id,
          audio: environment.isClientSideRendering ? audio.data : Array.from(audio.data),
          frame: absoluteFrame,
          timestamp: audio.timestamp,
          duration: audio.numberOfFrames / getTargetSampleRate() * 1e6,
          toneFrequency: toneFrequency ?? 1
        });
      }
      continueRender(newHandle);
    }).catch((error) => {
      cancelRender2(error);
    });
    return () => {
      continueRender(newHandle);
      unregisterRenderAsset(id);
    };
  }, [
    absoluteFrame,
    continueRender,
    delayRender,
    delayRenderRetries,
    delayRenderTimeoutInMilliseconds,
    disallowFallbackToHtml5Audio,
    environment.isClientSideRendering,
    fps,
    frame,
    id,
    logLevel,
    loop,
    loopVolumeCurveBehavior,
    muted,
    playbackRate,
    registerRenderAsset,
    src,
    startsAt,
    unregisterRenderAsset,
    volumeProp,
    audioStreamIndex,
    toneFrequency,
    trimAfter,
    trimBefore,
    replaceWithHtml5Audio,
    maxCacheSize,
    audioEnabled,
    onError,
    credentials,
    initialRequestInit
  ]);
  if (replaceWithHtml5Audio) {
    return /* @__PURE__ */ jsx2(Html5Audio, {
      src,
      playbackRate,
      muted,
      loop,
      volume: volumeProp,
      delayRenderRetries,
      delayRenderTimeoutInMilliseconds,
      style,
      loopVolumeCurveBehavior,
      audioStreamIndex,
      useWebAudioApi: fallbackHtml5AudioProps?.useWebAudioApi,
      onError: fallbackHtml5AudioProps?.onError,
      toneFrequency,
      acceptableTimeShiftInSeconds: fallbackHtml5AudioProps?.acceptableTimeShiftInSeconds,
      preservePitch: fallbackHtml5AudioProps?.preservePitch ?? true,
      name,
      showInTimeline
    });
  }
  return null;
};

// src/audio/audio.tsx
import { jsx as jsx3 } from "react/jsx-runtime";
var { validateMediaProps } = Internals17;
var audioSchema = {
  ...Internals17.baseSchema,
  volume: {
    type: "number",
    min: 0,
    max: 20,
    step: 0.01,
    default: 1,
    description: "Volume",
    hiddenFromList: false
  },
  playbackRate: {
    type: "number",
    min: 0.1,
    step: 0.01,
    default: 1,
    description: "Playback rate",
    hiddenFromList: false,
    keyframable: false
  },
  loop: { type: "boolean", default: false, description: "Loop" }
};
var AudioInner = (props) => {
  const {
    name,
    stack,
    showInTimeline,
    controls,
    from,
    durationInFrames,
    freeze,
    hidden,
    ...otherProps
  } = props;
  const environment = useRemotionEnvironment2();
  const [mediaVolume] = Internals17.useMediaVolumeState();
  const mediaStartsAt = Internals17.useMediaStartsAt();
  const videoConfig = useVideoConfig2();
  const sequenceDurationInFrames = Math.min(durationInFrames ?? Infinity, Math.max(0, videoConfig.durationInFrames - (from ?? 0)));
  const basicInfo = Internals17.useBasicMediaInTimeline({
    src: props.src,
    volume: props.volume,
    playbackRate: props.playbackRate ?? 1,
    trimBefore: props.trimBefore,
    trimAfter: props.trimAfter,
    sequenceDurationInFrames,
    mediaType: "audio",
    displayName: name ?? "<Audio>",
    mediaVolume,
    mediaStartsAt,
    loop: props.loop ?? false
  });
  const [mediaDurationInSeconds, setMediaDurationInSeconds] = useState3(null);
  const loopDisplay = useMemo3(() => getLoopDisplay({
    loop: props.loop ?? false,
    mediaDurationInSeconds,
    playbackRate: props.playbackRate ?? 1,
    trimAfter: props.trimAfter,
    trimBefore: props.trimBefore,
    sequenceDurationInFrames,
    compFps: videoConfig.fps
  }), [
    props.loop,
    mediaDurationInSeconds,
    props.playbackRate,
    props.trimAfter,
    props.trimBefore,
    sequenceDurationInFrames,
    videoConfig.fps
  ]);
  const isMedia = useMemo3(() => ({
    type: "audio",
    data: basicInfo
  }), [basicInfo]);
  if (typeof props.src !== "string") {
    throw new TypeError(`The \`<Audio>\` tag requires a string for \`src\`, but got ${JSON.stringify(props.src)} instead.`);
  }
  validateMediaProps({ playbackRate: props.playbackRate, volume: props.volume }, "Audio");
  if (sequenceDurationInFrames === 0) {
    return null;
  }
  return /* @__PURE__ */ jsx3(Sequence, {
    layout: "none",
    from: from ?? 0,
    durationInFrames: basicInfo.duration,
    freeze,
    _remotionInternalStack: stack,
    _remotionInternalIsMedia: isMedia,
    name: name ?? "<Audio>",
    _remotionInternalDocumentationLink: name === undefined ? "https://www.remotion.dev/docs/media/audio" : undefined,
    controls,
    _remotionInternalLoopDisplay: loopDisplay,
    showInTimeline: showInTimeline ?? true,
    hidden,
    children: environment.isRendering ? /* @__PURE__ */ jsx3(AudioForRendering, {
      ...otherProps
    }) : /* @__PURE__ */ jsx3(AudioForPreview, {
      name,
      ...otherProps,
      stack: stack ?? null,
      setMediaDurationInSeconds
    })
  });
};
var Audio = Interactive.withSchema({
  Component: AudioInner,
  componentName: "<Audio>",
  componentIdentity: "dev.remotion.media.Audio",
  schema: audioSchema,
  supportsEffects: false
});
Internals17.addSequenceStackTraces(Audio);

// src/video/video.tsx
import React6, { useMemo as useMemo6, useState as useState6 } from "react";
import {
  Internals as Internals21,
  Interactive as Interactive2,
  Sequence as Sequence2,
  useRemotionEnvironment as useRemotionEnvironment4,
  useVideoConfig as useVideoConfig5
} from "remotion";

// src/video/video-for-preview.tsx
import {
  useCallback,
  useContext as useContext4,
  useEffect as useEffect2,
  useLayoutEffect as useLayoutEffect3,
  useMemo as useMemo4,
  useRef as useRef2,
  useState as useState4
} from "react";
import {
  Html5Video,
  Internals as Internals19,
  useBufferState as useBufferState2,
  useCurrentFrame as useCurrentFrame3,
  useVideoConfig as useVideoConfig3
} from "remotion";

// src/video/video-frame-cache.ts
var cache = new Map;
var cacheVideoFrame = (src, sourceCanvas) => {
  const { width, height } = sourceCanvas;
  if (width === 0 || height === 0) {
    return;
  }
  let cached = cache.get(src);
  if (!cached || cached.width !== width || cached.height !== height) {
    cached = new OffscreenCanvas(width, height);
    cache.set(src, cached);
  }
  const ctx = cached.getContext("2d");
  if (!ctx) {
    return;
  }
  ctx.clearRect(0, 0, width, height);
  ctx.drawImage(sourceCanvas, 0, 0);
};
var getCachedVideoFrame = (src) => {
  return cache.get(src) ?? null;
};

// src/video/warn-object-fit-css.ts
import { Internals as Internals18 } from "remotion";
var OBJECT_FIT_CLASS_PATTERN = /\bobject-(contain|cover|fill|none|scale-down)\b/;
var warnedStyle = false;
var warnedClassName = false;
var warnAboutObjectFitInStyleOrClassName = ({
  style,
  className,
  logLevel
}) => {
  if (!warnedStyle && style?.objectFit) {
    warnedStyle = true;
    Internals18.Log.warn({ logLevel, tag: "@remotion/media" }, "Use the `objectFit` prop instead of the `style` prop.");
  }
  if (!warnedClassName && className && OBJECT_FIT_CLASS_PATTERN.test(className)) {
    warnedClassName = true;
    Internals18.Log.warn({ logLevel, tag: "@remotion/media" }, "Use the `objectFit` prop instead of `object-*` CSS class names.");
  }
};

// src/video/video-for-preview.tsx
import { jsx as jsx4 } from "react/jsx-runtime";
var {
  useUnsafeVideoConfig: useUnsafeVideoConfig2,
  Timeline: Timeline2,
  SharedAudioContext: SharedAudioContext2,
  usePlayerMutedState: usePlayerMutedState2,
  useMediaVolumeState: useMediaVolumeState2,
  useFrameForVolumeProp: useFrameForVolumeProp2,
  evaluateVolume: evaluateVolume2,
  warnAboutTooHighVolume: warnAboutTooHighVolume2,
  usePreload: usePreload2,
  SequenceContext: SequenceContext2,
  useEffectChainState
} = Internals19;
var VideoForPreviewAssertedShowing = ({
  src: unpreloadedSrc,
  style,
  playbackRate,
  logLevel,
  className,
  muted,
  volume,
  loopVolumeCurveBehavior,
  onVideoFrame,
  showInTimeline,
  loop,
  trimAfter,
  trimBefore,
  stack,
  disallowFallbackToOffthreadVideo,
  fallbackOffthreadVideoProps,
  audioStreamIndex,
  debugOverlay,
  headless,
  onError,
  credentials,
  requestInit,
  objectFit: objectFitProp,
  _experimentalInitiallyDrawCachedFrame,
  effects,
  setMediaDurationInSeconds,
  refForOutline,
  ...props
}) => {
  const src = usePreload2(unpreloadedSrc);
  const canvasRef = useRef2(null);
  const videoConfig = useUnsafeVideoConfig2();
  const frame = useCurrentFrame3();
  const mediaPlayerRef = useRef2(null);
  const initialTrimBeforeRef = useRef2(trimBefore);
  const initialTrimAfterRef = useRef2(trimAfter);
  const initialOnVideoFrameRef = useRef2(onVideoFrame);
  const [initialRequestInit] = useState4(requestInit);
  const [mediaPlayerReady, setMediaPlayerReady] = useState4(false);
  const [shouldFallbackToNativeVideo, setShouldFallbackToNativeVideo] = useState4(false);
  const [playing] = Timeline2.usePlayingState();
  const { playbackRate: globalPlaybackRate } = Internals19.usePlaybackRate();
  const sharedAudioContext = useContext4(SharedAudioContext2);
  const buffer = useBufferState2();
  const canvasRefCallback = useCallback((canvas) => {
    canvasRef.current = canvas;
    refForOutline.current = canvas;
  }, [refForOutline]);
  const fallbackVideoRef = useCallback((video) => {
    refForOutline.current = video;
  }, [refForOutline]);
  const [playerMuted] = usePlayerMutedState2();
  const [mediaVolume] = useMediaVolumeState2();
  const volumePropFrame = useFrameForVolumeProp2(loopVolumeCurveBehavior);
  const userPreferredVolume = evaluateVolume2({
    frame: volumePropFrame,
    volume,
    mediaVolume
  });
  if (!videoConfig) {
    throw new Error("No video config found");
  }
  warnAboutTooHighVolume2(userPreferredVolume);
  const effectChainState = useEffectChainState();
  const effectsRef = useRef2(effects);
  effectsRef.current = effects;
  const onErrorRef = useRef2(onError);
  onErrorRef.current = onError;
  const effectChainStateRef = useRef2(effectChainState);
  effectChainStateRef.current = effectChainState;
  const parentSequence = useContext4(SequenceContext2);
  const isPremounting = Boolean(parentSequence?.premounting);
  const isPostmounting = Boolean(parentSequence?.postmounting);
  const sequenceOffset = (parentSequence?.absoluteFrom ?? 0) / videoConfig.fps;
  const currentTime = frame / videoConfig.fps;
  const currentTimeRef = useRef2(currentTime);
  currentTimeRef.current = currentTime;
  const preloadedSrc = usePreload2(src);
  const buffering = useContext4(Internals19.BufferingContextReact);
  if (!buffering) {
    throw new Error("useMediaPlayback must be used inside a <BufferingContext>");
  }
  const effectiveMuted = muted || playerMuted || userPreferredVolume <= 0;
  const isPlayerBuffering = Internals19.useIsPlayerBuffering(buffering);
  const initialPlaying = useRef2(playing && !isPlayerBuffering);
  const initialIsPremounting = useRef2(isPremounting);
  const initialIsPostmounting = useRef2(isPostmounting);
  const initialGlobalPlaybackRate = useRef2(globalPlaybackRate);
  const initialPlaybackRate = useRef2(playbackRate);
  const initialMuted = useRef2(effectiveMuted);
  const initialSequenceDuration = useRef2(videoConfig.durationInFrames);
  const initialSequenceOffset = useRef2(sequenceOffset);
  const hasDrawnRealFrameRef = useRef2(false);
  const isPremountingRef = useRef2(isPremounting);
  isPremountingRef.current = isPremounting;
  useLayoutEffect3(() => {
    if (!_experimentalInitiallyDrawCachedFrame) {
      return;
    }
    const canvas = canvasRef.current;
    if (!canvas) {
      return;
    }
    const cached = getCachedVideoFrame(src);
    if (!cached) {
      return;
    }
    canvas.width = cached.width;
    canvas.height = cached.height;
    const ctx = canvas.getContext("2d", {
      alpha: true,
      desynchronized: true
    });
    if (!ctx) {
      return;
    }
    ctx.drawImage(cached, 0, 0);
  }, [_experimentalInitiallyDrawCachedFrame, src]);
  useLayoutEffect3(() => {
    if (!_experimentalInitiallyDrawCachedFrame) {
      return;
    }
    return () => {
      const canvas = canvasRef.current;
      if (!canvas || !hasDrawnRealFrameRef.current || isPremountingRef.current) {
        return;
      }
      cacheVideoFrame(src, canvas);
    };
  }, [_experimentalInitiallyDrawCachedFrame, src]);
  useEffect2(() => {
    const sharedAudioContextForMediaPlayer = sharedAudioContext?.audioContext && sharedAudioContext.gainNode ? {
      audioContext: sharedAudioContext.audioContext,
      gainNode: sharedAudioContext.gainNode,
      audioSyncAnchor: sharedAudioContext.audioSyncAnchor,
      scheduleAudioNode: sharedAudioContext.scheduleAudioNode,
      unscheduleAudioNode: sharedAudioContext.unscheduleAudioNode
    } : null;
    try {
      const player = new MediaPlayer({
        canvas: canvasRef.current,
        src: preloadedSrc,
        logLevel,
        sharedAudioContext: sharedAudioContextForMediaPlayer,
        loop,
        trimAfter: initialTrimAfterRef.current,
        trimBefore: initialTrimBeforeRef.current,
        fps: videoConfig.fps,
        playbackRate: initialPlaybackRate.current,
        audioStreamIndex,
        debugOverlay,
        bufferState: buffer,
        isPremounting: initialIsPremounting.current,
        isPostmounting: initialIsPostmounting.current,
        globalPlaybackRate: initialGlobalPlaybackRate.current,
        durationInFrames: initialSequenceDuration.current,
        onVideoFrameCallback: initialOnVideoFrameRef.current ?? null,
        playing: initialPlaying.current,
        sequenceOffset: initialSequenceOffset.current,
        credentials,
        requestInit: initialRequestInit,
        tagType: "video",
        getEffects: () => effectsRef.current,
        getEffectChainState: (width, height) => effectChainStateRef.current?.get(width, height)
      });
      mediaPlayerRef.current = player;
      player.initialize(currentTimeRef.current, initialMuted.current).then((result) => {
        if (result.type === "disposed") {
          return;
        }
        const handleError = (error, fallbackMessage) => {
          const [action, errorToUse] = callOnErrorAndResolve({
            onError: onErrorRef.current,
            error,
            disallowFallback: disallowFallbackToOffthreadVideo,
            isClientSideRendering: false,
            clientSideError: error
          });
          if (action === "fail") {
            throw errorToUse;
          }
          Internals19.Log.warn({ logLevel, tag: "@remotion/media" }, fallbackMessage);
          setShouldFallbackToNativeVideo(true);
        };
        if (result.type === "unknown-container-format") {
          handleError(new Error(`Unknown container format ${preloadedSrc}.`), `Unknown container format for ${preloadedSrc} (Supported formats: https://www.remotion.dev/docs/mediabunny/formats), falling back to <OffthreadVideo>`);
          return;
        }
        if (result.type === "network-error") {
          handleError(new Error(`Network error fetching ${preloadedSrc}.`), `Network error fetching ${preloadedSrc}, falling back to <OffthreadVideo>`);
          return;
        }
        if (result.type === "cannot-decode") {
          handleError(new Error(`Cannot decode ${preloadedSrc}.`), `Cannot decode ${preloadedSrc}, falling back to <OffthreadVideo>`);
          return;
        }
        if (result.type === "no-tracks") {
          handleError(new Error(`No video or audio tracks found for ${preloadedSrc}.`), `No video or audio tracks found for ${preloadedSrc}, falling back to <OffthreadVideo>`);
          return;
        }
        if (result.type === "success") {
          setMediaPlayerReady(true);
          setMediaDurationInSeconds(result.durationInSeconds);
          hasDrawnRealFrameRef.current = true;
        }
      }).catch((error) => {
        const [action, errorToUse] = callOnErrorAndResolve({
          onError: onErrorRef.current,
          error,
          disallowFallback: disallowFallbackToOffthreadVideo,
          isClientSideRendering: false,
          clientSideError: error
        });
        if (action === "fail") {
          throw errorToUse;
        }
        Internals19.Log.error({ logLevel, tag: "@remotion/media" }, "[VideoForPreview] Failed to initialize MediaPlayer", errorToUse);
        setShouldFallbackToNativeVideo(true);
      });
    } catch (error) {
      const [action, errorToUse] = callOnErrorAndResolve({
        error,
        onError: onErrorRef.current,
        disallowFallback: disallowFallbackToOffthreadVideo,
        isClientSideRendering: false,
        clientSideError: error
      });
      if (action === "fail") {
        throw errorToUse;
      }
      Internals19.Log.error({ logLevel, tag: "@remotion/media" }, "[VideoForPreview] MediaPlayer initialization failed", errorToUse);
      setShouldFallbackToNativeVideo(true);
    }
    return () => {
      if (mediaPlayerRef.current) {
        Internals19.Log.trace({ logLevel, tag: "@remotion/media" }, `[VideoForPreview] Disposing MediaPlayer`);
        mediaPlayerRef.current.dispose();
        mediaPlayerRef.current = null;
      }
      setMediaPlayerReady(false);
      setShouldFallbackToNativeVideo(false);
      hasDrawnRealFrameRef.current = false;
    };
  }, [
    audioStreamIndex,
    buffer,
    debugOverlay,
    disallowFallbackToOffthreadVideo,
    logLevel,
    loop,
    preloadedSrc,
    sharedAudioContext,
    videoConfig.fps,
    credentials,
    initialRequestInit,
    setMediaDurationInSeconds
  ]);
  warnAboutObjectFitInStyleOrClassName({ style, className, logLevel });
  const classNameValue = useMemo4(() => {
    return [Internals19.OBJECTFIT_CONTAIN_CLASS_NAME, className].filter(Internals19.truthy).join(" ");
  }, [className]);
  useCommonEffects({
    mediaPlayerRef,
    mediaPlayerReady,
    currentTimeRef,
    playing,
    isPlayerBuffering,
    frame,
    trimBefore,
    trimAfter,
    effectiveMuted,
    userPreferredVolume,
    playbackRate,
    globalPlaybackRate,
    fps: videoConfig.fps,
    sequenceOffset,
    loop,
    durationInFrames: videoConfig.durationInFrames,
    isPremounting,
    isPostmounting,
    currentTime,
    logLevel,
    label: "VideoForPreview"
  });
  useLayoutEffect3(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setDebugOverlay(debugOverlay);
  }, [debugOverlay, mediaPlayerReady]);
  useLayoutEffect3(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.setVideoFrameCallback(onVideoFrame ?? null);
  }, [onVideoFrame, mediaPlayerReady]);
  useLayoutEffect3(() => {
    const mediaPlayer = mediaPlayerRef.current;
    if (!mediaPlayer || !mediaPlayerReady) {
      return;
    }
    mediaPlayer.redrawVideoEffects().catch(() => {});
  }, [effects, mediaPlayerReady, mediaPlayerRef]);
  const actualStyle = useMemo4(() => {
    return {
      ...style,
      objectFit: objectFitProp
    };
  }, [objectFitProp, style]);
  if (shouldFallbackToNativeVideo && !disallowFallbackToOffthreadVideo) {
    return /* @__PURE__ */ jsx4(Html5Video, {
      ...props,
      ref: fallbackVideoRef,
      src,
      style: actualStyle,
      className,
      muted,
      volume,
      trimAfter,
      trimBefore,
      playbackRate,
      loopVolumeCurveBehavior,
      name: "<Html5Video> (fallback)",
      loop,
      showInTimeline,
      stack: stack ?? undefined,
      ...fallbackOffthreadVideoProps
    });
  }
  if (headless) {
    return null;
  }
  return /* @__PURE__ */ jsx4("canvas", {
    ...props,
    ref: canvasRefCallback,
    style: actualStyle,
    className: classNameValue
  });
};
var VideoForPreview = (props) => {
  const frame = useCurrentFrame3();
  const videoConfig = useVideoConfig3();
  const currentTime = frame / videoConfig.fps;
  const showShow = useMemo4(() => {
    return getTimeInSeconds({
      unloopedTimeInSeconds: currentTime,
      playbackRate: props.playbackRate,
      loop: props.loop,
      trimBefore: props.trimBefore,
      trimAfter: props.trimAfter,
      mediaDurationInSeconds: Infinity,
      fps: videoConfig.fps,
      ifNoMediaDuration: "infinity",
      src: props.src
    }) !== null;
  }, [
    currentTime,
    props.loop,
    props.playbackRate,
    props.src,
    videoConfig.fps,
    props.trimBefore,
    props.trimAfter
  ]);
  if (!showShow) {
    return null;
  }
  return /* @__PURE__ */ jsx4(VideoForPreviewAssertedShowing, {
    ...props
  });
};

// src/video/video-for-rendering.tsx
import {
  useContext as useContext5,
  useLayoutEffect as useLayoutEffect4,
  useMemo as useMemo5,
  useRef as useRef3,
  useState as useState5
} from "react";
import {
  Internals as Internals20,
  Loop,
  random as random2,
  useCurrentFrame as useCurrentFrame4,
  useDelayRender as useDelayRender2,
  useRemotionEnvironment as useRemotionEnvironment3,
  useVideoConfig as useVideoConfig4
} from "remotion";
import { jsx as jsx5 } from "react/jsx-runtime";
var VideoForRendering = ({
  volume: volumeProp,
  playbackRate,
  src,
  muted,
  loopVolumeCurveBehavior,
  delayRenderRetries,
  delayRenderTimeoutInMilliseconds,
  onVideoFrame,
  logLevel,
  loop,
  style,
  className,
  fallbackOffthreadVideoProps,
  audioStreamIndex,
  disallowFallbackToOffthreadVideo,
  stack,
  toneFrequency,
  trimAfterValue,
  trimBeforeValue,
  headless,
  onError,
  credentials,
  requestInit,
  objectFit: objectFitProp,
  effects,
  ...props
}) => {
  if (!src) {
    throw new TypeError("No `src` was passed to <Video>.");
  }
  const frame = useCurrentFrame4();
  const absoluteFrame = Internals20.useTimelinePosition();
  const { fps } = useVideoConfig4();
  const { registerRenderAsset, unregisterRenderAsset } = useContext5(Internals20.RenderAssetManager);
  const startsAt = Internals20.useMediaStartsAt();
  const sequenceContext = useContext5(Internals20.SequenceContext);
  const id = useMemo5(() => `media-video-${random2(src)}-${sequenceContext?.cumulatedFrom}-${sequenceContext?.relativeFrom}-${sequenceContext?.durationInFrames}`, [
    src,
    sequenceContext?.cumulatedFrom,
    sequenceContext?.relativeFrom,
    sequenceContext?.durationInFrames
  ]);
  const environment = useRemotionEnvironment3();
  const { delayRender, continueRender, cancelRender: cancelRender3 } = useDelayRender2();
  const canvasRef = useRef3(null);
  const [replaceWithOffthreadVideo, setReplaceWithOffthreadVideo] = useState5(false);
  const [initialRequestInit] = useState5(requestInit);
  const audioEnabled = Internals20.useAudioEnabled();
  const videoEnabled = Internals20.useVideoEnabled();
  const maxCacheSize = useMaxMediaCacheSize(logLevel);
  const effectChainState = Internals20.useEffectChainState();
  const [error, setError] = useState5(null);
  if (error) {
    throw error;
  }
  useLayoutEffect4(() => {
    if (!canvasRef.current && !headless) {
      return;
    }
    if (replaceWithOffthreadVideo) {
      return;
    }
    if (!canvasRef.current?.getContext && !headless) {
      return setError(new Error("Canvas does not have .getContext() method available. This could be because <Video> was mounted inside an <svg> tag."));
    }
    const timestamp = frame / fps;
    const durationInSeconds = 1 / fps;
    const newHandle = delayRender(`Extracting frame at time ${timestamp} from ${src}`, {
      retries: delayRenderRetries ?? undefined,
      timeoutInMilliseconds: delayRenderTimeoutInMilliseconds ?? undefined
    });
    const shouldRenderAudio = (() => {
      if (!audioEnabled) {
        return false;
      }
      if (muted) {
        return false;
      }
      return true;
    })();
    extractFrameViaBroadcastChannel({
      src,
      timeInSeconds: timestamp,
      durationInSeconds,
      playbackRate,
      logLevel,
      includeAudio: shouldRenderAudio,
      includeVideo: videoEnabled,
      isClientSideRendering: environment.isClientSideRendering,
      loop,
      audioStreamIndex,
      trimAfter: trimAfterValue,
      trimBefore: trimBeforeValue,
      fps,
      maxCacheSize,
      credentials,
      requestInit: initialRequestInit
    }).then(async (result) => {
      const handleError = (err, clientSideError, fallbackMessage, mediaDurationInSeconds) => {
        if (environment.isClientSideRendering) {
          cancelRender3(clientSideError);
          return;
        }
        const [action, errorToUse] = callOnErrorAndResolve({
          onError,
          error: err,
          disallowFallback: disallowFallbackToOffthreadVideo,
          isClientSideRendering: environment.isClientSideRendering,
          clientSideError: err
        });
        if (action === "fail") {
          cancelRender3(errorToUse);
          return;
        }
        if (window.remotion_isMainTab) {
          Internals20.Log.warn({ logLevel, tag: "@remotion/media" }, fallbackMessage);
        }
        setReplaceWithOffthreadVideo({
          durationInSeconds: mediaDurationInSeconds
        });
      };
      if (result.type === "unknown-container-format") {
        handleError(new Error(`Unknown container format ${src}.`), new Error(`Cannot render video "${src}": Unknown container format. See supported formats: https://www.remotion.dev/docs/mediabunny/formats`), `Unknown container format for ${src} (Supported formats: https://www.remotion.dev/docs/mediabunny/formats), falling back to <OffthreadVideo>`, null);
        return;
      }
      if (result.type === "cannot-decode") {
        handleError(new Error(`Cannot decode ${src}.`), new Error(`Cannot render video "${src}": The video could not be decoded by the browser.`), `Cannot decode ${src}, falling back to <OffthreadVideo>`, result.durationInSeconds);
        return;
      }
      if (result.type === "cannot-decode-alpha") {
        handleError(new Error(`Cannot decode alpha component for ${src}.`), new Error(`Cannot render video "${src}": The alpha channel could not be decoded by the browser.`), `Cannot decode alpha component for ${src}, falling back to <OffthreadVideo>`, result.durationInSeconds);
        return;
      }
      if (result.type === "network-error") {
        handleError(new Error(`Network error fetching ${src}.`), new Error(`Cannot render video "${src}": Network error while fetching the video (possibly CORS).`), `Network error fetching ${src} (no CORS?), falling back to <OffthreadVideo>`, null);
        return;
      }
      const {
        frame: imageBitmap,
        audio,
        durationInSeconds: assetDurationInSeconds
      } = result;
      if (imageBitmap) {
        onVideoFrame?.(imageBitmap);
        const context = canvasRef.current?.getContext("2d", {
          alpha: true
        });
        if (context) {
          context.canvas.width = imageBitmap.width;
          context.canvas.height = imageBitmap.height;
          context.canvas.style.aspectRatio = `${context.canvas.width} / ${context.canvas.height}`;
          context.drawImage(imageBitmap, 0, 0);
          if (effects.length > 0) {
            const completed = await Internals20.runEffectChain({
              state: effectChainState.get(imageBitmap.width, imageBitmap.height),
              source: context.canvas,
              effects,
              output: context.canvas,
              width: imageBitmap.width,
              height: imageBitmap.height
            });
            if (!completed) {
              imageBitmap.close();
              return;
            }
          }
        }
        imageBitmap.close();
      } else if (videoEnabled) {
        const context = canvasRef.current?.getContext("2d", {
          alpha: true
        });
        if (context) {
          context.clearRect(0, 0, context.canvas.width, context.canvas.height);
        }
      }
      const volumePropsFrame = frameForVolumeProp({
        behavior: loopVolumeCurveBehavior,
        loop,
        assetDurationInSeconds: assetDurationInSeconds ?? 0,
        fps,
        frame,
        startsAt
      });
      const volume = Internals20.evaluateVolume({
        volume: volumeProp,
        frame: volumePropsFrame,
        mediaVolume: 1
      });
      Internals20.warnAboutTooHighVolume(volume);
      if (audio && volume > 0) {
        applyVolume(audio.data, volume);
        registerRenderAsset({
          type: "inline-audio",
          id,
          audio: environment.isClientSideRendering ? audio.data : Array.from(audio.data),
          frame: absoluteFrame,
          timestamp: audio.timestamp,
          duration: audio.numberOfFrames / getTargetSampleRate() * 1e6,
          toneFrequency
        });
      }
      continueRender(newHandle);
    }).catch((err) => {
      cancelRender3(err);
    });
    return () => {
      continueRender(newHandle);
      unregisterRenderAsset(id);
    };
  }, [
    absoluteFrame,
    continueRender,
    delayRender,
    delayRenderRetries,
    delayRenderTimeoutInMilliseconds,
    environment.isClientSideRendering,
    fps,
    frame,
    id,
    logLevel,
    loop,
    loopVolumeCurveBehavior,
    muted,
    onVideoFrame,
    playbackRate,
    registerRenderAsset,
    src,
    startsAt,
    unregisterRenderAsset,
    volumeProp,
    replaceWithOffthreadVideo,
    audioStreamIndex,
    disallowFallbackToOffthreadVideo,
    toneFrequency,
    trimAfterValue,
    trimBeforeValue,
    audioEnabled,
    videoEnabled,
    maxCacheSize,
    cancelRender3,
    headless,
    onError,
    credentials,
    effectChainState,
    effects,
    initialRequestInit
  ]);
  warnAboutObjectFitInStyleOrClassName({ style, className, logLevel });
  const classNameValue = useMemo5(() => {
    return [Internals20.OBJECTFIT_CONTAIN_CLASS_NAME, className].filter(Internals20.truthy).join(" ");
  }, [className]);
  const styleWithObjectFit = useMemo5(() => {
    return {
      ...style,
      objectFit: objectFitProp
    };
  }, [objectFitProp, style]);
  if (replaceWithOffthreadVideo) {
    const fallback = /* @__PURE__ */ jsx5(Internals20.InnerOffthreadVideo, {
      ...props,
      src,
      playbackRate: playbackRate ?? 1,
      muted: muted ?? false,
      acceptableTimeShiftInSeconds: fallbackOffthreadVideoProps?.acceptableTimeShiftInSeconds,
      loopVolumeCurveBehavior: loopVolumeCurveBehavior ?? "repeat",
      delayRenderRetries: delayRenderRetries ?? undefined,
      delayRenderTimeoutInMilliseconds: delayRenderTimeoutInMilliseconds ?? undefined,
      name: "Fallback to <OffthreadVideo>",
      style: styleWithObjectFit,
      allowAmplificationDuringRender: true,
      transparent: fallbackOffthreadVideoProps?.transparent ?? true,
      toneMapped: fallbackOffthreadVideoProps?.toneMapped ?? true,
      audioStreamIndex: audioStreamIndex ?? 0,
      className,
      onVideoFrame,
      volume: volumeProp,
      id,
      onError: fallbackOffthreadVideoProps?.onError,
      toneFrequency,
      showInTimeline: false,
      crossOrigin: fallbackOffthreadVideoProps?.crossOrigin,
      onAutoPlayError: fallbackOffthreadVideoProps?.onAutoPlayError ?? null,
      pauseWhenBuffering: fallbackOffthreadVideoProps?.pauseWhenBuffering ?? false,
      trimAfter: trimAfterValue,
      trimBefore: trimBeforeValue,
      useWebAudioApi: fallbackOffthreadVideoProps?.useWebAudioApi ?? false,
      preservePitch: fallbackOffthreadVideoProps?.preservePitch ?? true,
      startFrom: undefined,
      endAt: undefined,
      stack,
      _remotionInternalNativeLoopPassed: false
    });
    if (loop) {
      if (!replaceWithOffthreadVideo.durationInSeconds) {
        const err = new Error(`Cannot render video ${src}: @remotion/media was unable to render, and fell back to <OffthreadVideo>. Also, "loop" was set, but <OffthreadVideo> does not support looping and @remotion/media could also not determine the duration of the video.`);
        cancelRender3(err);
        throw err;
      }
      return /* @__PURE__ */ jsx5(Loop, {
        layout: "none",
        durationInFrames: Internals20.calculateMediaDuration({
          trimAfter: trimAfterValue,
          mediaDurationInFrames: replaceWithOffthreadVideo.durationInSeconds * fps,
          playbackRate,
          trimBefore: trimBeforeValue
        }),
        children: fallback
      });
    }
    return fallback;
  }
  if (headless) {
    return null;
  }
  return /* @__PURE__ */ jsx5("canvas", {
    ...props,
    ref: canvasRef,
    style: styleWithObjectFit,
    className: classNameValue
  });
};

// src/video/video.tsx
import { jsx as jsx6 } from "react/jsx-runtime";
var { validateMediaTrimProps, resolveTrimProps, validateMediaProps: validateMediaProps2 } = Internals21;
var videoSchema = {
  ...Internals21.baseSchema,
  volume: {
    type: "number",
    min: 0,
    max: 20,
    step: 0.01,
    default: 1,
    description: "Volume",
    hiddenFromList: false
  },
  playbackRate: {
    type: "number",
    min: 0.1,
    step: 0.01,
    default: 1,
    description: "Playback rate",
    hiddenFromList: false,
    keyframable: false
  },
  loop: { type: "boolean", default: false, description: "Loop" },
  ...Internals21.transformSchema
};
var InnerVideo = ({
  src,
  audioStreamIndex,
  className,
  delayRenderRetries,
  delayRenderTimeoutInMilliseconds,
  disallowFallbackToOffthreadVideo,
  fallbackOffthreadVideoProps,
  logLevel,
  loop,
  loopVolumeCurveBehavior,
  muted,
  onVideoFrame,
  playbackRate,
  style,
  trimAfter,
  trimBefore,
  volume,
  stack,
  toneFrequency,
  showInTimeline,
  debugOverlay,
  headless,
  onError,
  credentials,
  requestInit,
  controls,
  objectFit,
  _experimentalInitiallyDrawCachedFrame,
  effects,
  setMediaDurationInSeconds,
  refForOutline,
  ...props
}) => {
  const environment = useRemotionEnvironment4();
  if (typeof src !== "string") {
    throw new TypeError(`The \`<Video>\` tag requires a string for \`src\`, but got ${JSON.stringify(src)} instead.`);
  }
  validateMediaTrimProps({
    startFrom: undefined,
    endAt: undefined,
    trimBefore,
    trimAfter
  });
  const { trimBeforeValue, trimAfterValue } = resolveTrimProps({
    startFrom: undefined,
    endAt: undefined,
    trimBefore,
    trimAfter
  });
  validateMediaProps2({ playbackRate, volume }, "Video");
  if (environment.isRendering) {
    return /* @__PURE__ */ jsx6(VideoForRendering, {
      ...props,
      audioStreamIndex: audioStreamIndex ?? 0,
      className,
      delayRenderRetries: delayRenderRetries ?? null,
      delayRenderTimeoutInMilliseconds: delayRenderTimeoutInMilliseconds ?? null,
      disallowFallbackToOffthreadVideo: disallowFallbackToOffthreadVideo ?? false,
      fallbackOffthreadVideoProps,
      logLevel,
      loop,
      loopVolumeCurveBehavior,
      muted,
      onVideoFrame,
      playbackRate,
      src,
      stack,
      style,
      volume,
      toneFrequency,
      trimAfterValue,
      trimBeforeValue,
      headless,
      onError,
      credentials,
      requestInit,
      objectFit,
      effects
    });
  }
  return /* @__PURE__ */ jsx6(VideoForPreview, {
    ...props,
    setMediaDurationInSeconds,
    audioStreamIndex: audioStreamIndex ?? 0,
    className,
    logLevel,
    loop,
    loopVolumeCurveBehavior,
    muted,
    onVideoFrame,
    playbackRate,
    src,
    style,
    volume,
    showInTimeline,
    trimAfter: trimAfterValue,
    trimBefore: trimBeforeValue,
    stack: stack ?? null,
    disallowFallbackToOffthreadVideo,
    fallbackOffthreadVideoProps,
    debugOverlay: debugOverlay ?? false,
    headless: headless ?? false,
    onError,
    credentials,
    requestInit,
    controls,
    objectFit,
    effects,
    _experimentalInitiallyDrawCachedFrame,
    refForOutline
  });
};
var VideoInner = ({
  src,
  audioStreamIndex,
  className,
  delayRenderRetries,
  delayRenderTimeoutInMilliseconds,
  disallowFallbackToOffthreadVideo,
  fallbackOffthreadVideoProps,
  logLevel,
  loop,
  loopVolumeCurveBehavior,
  muted,
  name,
  onVideoFrame,
  playbackRate,
  showInTimeline,
  style,
  trimAfter,
  trimBefore,
  volume,
  stack,
  toneFrequency,
  debugOverlay,
  headless,
  onError,
  credentials,
  requestInit,
  controls,
  objectFit,
  _experimentalInitiallyDrawCachedFrame,
  effects,
  durationInFrames,
  from,
  freeze,
  hidden,
  ...props
}) => {
  const fallbackLogLevel = Internals21.useLogLevel();
  const [mediaVolume] = Internals21.useMediaVolumeState();
  const mediaStartsAt = Internals21.useMediaStartsAt();
  const videoConfig = useVideoConfig5();
  const sequenceDurationInFrames = Math.min(durationInFrames ?? Infinity, Math.max(0, videoConfig.durationInFrames - (from ?? 0)));
  const basicInfo = Internals21.useBasicMediaInTimeline({
    src,
    volume,
    playbackRate: playbackRate ?? 1,
    trimBefore,
    trimAfter,
    sequenceDurationInFrames,
    mediaType: "video",
    displayName: name ?? "<Video>",
    mediaVolume,
    mediaStartsAt,
    loop: loop ?? false
  });
  const [mediaDurationInSeconds, setMediaDurationInSeconds] = useState6(null);
  const loopDisplay = useMemo6(() => getLoopDisplay({
    loop: loop ?? false,
    mediaDurationInSeconds,
    playbackRate: playbackRate ?? 1,
    trimAfter,
    trimBefore,
    sequenceDurationInFrames,
    compFps: videoConfig.fps
  }), [
    loop,
    mediaDurationInSeconds,
    playbackRate,
    trimAfter,
    trimBefore,
    sequenceDurationInFrames,
    videoConfig.fps
  ]);
  const isMedia = useMemo6(() => ({
    type: "video",
    data: basicInfo
  }), [basicInfo]);
  const memoizedEffects = Internals21.useMemoizedEffects({
    effects: effects ?? [],
    overrideId: controls?.overrideId ?? null
  });
  const memoizedEffectDefinitions = Internals21.useMemoizedEffectDefinitions(effects ?? []);
  const refForOutline = React6.useRef(null);
  if (sequenceDurationInFrames === 0) {
    return null;
  }
  return /* @__PURE__ */ jsx6(Sequence2, {
    layout: "none",
    from: from ?? 0,
    durationInFrames: basicInfo.duration,
    freeze,
    _remotionInternalStack: stack,
    _remotionInternalIsMedia: isMedia,
    name: name ?? "<Video>",
    _remotionInternalDocumentationLink: name === undefined ? "https://www.remotion.dev/docs/media/video" : undefined,
    controls,
    _remotionInternalLoopDisplay: loopDisplay,
    _remotionInternalEffects: memoizedEffectDefinitions,
    outlineRef: refForOutline,
    showInTimeline: showInTimeline ?? true,
    hidden,
    children: /* @__PURE__ */ jsx6(InnerVideo, {
      ...props,
      audioStreamIndex: audioStreamIndex ?? 0,
      className,
      delayRenderRetries: delayRenderRetries ?? null,
      delayRenderTimeoutInMilliseconds: delayRenderTimeoutInMilliseconds ?? null,
      disallowFallbackToOffthreadVideo: disallowFallbackToOffthreadVideo ?? false,
      fallbackOffthreadVideoProps: fallbackOffthreadVideoProps ?? {},
      logLevel: logLevel ?? fallbackLogLevel,
      loop: loop ?? false,
      loopVolumeCurveBehavior: loopVolumeCurveBehavior ?? "repeat",
      muted: muted ?? false,
      onVideoFrame,
      playbackRate: playbackRate ?? 1,
      showInTimeline: showInTimeline ?? true,
      src,
      style: style ?? {},
      trimAfter,
      trimBefore,
      volume: volume ?? 1,
      toneFrequency: toneFrequency ?? 1,
      stack,
      debugOverlay: debugOverlay ?? false,
      headless: headless ?? false,
      onError,
      credentials,
      requestInit,
      controls,
      objectFit: objectFit ?? "contain",
      _experimentalInitiallyDrawCachedFrame: _experimentalInitiallyDrawCachedFrame ?? false,
      effects: memoizedEffects,
      setMediaDurationInSeconds,
      refForOutline
    })
  });
};
var Video = Interactive2.withSchema({
  Component: VideoInner,
  componentName: "<Video>",
  componentIdentity: "dev.remotion.media.Video",
  schema: videoSchema,
  supportsEffects: true
});
Internals21.addSequenceStackTraces(Video);

// src/index.ts
var experimental_Audio = Audio;
var experimental_Video = Video;
export {
  getTargetSampleRate,
  experimental_Video,
  experimental_Audio,
  Video,
  AudioForPreview,
  Audio
};
