// @remotion/studio/dist/components/RenderModal/OptionExplainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionExplainer = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const colors_1 = require("../../helpers/colors");
const layout_1 = require("../layout");
const MenuDivider_1 = require("../Menu/MenuDivider");
const CliCopyButton_1 = require("./CliCopyButton");
const container = {
    fontSize: 14,
    paddingTop: 10,
    paddingBottom: 10,
    backgroundColor: colors_1.INPUT_BACKGROUND,
    color: colors_1.WHITE,
};
const padding = {
    paddingLeft: 20,
    paddingRight: 20,
};
const title = {
    fontSize: 14,
};
const DESCRIPTION_FONT_SIZE_PX = 14;
const description = {
    color: colors_1.LIGHT_TEXT,
    fontSize: DESCRIPTION_FONT_SIZE_PX,
    lineHeight: 1.5,
    maxWidth: 400,
};
const link = {
    fontSize: 14,
    maxWidth: 200,
    color: colors_1.BLUE,
    wordWrap: 'break-word',
    textDecoration: 'none',
};
const infoRow = {
    ...padding,
    fontSize: 14,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
};
const infoRowLabel = {
    width: 150,
    fontSize: 14,
    color: colors_1.WHITE,
};
const flexSpacer = {
    display: 'flex',
    flex: 1,
};
const copyWrapper = {
    display: 'flex',
    justifyContent: 'flex-end',
};
const OptionExplainer = ({ option, extraInfoRows = [], showApiOption = true, showCliFlag = true, }) => {
    const shouldShowApiOption = showApiOption && option.ssrName !== null;
    const shouldShowCliFlag = showCliFlag && option.cliFlag !== undefined;
    const hasInfoRows = extraInfoRows.length > 0 || shouldShowCliFlag || shouldShowApiOption;
    return (jsx_runtime_1.jsxs("div", { style: container, className: "__remotion-info-button-container", children: [
            jsx_runtime_1.jsxs("div", { style: padding, children: [
                    jsx_runtime_1.jsxs("div", { children: [
                            jsx_runtime_1.jsx("strong", { style: title, children: option.name }), option.docLink ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                    jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsx("a", { style: link, href: option.docLink, target: "_blank", children: "Docs" })
                                ] })) : null] }), jsx_runtime_1.jsxs("div", { style: description, children: [
                            jsx_runtime_1.jsx("style", { children: `
							.__remotion-option-explainer-description,
							.__remotion-option-explainer-description * {
								color: inherit;
								font-size: ${DESCRIPTION_FONT_SIZE_PX}px;
								line-height: inherit;
							}

							.__remotion-option-explainer-description a {
								color: ${colors_1.BLUE};
								text-decoration: none;
							}

							.__remotion-option-explainer-description code {
								color: ${colors_1.BLUE};
								font-family: monospace;
							}
						` }), jsx_runtime_1.jsx("div", { className: "__remotion-option-explainer-description", children: option.description('ssr') })
                        ] })
                ] }), hasInfoRows ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                    jsx_runtime_1.jsx(layout_1.Spacing, { y: 0.5, block: true }), jsx_runtime_1.jsx(MenuDivider_1.MenuDivider, {}), jsx_runtime_1.jsx(layout_1.Spacing, { y: 0.5, block: true }), jsx_runtime_1.jsxs("div", { children: [extraInfoRows.map((row) => (jsx_runtime_1.jsxs("div", { style: infoRow, children: [
                                    jsx_runtime_1.jsx("div", { style: infoRowLabel, children: row.label }), jsx_runtime_1.jsx("div", { style: flexSpacer }), jsx_runtime_1.jsx("code", { children: row.value }), jsx_runtime_1.jsx("div", { style: copyWrapper, children: jsx_runtime_1.jsx(CliCopyButton_1.CliCopyButton, { valueToCopy: row.value }) })
                                ] }, row.label))), shouldShowCliFlag ? (jsx_runtime_1.jsxs("div", { style: infoRow, children: [
                                    jsx_runtime_1.jsx("div", { style: infoRowLabel, children: "CLI flag" }), jsx_runtime_1.jsx("div", { style: flexSpacer }), jsx_runtime_1.jsxs("code", { children: ["--", option.cliFlag] }), jsx_runtime_1.jsx("div", { style: copyWrapper, children: jsx_runtime_1.jsx(CliCopyButton_1.CliCopyButton, { valueToCopy: option.cliFlag }) })
                                ] })) : null, shouldShowApiOption ? (jsx_runtime_1.jsxs("div", { style: infoRow, children: [
                                    jsx_runtime_1.jsx("div", { style: infoRowLabel, children: "API option" }), jsx_runtime_1.jsx("div", { style: flexSpacer }), jsx_runtime_1.jsx("code", { children: option.ssrName }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 3.75 })
                                ] })) : null, jsx_runtime_1.jsx("div", { style: infoRow })
                        ] })
                ] })) : null] }));
};
exports.OptionExplainer = OptionExplainer;
