// @remotion/studio/dist/components/RenderModal/WebRenderModalAdvanced.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebRenderModalAdvanced = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const Checkmark_1 = require("../../icons/Checkmark");
const Checkbox_1 = require("../Checkbox");
const layout_1 = require("../layout");
const ComboBox_1 = require("../NewComposition/ComboBox");
const layout_2 = require("./layout");
const NumberSetting_1 = require("./NumberSetting");
const OptionExplainerBubble_1 = require("./OptionExplainerBubble");
const tabContainer = {
    flex: 1,
};
const paragraph = {
    margin: 0,
    marginBottom: 8,
};
const lastParagraph = {
    margin: 0,
};
const PageResponsivenessDescription = () => {
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("p", { style: paragraph, children: "The Web Renderer runs in the same browser tab as the Studio. Rendering can block the tab while Remotion captures frames." }), jsx_runtime_1.jsxs("p", { style: paragraph, children: ["The default is ",
                    jsx_runtime_1.jsx("code", { children: "Medium" }),
                    ", which tries to free the event loop every 33ms of rendering work so progress updates and UI interactions can stay responsive."] }), jsx_runtime_1.jsxs("p", { style: lastParagraph, children: ["Choose ",
                    jsx_runtime_1.jsx("code", { children: "Disabled" }),
                    " to prioritize render speed, or", ' ', jsx_runtime_1.jsx("code", { children: "High" }),
                    " to give the browser more chances to update."] })
        ] }));
};
const WebRenderModalAdvanced = ({ renderMode, delayRenderTimeout, setDelayRenderTimeout, mediaCacheSizeInBytes, setMediaCacheSizeInBytes, hardwareAcceleration, setHardwareAcceleration, allowHtmlInCanvas, setAllowHtmlInCanvas, pageResponsiveness, setPageResponsiveness, }) => {
    const toggleCustomMediaCacheSizeInBytes = (0, react_1.useCallback)(() => {
        setMediaCacheSizeInBytes((previous) => {
            if (previous === null) {
                return 1000 * 1000 * 1000;
            }
            return null;
        });
    }, [setMediaCacheSizeInBytes]);
    const toggleAllowHtmlInCanvas = (0, react_1.useCallback)(() => {
        setAllowHtmlInCanvas((prev) => !prev);
    }, [setAllowHtmlInCanvas]);
    const changeMediaCacheSizeInBytes = (0, react_1.useCallback)((cb) => {
        setMediaCacheSizeInBytes((prev) => {
            if (prev === null) {
                throw new TypeError('Expected previous value');
            }
            if (typeof cb === 'function') {
                return cb(prev);
            }
            return cb;
        });
    }, [setMediaCacheSizeInBytes]);
    const hardwareAccelerationOptions = (0, react_1.useMemo)(() => {
        return [
            {
                label: 'No Preference',
                onClick: () => setHardwareAcceleration('no-preference'),
                leftItem: hardwareAcceleration === 'no-preference' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'no-preference',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'no-preference',
            },
            {
                label: 'Prefer Hardware',
                onClick: () => setHardwareAcceleration('prefer-hardware'),
                leftItem: hardwareAcceleration === 'prefer-hardware' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'prefer-hardware',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'prefer-hardware',
            },
            {
                label: 'Prefer Software',
                onClick: () => setHardwareAcceleration('prefer-software'),
                leftItem: hardwareAcceleration === 'prefer-software' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'prefer-software',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'prefer-software',
            },
        ];
    }, [hardwareAcceleration, setHardwareAcceleration]);
    const selectedPageResponsiveness = typeof pageResponsiveness === 'number' ? 'custom' : pageResponsiveness;
    const pageResponsivenessOptions = (0, react_1.useMemo)(() => {
        return [
            {
                label: 'Disabled',
                onClick: () => setPageResponsiveness('disabled'),
                leftItem: selectedPageResponsiveness === 'disabled' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'disabled',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'disabled',
            },
            {
                label: 'Low (100ms)',
                onClick: () => setPageResponsiveness('low'),
                leftItem: selectedPageResponsiveness === 'low' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'low',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'low',
            },
            {
                label: 'Medium (33ms)',
                onClick: () => setPageResponsiveness('medium'),
                leftItem: selectedPageResponsiveness === 'medium' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'medium',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'medium',
            },
            {
                label: 'High (16ms)',
                onClick: () => setPageResponsiveness('high'),
                leftItem: selectedPageResponsiveness === 'high' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'high',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'high',
            },
            {
                label: 'Custom',
                onClick: () => setPageResponsiveness((previous) => typeof previous === 'number' ? previous : 33),
                leftItem: selectedPageResponsiveness === 'custom' ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                id: 'custom',
                keyHint: null,
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'custom',
            },
        ];
    }, [selectedPageResponsiveness, setPageResponsiveness]);
    const setCustomPageResponsiveness = (0, react_1.useCallback)((value) => {
        setPageResponsiveness((previous) => {
            const currentValue = typeof previous === 'number' ? previous : 33;
            return typeof value === 'function' ? value(currentValue) : value;
        });
    }, [setPageResponsiveness]);
    return (jsx_runtime_1.jsxs("div", { style: tabContainer, children: [
            jsx_runtime_1.jsx(NumberSetting_1.NumberSetting, { name: "Delay Render Timeout", formatter: (v) => `${v}ms`, min: 0, max: 1000000000, step: 1000, value: delayRenderTimeout, onValueChanged: setDelayRenderTimeout, hint: "delayRenderTimeoutInMillisecondsOption" }), jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                    jsx_runtime_1.jsxs("div", { style: layout_2.label, children: ["Custom @remotion/media cache size ",
                            jsx_runtime_1.jsx(layout_1.Spacing, { x: 0.5 }), jsx_runtime_1.jsx(OptionExplainerBubble_1.WebRendererOptionExplainerBubble, { apiName: "mediaCacheSizeInBytes", id: "mediaCacheSizeInBytesOption" })
                        ] }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsx(Checkbox_1.Checkbox, { checked: mediaCacheSizeInBytes !== null, onChange: toggleCustomMediaCacheSizeInBytes, name: "media-cache-size" }) })
                ] }), mediaCacheSizeInBytes === null ? null : (jsx_runtime_1.jsx(NumberSetting_1.NumberSetting, { name: "@remotion/media cache size", formatter: (w) => `${w} bytes`, min: 0, max: 10000000000, step: 10 * 1024 * 1024, value: mediaCacheSizeInBytes, onValueChanged: changeMediaCacheSizeInBytes })), renderMode === 'video' ? (jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Hardware Acceleration" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsx(ComboBox_1.Combobox, { values: hardwareAccelerationOptions, selectedId: hardwareAcceleration, title: "Hardware Acceleration" }) })
                ] })) : null, jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                    jsx_runtime_1.jsxs("div", { style: layout_2.label, children: ["Allow HTML-in-canvas ",
                            jsx_runtime_1.jsx(layout_1.Spacing, { x: 0.5 }), jsx_runtime_1.jsx(OptionExplainerBubble_1.WebRendererOptionExplainerBubble, { apiName: "allowHtmlInCanvas", id: "allowHtmlInCanvasOption" })
                        ] }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsx(Checkbox_1.Checkbox, { checked: allowHtmlInCanvas, onChange: toggleAllowHtmlInCanvas, name: "allow-html-in-canvas" }) })
                ] }), renderMode === 'still' ? null : (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                    jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                            jsx_runtime_1.jsxs("div", { style: layout_2.label, children: ["Page Responsiveness ",
                                    jsx_runtime_1.jsx(layout_1.Spacing, { x: 0.5 }), jsx_runtime_1.jsx(OptionExplainerBubble_1.WebRendererCustomOptionExplainerBubble, { apiName: "pageResponsiveness", description: jsx_runtime_1.jsx(PageResponsivenessDescription, {}), docLink: "https://www.remotion.dev/docs/client-side-rendering/page-responsiveness", name: "Page responsiveness" })
                                ] }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsx(ComboBox_1.Combobox, { values: pageResponsivenessOptions, selectedId: selectedPageResponsiveness, title: "Page Responsiveness" }) })
                        ] }), typeof pageResponsiveness === 'number' ? (jsx_runtime_1.jsx(NumberSetting_1.NumberSetting, { name: "Responsiveness Interval", formatter: (v) => `${v}ms`, min: 1, max: 1000000000, step: 1, value: pageResponsiveness, onValueChanged: setCustomPageResponsiveness })) : null] }))] }));
};
exports.WebRenderModalAdvanced = WebRenderModalAdvanced;
