// @remotion/studio/dist/components/selected-outline-snap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findSelectedOutlineSnap = exports.getSelectedOutlineSnapTargets = exports.selectedOutlineSnapIndicatorColor = exports.selectedOutlineSnapThresholdPx = void 0;
exports.selectedOutlineSnapThresholdPx = 10;
exports.selectedOutlineSnapIndicatorColor = '#ff00ff';
const EPSILON = 0.000001;
const getSelectedOutlineSnapTargets = ({ compositionHeight, compositionWidth, guides, }) => {
    return [
        { axis: 'x', position: 0, type: 'canvas-left' },
        {
            axis: 'x',
            position: compositionWidth / 2,
            type: 'canvas-horizontal-center',
        },
        { axis: 'x', position: compositionWidth, type: 'canvas-right' },
        { axis: 'y', position: 0, type: 'canvas-top' },
        {
            axis: 'y',
            position: compositionHeight / 2,
            type: 'canvas-vertical-center',
        },
        { axis: 'y', position: compositionHeight, type: 'canvas-bottom' },
        ...guides.flatMap((guide) => {
            if (!guide.show) {
                return [];
            }
            return [
                guide.orientation === 'vertical'
                    ? {
                        axis: 'x',
                        position: guide.position,
                        type: 'guide-vertical',
                    }
                    : {
                        axis: 'y',
                        position: guide.position,
                        type: 'guide-horizontal',
                    },
            ];
        }),
    ];
};
exports.getSelectedOutlineSnapTargets = getSelectedOutlineSnapTargets;
const getSelectedOutlineBounds = ({ deltaX, deltaY, outlines, scale, }) => {
    if (outlines.length === 0) {
        return null;
    }
    const points = outlines.flatMap((outline) => outline.points);
    const left = Math.min(...points.map((point) => point.x)) / scale + deltaX;
    const right = Math.max(...points.map((point) => point.x)) / scale + deltaX;
    const top = Math.min(...points.map((point) => point.y)) / scale + deltaY;
    const bottom = Math.max(...points.map((point) => point.y)) / scale + deltaY;
    return {
        left,
        right,
        top,
        bottom,
        centerX: left + (right - left) / 2,
        centerY: top + (bottom - top) / 2,
    };
};
const canSnapEdgeToTarget = (edge, target) => {
    if (target.type === 'canvas-horizontal-center') {
        return edge === 'center-x';
    }
    if (target.type === 'canvas-vertical-center') {
        return edge === 'center-y';
    }
    if (target.type === 'canvas-left' || target.type === 'canvas-right') {
        return edge === 'left' || edge === 'right';
    }
    if (target.type === 'guide-vertical') {
        return edge === 'left' || edge === 'right' || edge === 'center-x';
    }
    if (target.type === 'canvas-top' || target.type === 'canvas-bottom') {
        return edge === 'top' || edge === 'bottom';
    }
    return edge === 'top' || edge === 'bottom' || edge === 'center-y';
};
const findBestSnapForAxis = ({ edges, targets, threshold, }) => {
    var _a;
    let bestDistance = Infinity;
    let bestSnaps = [];
    for (const target of targets) {
        for (const edge of edges) {
            if (!canSnapEdgeToTarget(edge.edge, target)) {
                continue;
            }
            const distance = Math.abs(edge.position - target.position);
            if (distance > threshold || distance > bestDistance + EPSILON) {
                continue;
            }
            if (distance < bestDistance - EPSILON) {
                bestSnaps = [];
                bestDistance = distance;
            }
            bestSnaps.push({
                offset: target.position - edge.position,
                snapPoint: {
                    distance,
                    edge: edge.edge,
                    target,
                },
            });
        }
    }
    if (bestSnaps.length > 1) {
        const centerSnap = bestSnaps.find((snap) => snap.snapPoint.target.type === 'canvas-horizontal-center' ||
            snap.snapPoint.target.type === 'canvas-vertical-center');
        if (centerSnap) {
            return centerSnap;
        }
    }
    return (_a = bestSnaps[0]) !== null && _a !== void 0 ? _a : null;
};
const findSelectedOutlineSnap = ({ allowX, allowY, deltaX, deltaY, outlines, scale, targets, }) => {
    var _a, _b, _c, _d;
    const bounds = getSelectedOutlineBounds({ deltaX, deltaY, outlines, scale });
    if (bounds === null) {
        return {
            activeSnapPoints: [],
            snapOffsetX: null,
            snapOffsetY: null,
        };
    }
    const threshold = exports.selectedOutlineSnapThresholdPx / scale;
    const snapX = allowX
        ? findBestSnapForAxis({
            edges: [
                { edge: 'left', position: bounds.left },
                { edge: 'center-x', position: bounds.centerX },
                { edge: 'right', position: bounds.right },
            ],
            targets: targets.filter((target) => target.axis === 'x'),
            threshold,
        })
        : null;
    const snapY = allowY
        ? findBestSnapForAxis({
            edges: [
                { edge: 'top', position: bounds.top },
                { edge: 'center-y', position: bounds.centerY },
                { edge: 'bottom', position: bounds.bottom },
            ],
            targets: targets.filter((target) => target.axis === 'y'),
            threshold,
        })
        : null;
    return {
        activeSnapPoints: [
            (_a = snapX === null || snapX === void 0 ? void 0 : snapX.snapPoint) !== null && _a !== void 0 ? _a : null,
            (_b = snapY === null || snapY === void 0 ? void 0 : snapY.snapPoint) !== null && _b !== void 0 ? _b : null,
        ].filter((snapPoint) => snapPoint !== null),
        snapOffsetX: (_c = snapX === null || snapX === void 0 ? void 0 : snapX.offset) !== null && _c !== void 0 ? _c : null,
        snapOffsetY: (_d = snapY === null || snapY === void 0 ? void 0 : snapY.offset) !== null && _d !== void 0 ? _d : null,
    };
};
exports.findSelectedOutlineSnap = findSelectedOutlineSnap;
